---
title: "pja_ipmn_wta"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pja_ipmn_wta}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Setup

## Library imports

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(tidyverse)
library(readxl)
library(writexl)
library(patchwork)
library(ggrepel)
library(ggridges)
library(plotly)

# bioinformatics packages
library(umap)
library(pheatmap)

# bioconductor packages
library(factoextra)
library(limma)
library(fgsea)
library(GSVA)

# link to stgeomx
library(devtools)
stgeomx_path <- '/Users/mkiyer/proj/stgeomx'
devtools::load_all(path=stgeomx_path)
library(stgeomx)

```

## Input files and parameter settings

```{r input files, include=FALSE}

# master directory
working_dir <- "/Users/mkiyer/Dropbox (University of Michigan)/manuscripts/aso_stgeomx_ipmn_wta/analysis"

# input directories
geomx_dir <- file.path(working_dir, "geomx")
data_dir <- file.path(working_dir, "external_data")

# ipmn dataset and annotations
pja_ipmn_geomx_xlsx <- file.path(geomx_dir, "pja_ipmn_wta_2022-12-02.xlsx")
pja_ipmn_geomx_annot_xlsx <- file.path(geomx_dir, "pja_ipmn_wta_annot_2023-05-03.xlsx")

# external resource files
hpa_path_file <- file.path(data_dir, "hpa_pathology.tsv")
cptac_panc_file <- file.path(data_dir, "cptac_panc_supp_mmc3.xlsx")
pdacr_gene_list_file <- file.path(data_dir, "pdacR_gene_lists.rds")

# qc parameters
geomx_negprobe_lod_quantile = 0.9
geomx_min_counts = 20000
geomx_min_auc = 0.60
geomx_aoi_min_frac_expr = 0.2
geomx_gene_min_frac_expr = 0.1

# normalization
bgcorrect_method = "qq"
bgcorrect_bw_adjust = 2
bgcorrect_bg_quant = 0.5
norm_method = "qn"

# de parameters
de_padj_cutoff <- 0.05
de_log2fc_cutoff <- 1

# gsea params
gsea_padj_cutoff <- 0.05
gsea_log2fc_cutoff <- 1

# gsva params
gsva_tau <- 0.25

# results
results_dir <- file.path(working_dir, "results")
plot_dir <- file.path(results_dir, "plots")
gsea_plot_dir <- file.path(results_dir, "gsea_plots")

for (d in c(results_dir, plot_dir, gsea_plot_dir)) {
  if (!dir.exists(d)) {
    dir.create(d)
  }
}

```

## Color scales for plots

```{r color scales, include=FALSE}

color_scales = list(
  cluster = c(c1="#fc6a03", c2="#63c5da", c3="#710193"),
  cluster_name = c(cHR="#fc6a03", cLR="#63c5da", cNL="#710193"),
  cluster_nllgd_name = c(cHR="#fc6a03", cLR="#63c5da", cNL="#710193"),
  path_aoi = c(nl="#0044ff", lgd="#00ffff", hgd="#ffff00", inv="#ff0000"),
  path_specimen = c(ipmn_lgd="#00ffff", ipmn_inv="#cc0000"),
  path = c(ipmn_lgd_nl_epithelial ="#0044ff", 
           ipmn_lgd_lgd_epithelial ="#00ffff", 
           ipmn_inv_nl_epithelial ="#ae00ff", 
           ipmn_inv_lgd_epithelial ="#42ff00", 
           ipmn_inv_hgd_epithelial ="#ffff00", 
           ipmn_inv_inv_epithelial ="#ff0000"),
  de_cptac_rna = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"), 
  de_cptac_prot = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"),
  prognostic = c(no="#aaaaaa", fav="#00FFFF", unfav="#FFFF00", na="#FFFFFF"),
  moffitt_basallike = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"),
  moffitt_classical = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"),
  moffitt_exocrine = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF")
)

shape_scales <- list(
  path_specimen = c(ipmn_inv=16, ipmn_lgd=17)
)

```

## Helper functions

```{r helper functions}

plot_gene <- function(s, x, gene, aes_x, aes_color) {
  exprs <- unlist(x[gene, ])
  x <- s %>% add_column(gene = exprs)
  p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
    geom_boxplot(outlier.shape = NA, width=0.75) +
#    geom_violin(scale="width", width=0.5) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
    ggtitle(paste0("Gene: ", gene)) +
    theme_minimal() + 
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
    theme(axis.line = element_line(color = "black"))
  return(p)
}

plot_gene2 <- function(s, x, gene, aes_x, aes_color) {
  exprs <- unlist(x[gene, ])
  x <- s %>% add_column(gene = exprs)
  p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
    geom_boxplot(outlier.shape=NA, width=0.75) +
    geom_point(position=position_jitterdodge(jitter.width=0.1), 
               aes(group={{aes_color}}), pch=21) +
    ggtitle(paste0("Gene: ", gene)) +
    theme_bw() + 
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
    theme(axis.line = element_line(color = "black"))
  return(p)
}

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```


# Read and process geomx data

```{r ipmn wta geomx data, include=FALSE}

# read data
ds <- stgeomx::read_xlsx(pja_ipmn_geomx_xlsx)
# read and join sample annotation
annot <- read_excel(pja_ipmn_geomx_annot_xlsx)
ds$samples <- ds$samples %>% 
  inner_join(annot, by="SegmentDisplayName")

# subset epithelial segment
s <- ds$samples %>% filter(segment == "panck")
ds <- stgeomx::init(samples = s,
                    meta = ds$meta, 
                    counts = ds$counts[, s$aoi],
                    x = ds$counts[, s$aoi])

# process dataset
ds <- stgeomx::preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx::merge_probes(ds)

# set filtering thresholds
ds <- stgeomx::set_thresholds(ds,
                              min_counts = geomx_min_counts,
                              min_auc = geomx_min_auc,
                              aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                              gene_min_frac_expr = geomx_gene_min_frac_expr)

table(ds$samples$path)

```

## QC plots

```{r ipmn qc, include=FALSE}

# nuclei count
summary(ds$samples$AOINucleiCount)

# read counts
summary(ds$samples$num_counts)

# density
ds$samples$AOIDensity <- ds$samples$AOINucleiCount / ds$samples$AOISurfaceArea

# figure: scatter plot of nuclei count vs total counts
x <- cor.test(ds$samples$AOINucleiCount, ds$samples$num_counts)
corlab <- paste0("r = ", round(x$estimate, 2))
p <- ggplot(ds$samples, aes(AOINucleiCount, num_counts, color=path)) + 
  geom_point() +
  geom_smooth(method = 'lm', 
              formula = y ~ x, 
              se = FALSE, 
              color = "black", 
              linetype = "dashed") +
  scale_x_log10() +
  scale_y_log10() +
  scale_color_manual(values = color_scales$path) +
  labs(x="Nuclei Count", y="Total Counts") +
  annotate("text", x=500, y=1e4, label = corlab) +
  theme_bw()
p
f <- file.path(plot_dir, "qc_scatter_nuclei_vs_counts_bypath.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 3)

# figure: scatter plot of surface area versus total counts
x <- cor.test(ds$samples$AOISurfaceArea, ds$samples$num_counts)
corlab <- paste0("r = ", round(x$estimate, 2))
p <- ggplot(ds$samples, aes(AOISurfaceArea, num_counts, color=path)) + 
  geom_point() +
  geom_smooth(method = 'lm', 
              formula = y ~ x, 
              se = FALSE, 
              color = "black", 
              linetype = "dashed") +
  scale_x_log10() +
  scale_y_log10() +
  scale_color_manual(values = color_scales$path) +
  labs(x="Surface Area", y="Total Counts") +
  annotate("text", x=1e5, y=1e4, label = corlab) +
  theme_bw()
p
f <- file.path(plot_dir, "qc_scatter_surfacearea_vs_counts_bypath.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 3)


# figure: boxplot slide vs num_counts
p <- ggplot(ds$samples, aes(slide, num_counts, fill=path_specimen)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(aes(color=path_aoi), size=2) +
  geom_point(pch=21, color="black", fill="#ff000000", size=2) +
#  geom_point(aes(color=path_aoi), size=2, position=position_jitterdodge(jitter.width = 0, dodge.width = 0.5)) +
  scale_y_log10() +
  scale_color_manual(values = color_scales$path_aoi) +
  scale_fill_manual(values = color_scales$path_specimen) +
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  labs(x="Slide", y="Total Counts", color="AOI Pathology", fill="Specimen Pathology")
p
f <- file.path(plot_dir, "qc_boxplot_slide_vs_num_counts.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 5)


# figure: boxplot slide vs auc
p <- ggplot(ds$samples, aes(slide, bg_auc, fill=path_specimen)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(aes(color=path_aoi), size=2) +
  geom_point(pch=21, color="black", fill="#ff000000", size=2) +
#  geom_point(position=position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
#  scale_y_log10() +
  scale_color_manual(values = color_scales$path_aoi) +
  scale_fill_manual(values = color_scales$path_specimen) +
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  labs(x="Slide", y="snAUC", color="AOI Pathology", fill="Specimen Pathology")
p
f <- file.path(plot_dir, "qc_boxplot_slide_vs_auc.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 5)


# figure: boxplot path vs counts
p <- ggplot(ds$samples, aes(path, num_counts, fill=path)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
  scale_y_log10() +
  scale_fill_manual(values = color_scales$path) +
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1))
p
f <- file.path(plot_dir, "qc_boxplot_path_vs_num_counts.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 5)

# figure: boxplot path vs auc
p <- ggplot(ds$samples, aes(path, bg_auc, fill=path)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
  scale_fill_manual(values = color_scales$path) +
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1))
p
f <- file.path(plot_dir, "qc_boxplot_path_vs_auc.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 5)

# counts vs auc (path)
p <- ggplot(ds$samples, aes(num_counts, bg_auc, color=path)) +
  geom_point() +
  scale_x_log10() +
  scale_color_manual(values = color_scales$path) +
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  labs(x="Total Counts", y="snAUC", color="AOI Pathology")
p
f <- file.path(plot_dir, "qc_scatter_counts_vs_auc_bypath.pdf")
ggsave(f, plot=p, device="pdf", width = 6, height = 4)

# figure: frac expressed cutoff
p <- ggplot(ds$meta, aes(x=frac_expr)) +
  geom_histogram(binwidth=0.05, color="black", fill="grey") +
  geom_vline(xintercept=geomx_gene_min_frac_expr, linetype="dashed", color="red") +
  theme_bw() +
  xlab("Fraction Expressed Above Background") +
  ylab("# of Probes") +
  facet_wrap(~ bg, scales = "free_y")
p
f <- file.path(plot_dir, "histogram_frac_expr.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 4)


```

## QC AOI and gene filtering

```{r qc filter aoi}

p <- stgeomx::plot_aoi_filter(ds, 
                              min_counts=geomx_min_counts,
                              min_auc=geomx_min_auc,
                              min_frac_expr=geomx_aoi_min_frac_expr)
p
w <- 9
h <- 4
f <- file.path(plot_dir, "qc_aoi_filtering")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

p <- stgeomx::plot_gene_filter(ds, geomx_gene_min_frac_expr)
p
w <- 7
h <- 7
f <- file.path(plot_dir, "qc_gene_filtering")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

```

## BG correction and normalization

```{r background correction}

# background correction must use original unfiltered matrix
fds <- stgeomx::bgcorrect(ds, method=bgcorrect_method, 
                          bw.adjust=bgcorrect_bw_adjust, 
                          bg.quant=bgcorrect_bg_quant)
# normalize
fds <- stgeomx::normalize(fds, norm_method)

# figure: expression distribution
p <- stgeomx::plot_expr_dist(fds)
p
w <- 8
h <- 8
f <- file.path(plot_dir, "qc_expr_dist")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

# figure: background correction
p <- stgeomx::plot_bgcorrect2(fds)
p
w <- 8
h <- 8
f <- file.path(plot_dir, "qc_bgcorrect")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)


# apply filters to permanently remove aois/genes
# fds <- stgeomx::apply_filters(fds)

# background correction must use original unfiltered matrix
fds <- stgeomx::bgcorrect(ds, method=bgcorrect_method,
                          bw.adjust=bgcorrect_bw_adjust,
                          bg.quant=bgcorrect_bg_quant)
# apply filters before normalizing
fds <- stgeomx::apply_filters(fds)
# normalize
fds <- stgeomx::normalize(fds, norm_method)

# filtering stats
nrow(ds$samples)
nrow(fds$samples)
table(ds$meta$keep)

# final normalized gene expression matrix
log2_norm_cpm <- log2(fds$x + 1)

```
# Color scales for patients

```{r patient color scales}

# set patient color scales
x <- pals::cols25(length(unique(fds$samples$patient)))
color_scales$patient = setNames(x, sort(unique(fds$samples$patient)))


```

# Read external datasets

```{r read external data}

get_top_gene_sets <- function(tbl, log2fc_cutoff, padj_cutoff) {
  ntopgenes <- c(25, 50, 100, 250)
  gs <- list()
  for (a in unique(tbl$analysis)) {
    up_genes <- tbl %>% filter(
      analysis == a,
      log2fc >= log2fc_cutoff,
      fdr < padj_cutoff
    )
    dn_genes <- tbl %>% filter(
      analysis == a,
      log2fc <= -log2fc_cutoff,
      fdr < padj_cutoff
    )
    gs_name <- paste0(toupper(a), "_ALL_UP")
    gs[[gs_name]] <- pull(up_genes, gene_symbol)
    gs_name <- paste0(toupper(a), "_ALL_DN")
    gs[[gs_name]] <- pull(dn_genes, gene_symbol)

    for (n in ntopgenes) {
      x <- up_genes %>% slice_max(log2fc, n=n) %>% pull(gene_symbol)
      if (length(x) < n) { next }
      gs_name <- paste0(toupper(a), "_TOP", n, "_UP")
      gs[[gs_name]] <- x
      
      x <- dn_genes %>% slice_min(log2fc, n=n) %>% pull(gene_symbol)
      if (length(x) < n) { next }
      gs_name <- paste0(toupper(a), "_TOP", n, "_DN")
      gs[[gs_name]] <- x 
    }
  }
  return(gs)
}

read_cptac_pdac_data <- function(filename, log2fc_cutoff, padj_cutoff) {
  rna <- read_excel(filename, sheet="Diff_exp_RNA")
  rna <- rna %>%
    dplyr::rename(gene_symbol = GeneSymbol,
                  log2fc = MedianLog2FoldChange,
                  pval = `p value`,
                  fdr = FDR) %>%
    dplyr::mutate(analysis = "cptac_rna", 
                  de = case_when(fdr > padj_cutoff ~ "no",
                                 abs(log2fc) <= log2fc_cutoff ~ "no",
                                 log2fc < -log2fc_cutoff ~ "dn",
                                 log2fc > log2fc_cutoff ~ "up",
                                 TRUE ~ "no"))
  
  # GeneSymbol	MedianLog2FoldChange	p value	FDR	
  # MedianLog2FoldChange.duct	p value.duct	FDR.duct	
  # Expression Coefficient	p value of adjusted expression	FDR of adjusted expression
  # add additional data?
  prot <- read_excel(cptac_panc_file, sheet="Diff_exp_prot")
  prot <- prot %>%
    dplyr::rename(
      gene_symbol = GeneSymbol,
      log2fc = MedianLog2FoldChange,
      pval = `p value`,
      fdr = FDR
    ) %>%
    select(gene_symbol, log2fc, pval, fdr) %>%
    mutate(analysis = "cptac_prot",
           de = case_when(fdr > padj_cutoff ~ "no",
                          abs(log2fc) <= log2fc_cutoff ~ "no",
                          log2fc < -log2fc_cutoff ~ "dn",
                          log2fc > log2fc_cutoff ~ "up",
                          TRUE ~ "no"))
  return(bind_rows(rna, prot))
}


read_hpa_data <- function(hpa_path_file) {
  x <- read.table(hpa_path_file, header=TRUE, sep="\t")
  colnames(x) <- c("ens_gene_id", "gene_symbol", "cancer_type",
                   "ihc_high", "ihc_medium", "ihc_low", "ihc_not_detected",
                   "prog_fav", "unprog_fav", "prog_unfav", "unprog_unfav")
  x <- as_tibble(x) %>%
    distinct(gene_symbol, cancer_type, .keep_all=TRUE) %>%
    mutate(cancer_type = str_replace(cancer_type, " ", "_")) %>%
    mutate(prognostic = case_when(!is.na(prog_fav) ~ "fav",
                                  !is.na(prog_unfav) ~ "unfav",
                                  TRUE ~ "no")) %>%
    select(ens_gene_id, gene_symbol, cancer_type, prognostic, prog_fav, prog_unfav)
  return(x)
}

get_hpa_gene_sets <- function(hpa) {
  gs = list()
  for (catype in unique(hpa$cancer_type)) {
    unfav <- filter(hpa, cancer_type == catype, prognostic == "unfav")  %>% select(gene_symbol) %>% pull()
    fav <- filter(hpa, cancer_type == catype, prognostic == "fav")  %>% select(gene_symbol) %>% pull()
    unfav_name <- paste0("HPA_", toupper(catype), "_UNFAV_PROGNOSIS")
    fav_name <- paste0("HPA_",  toupper(catype), "_FAV_PROGNOSIS")
    x <- setNames(list(unfav, fav), c(unfav_name, fav_name))
    gs <- append(gs, x)
  }
  gs
}

read_msigdb_json_file <- function(filename) {
  x <- jsonlite::fromJSON(filename)
  y <- tibble(gs_name = names(x), gs = x) %>% 
    unnest_wider(gs) %>%
    unnest_longer(geneSymbols) %>%
    select(gs_cat = collection,
           gs_name = gs_name,
           gene_symbol = geneSymbols)
}

# read cptac data
cptac_de_data <- read_cptac_pdac_data(cptac_panc_file, gsea_log2fc_cutoff, gsea_padj_cutoff)
cptac_de_data <- filter(cptac_de_data, gene_symbol %in% fds$meta$gene)
# pivot to merge protein/rna data
cptac_de_merged <- cptac_de_data %>%
  pivot_wider(names_from = analysis, values_from = c(log2fc, pval, fdr, de))
# get gene sets
gs_cptac <- get_top_gene_sets(cptac_de_data, gsea_log2fc_cutoff, gsea_padj_cutoff)

# read hpa data
hpa <- read_hpa_data(hpa_path_file)
hpa <- filter(hpa, gene_symbol %in% fds$meta$gene)
hpa_pdac <- filter(hpa, cancer_type == "pancreatic_cancer")
# get gene sets
gs_hpa <- get_hpa_gene_sets(hpa)
gs_hpa_panc <- gs_hpa[c("HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS", 
                        "HPA_PANCREATIC_CANCER_FAV_PROGNOSIS")]


# read msigdb data
f <- file.path(data_dir, "msigdb.v2023.2.Hs.json")
msigdb_gene_sets <- read_msigdb_json_file(f)
# determine gene universe
msigdb_genes <- intersect(fds$meta$gene, msigdb_gene_sets$gene_symbol)
# filter genes not in our dataset
msigdb_gene_sets <- msigdb_gene_sets %>%
  dplyr::filter(gene_symbol %in% fds$meta$gene) %>%
  distinct(gs_cat, gs_name, gene_symbol, .keep_all = TRUE)

# gene set categories
gs_hallmark <- dplyr::filter(msigdb_gene_sets, gs_cat == "H")
gs_hallmark <- split(x = gs_hallmark$gene_symbol, f = gs_hallmark$gs_name)
gs_gobp <- dplyr::filter(msigdb_gene_sets, gs_cat == "C5:GO:BP")
gs_gobp <- split(x = gs_gobp$gene_symbol, f = gs_gobp$gs_name)
gs_3ca <- dplyr::filter(msigdb_gene_sets, gs_cat == "C4:3CA")
gs_3ca <- split(x = gs_3ca$gene_symbol, f = gs_3ca$gs_name)

# pancreatic cancer subtypes gene lists
x <- readRDS(pdacr_gene_list_file)
gs_pdacr <- list()
for (gs_name in names(x)) {
  genes <- x[[gs_name]]
  if (typeof(genes) == "character") {
    genes <- intersect(x[[gs_name]], fds$meta$gene)
    gs_pdacr[[gs_name]] <- genes
  } else {
    print(paste0("error reading gene set ", gs_name))
  }
}

```

# GSVA (ssgsea)

```{r gsva analysis}

y <- log2_norm_cpm
s <- fds$samples
gs <- c(gs_cptac, gs_hpa_panc, gs_pdacr, gs_hallmark, gs_3ca) 

# run gsva
x <- gsva(GSVA::gsvaParam(y, gs, tau=gsva_tau))
rownames(x) <- gsub("HALLMARK_", "H_", rownames(x))
rownames(x) <- gsub("GAVISH_", "", rownames(x))
rownames(x) <- gsub("METAPROGRAM_", "", rownames(x))
gsva_res <- x

```


# Dimension reduction analysis

## Highly variable genes

```{r highly variable genes, include=FALSE}

most_var_genes <- function(x, ntop=500, span=0.50) {
  tbl <- bind_cols(
    u = rowMeans(x),
    v = apply(x, 1, function(x) var(x))
  )
  lovar <- loess(v ~ u, data=tbl, span=span)
  tbl$lovar <- predict(lovar)
  tbl$vadj <- tbl$v - tbl$lovar
  keep <- order(tbl$vadj, decreasing=TRUE)[seq_len(min(ntop, nrow(tbl)))]
  return(x[keep,])
}

maplot <- function(y, ntop=1000, nlabel=100, myspan=0.5) {
  tbl <- bind_cols(
    g = rownames(y),
    u = rowMeans(y),
    v = apply(y, 1, function(x) var(x))
  )
  lovar <- loess(v ~ u, data=tbl, span=myspan)
  tbl$lovar <- predict(lovar)
  tbl$vadj <- tbl$v - tbl$lovar
  tbl <- tbl %>% 
    mutate(rank = min_rank(-vadj)) %>%
    arrange(rank)
  label_data <- filter(tbl, rank <= nlabel)
  top_data <- filter(tbl, rank < ntop)
  bot_data <- filter(tbl, rank >= ntop)
  p <- ggplot(tbl) +
    geom_point(data=bot_data, aes(u, v), alpha=0.5, size=1, color="#aaaaaa") +
    geom_point(data=top_data, aes(u, v), alpha=0.8, size=2, color="#ffaaaa") +
    geom_line(aes(u, lovar), color="blue") +
    geom_text_repel(data=label_data, aes(u, v, label=g), color="black", size=3, max.overlaps=Inf) +
    theme_minimal()
  return(p)
}

heatmap_genes <- function(s, x, filename="heatmap.pdf", width=10, height=10) {
  s <- s %>% mutate(
    noise_quantile = ntile(desc(bg_auc), 8)
  )
  annot_col <- column_to_rownames(s, "aoi") %>%
    select(patient, path_specimen, path_aoi, noise_quantile) %>%
    as.data.frame()
  p <- pheatmap(x,
                scale="row",
                show_rownames=TRUE,
                show_colnames=FALSE,
                border_color=NA,
                clustering_distance_rows = "euclidean",
                clustering_distance_cols = "euclidean",
                clustering_method = "ward.D2",
                # clustering_method = "average",
                # clustering_distance_rows = "correlation",
                # clustering_distance_cols = "correlation",
                breaks = seq(-3, 3, 0.05),
                color = colorRampPalette(c("purple3", "black", "yellow2"))(120),
                annotation_col = annot_col,
                annotation_colors = color_scales,
                fontsize=4,
                filename=filename,
                width=width,
                height=height)
  return(p)
}


y <- log2_norm_cpm
ntop <- 1000
nlabel <- 100
myspan <- 0.5

# maplot
p <- maplot(log2_norm_cpm, ntop=ntop, nlabel=nlabel, myspan=myspan)
f <- file.path(plot_dir, paste0("maplot"))
ggsave(paste0(f, ".pdf"), plot=p, width=10, height=10)
ggsave(paste0(f, ".png"), plot=p, width=10, height=10)

# select most variable genes
ntop <- 500
myspan <- 0.5
s <- fds$samples
y <- most_var_genes(log2_norm_cpm, ntop=ntop, span=myspan)

# heatmap
f <- file.path(plot_dir, "heatmap_highvar_genes.pdf")
p <- heatmap_genes(s, y, f)


```

## PCA

```{r pca}

# subset most variable genes
ntop <- 250
myspan <- 0.5
s <- fds$samples
y <- most_var_genes(log2_norm_cpm, ntop=ntop, span=myspan)

# plot dimensions
width <- 4
height <- 3

# PCA
npcs <- 50
pca_res <- prcomp(t(y), center=TRUE, scale.=FALSE, rank.=npcs)
#pca_res <- prcomp(t(y), center=TRUE, scale.=TRUE)
pca_res.var <- round(100 * pca_res$sdev^2 / sum( pca_res$sdev^2 ), 2)
pca_res.var_txt <- paste(colnames(pca_res$x), " (", paste(as.character(pca_res.var), "%", ")", sep=""), sep="")

# fviz plots
fviz_ntop <- 50

# percent variance of each PC
p <- fviz_eig(pca_res, col.var="blue", addlabels=TRUE)
f <- file.path(plot_dir, paste0("pca_fviz_eig.pdf"))
ggsave(f, p)

# contributions of variables to each PC
p <- fviz_contrib(pca_res, choice="var", axes=1, top=fviz_ntop) +
  theme(axis.text.x = element_text(angle=90))
f <- file.path(plot_dir, paste0("pca_fviz_contrib1.pdf"))
ggsave(f, p)

p <- fviz_contrib(pca_res, choice="var", axes=2, top=fviz_ntop) +
  theme(axis.text.x = element_text(angle=90))
f <- file.path(plot_dir, paste0("pca_fviz_contrib2.pdf"))
ggsave(f, p)

p <- fviz_contrib(pca_res, choice="var", axes=3, top=fviz_ntop) +
  theme(axis.text.x = element_text(angle=90))
f <- file.path(plot_dir, paste0("pca_fviz_contrib3.pdf"))
ggsave(f, p)

# contribution to first 3 PCs
p <- fviz_contrib(pca_res, choice="var", axes = 1:3, top=fviz_ntop) +
  theme(axis.text.x = element_text(angle=90))
f <- file.path(plot_dir, paste0("pca_fviz_contrib123.pdf"))
ggsave(f, p)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  pca_res,
  axes = c(1,2),
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("pca_fviz_pca_var_pc12.pdf"))
ggsave(f, p, width=7, height=7)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  pca_res,
  axes = c(1,3),
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("pca_fviz_pca_var_pc13.pdf"))
ggsave(f, p, width=7, height=7)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  pca_res,
  axes = c(2,3),
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("pca_fviz_pca_var_pc23.pdf"))
ggsave(f, p, width=7, height=7)



#
# PCA plots
#
x <- bind_cols(s, as_tibble(pca_res$x))

# bg_ks_dist
p <- ggplot(x, aes(x=PC1, y=PC2, color=bg_auc)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  theme_bw() +
  theme(aspect.ratio=1) +
  theme(legend.text = element_text(size=7)) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_bg_auc.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# patient
p <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$patient) +
  theme_bw() +
  theme(aspect.ratio=1) +
  theme(legend.text = element_text(size=7)) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_patient.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path specimen
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  theme(legend.text = element_text(size=7)) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_path_specimen.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path aoi
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  theme_bw() +
  theme(aspect.ratio=1) +
  theme(legend.text = element_text(size=7)) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_path_aoi.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path
p <- ggplot(x, aes(x=PC1, y=PC2, color=path, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path) +
  theme_bw() +
  theme(aspect.ratio=1) +
  theme(legend.text = element_text(size=7)) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_path.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)



#
# Plots of PC1, PC2, and PC3
#
pca_plot <- function(x, xvar, yvar, color_by, shape_by) {
  color_by_string <- rlang::as_string(rlang::ensym(color_by))
  shape_by_string <- rlang::as_string(rlang::ensym(shape_by))
  
  if (is.character(pull(x, {{color_by}}))) {
    scale_color_func <- scale_color_manual(values = color_scales[[color_by_string]])
  } else {
    scale_color_func <- scale_color_viridis_c()
  }

  p <- ggplot(x, aes(x={{xvar}}, y={{yvar}}, color={{color_by}}, shape={{shape_by}})) +
    geom_point(alpha = 0.8, size=3) +
    scale_color_func +
    #scale_color_manual(values = color_scales[[color_by_string]]) +
    scale_shape_manual(values = shape_scales[[shape_by_string]]) +
    theme_bw() +
    theme(aspect.ratio=1,
          legend.text = element_text(size=7))
    # coord_fixed() +
  return(p)
}


# bg_auc
p12 <- pca_plot(x, PC1, PC2, bg_auc, path_specimen)
p13 <- pca_plot(x, PC1, PC3, bg_auc, path_specimen)
p23 <- pca_plot(x, PC2, PC3, bg_auc, path_specimen)
p <- p12 + p13 + p23 + 
  plot_layout(guides = "collect")
f <- file.path(plot_dir, paste0("pca_bg_auc_pc123.pdf"))
ggsave(f, plot=p, device="pdf", width=12, height=6)

# path aoi
p12 <- pca_plot(x, PC1, PC2, path_aoi, path_specimen)
p13 <- pca_plot(x, PC1, PC3, path_aoi, path_specimen)
p23 <- pca_plot(x, PC2, PC3, path_aoi, path_specimen)
p <- p12 + p13 + p23 + 
  plot_layout(guides = "collect")
f <- file.path(plot_dir, paste0("pca_path_aoi_pc123.pdf"))
ggsave(f, plot=p, device="pdf", width=12, height=6)

# patient
p12 <- pca_plot(x, PC1, PC2, patient, path_specimen)
p13 <- pca_plot(x, PC1, PC3, patient, path_specimen)
p23 <- pca_plot(x, PC2, PC3, patient, path_specimen)
p <- p12 + p13 + p23 + 
  plot_layout(guides = "collect")
f <- file.path(plot_dir, paste0("pca_patient_pc123.pdf"))
ggsave(f, plot=p, device="pdf", width=12, height=6)


```


## PCA clustering

```{r pca clustering}
# the following guides were used in this section:
# guide: https://rpubs.com/KarolinaSzczesna/862710
# guide: https://rpubs.com/Bury/ClusteringOnPcaResults

# npcs
npcs <- 3

# choose number of clusters
# x <- pca_res$x[,1:npcs]
x <- pca_res$x

p <- fviz_nbclust(x, FUNcluster=cluster::pam, method="silhouette")
f <- file.path(plot_dir, paste0("pca_fviz_nbclust_silhouette.pdf"))
ggsave(f, plot=p, device="pdf", width=4, height=4)

# p <- fviz_nbclust(x, FUNcluster=cluster::pam, method="wss")
# f <- file.path(plot_dir, paste0("pca_fviz_nbclust_wss.pdf"))
# ggsave(f, plot=p, device="pdf", width=4, height=4)
# 
# p <- fviz_nbclust(x, FUNcluster=cluster::pam, method="gap_stat", nboot=100)
# f <- file.path(plot_dir, paste0("pca_fviz_nbclust_gapstat.pdf"))
# ggsave(f, plot=p, device="pdf", width=4, height=4)

# set number of clusters
nclusters <- 3

# commented out: hierarchical clustering based on first two PCs
# x <- pca_res$x[, npcs]
# d <- dist(x)
# #hc <- hclust(d, method="complete")
# hc <- hclust(d, method="ward.D2")
# memb <- cutree(hc, k=nclusters)
# memb <- paste0("c", memb)

# commented out: decided to use hierarchical clustering
# kmeans clustering based on PCs
#km <- kmeans(x, centers=nclusters)
#memb <- paste0("c", km$cluster)

# pam
x <- pca_res$x[,1:npcs]
d <- dist(x)
pamx <- cluster::pam(d, k=3)
memb <- paste0("c", pamx$clustering)

# add cluster membership to sample properties
fds$samples$cluster <- memb

# plot clustering
x <- bind_cols(fds$samples, as_tibble(pca_res$x))

# cluster
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$path_aoi) +
  theme_bw() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster) +
  theme_bw() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
p1 + p2





# set number of clusters
nclusters <- 5

# pam
x <- pca_res$x[,1:npcs]
pamx <- cluster::pam(dist(x), k=nclusters)
memb <- paste0("c", pamx$clustering)

# add cluster membership to sample properties
fds$samples$cluster <- memb

# plot clustering
x <- bind_cols(fds$samples, as_tibble(pca_res$x))

# pca plot clustering
p12 <- pca_plot(x, PC1, PC2, cluster, path_specimen)
p13 <- pca_plot(x, PC1, PC3, cluster, path_specimen)
p23 <- pca_plot(x, PC2, PC3, cluster, path_specimen)
p <- p12 + p13 + p23 +
  plot_layout(guides = "collect")
f <- file.path(plot_dir, paste0("pca_cluster_pc123.pdf"))
ggsave(f, plot=p, device="pdf", width=12, height=6)









#
# identify clusters based on pca relationships
#
fds$samples <- fds$samples %>% mutate(
  cluster_name = case_when(cluster == "c1" ~ "cLR",
                           cluster == "c2" ~ "cHR",
                           cluster == "c3" ~ "cNL")
)
x <- bind_cols(fds$samples, as_tibble(pca_res$x))

#
# single multipanel plot for figure
# 
# path specimen
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
#  coord_fixed() +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

# path aoi
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
#  coord_fixed() +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

# cluster
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster_name) +
#  coord_fixed() +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

# patient
p4 <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = pals::cols25()) +
#  coord_fixed() +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

p <- p1 + p2 + p3 + p4 + plot_layout(nrow=1)
f <- file.path(plot_dir, paste0("pca_panel_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=16, height=6)


# f <- file.path(plot_dir, paste0("pca_cluster.pdf"))
# ggsave(f, plot=p, device="pdf", width=width, height=height)
# p <- p1 + theme(legend.position="none") + p2 + theme(aspect.ratio=1)
# f <- file.path(plot_dir, paste0("pca_cluster_12_23.pdf"))
# ggsave(f, plot=p, device="pdf", width=7, height=3)


# pairs plot of PCs

p1 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$path_aoi) +
  theme_bw() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
p2 <- ggplot(x, aes(x=PC3, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$path_aoi) +
  theme_bw() +
  labs(x=pca_res.var_txt[3], y=pca_res.var_txt[2])
p3 <- ggplot(x, aes(x=PC3, y=PC4, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$path_aoi) +
  theme_bw() +
  labs(x=pca_res.var_txt[3], y=pca_res.var_txt[4])
p1 + p2 + p3



```


# Differential expression analysis

## DE analysis

```{r}

run_limma_trend <- function(y, design, contrasts, padj_cutoff, log2fc_cutoff) {
  fit <- lmFit(y, design)
  fit <- contrasts.fit(fit, contrasts)
  fit <- eBayes(fit, trend=TRUE)

  x <- NULL
  for (coef in colnames(contrasts)) {
    res <- topTable(fit, coef=coef, number=Inf, sort.by="none")
    res <- as_tibble(res, rownames="gene")
    res <- res %>%
      select(gene,
             log2fc=logFC,
             avgexpr=AveExpr,
             pval=P.Value,
             padj=adj.P.Val) %>%
      mutate(de = case_when(padj > padj_cutoff ~ "no",
                            log2fc < -log2fc_cutoff ~ "dn",
                            log2fc > log2fc_cutoff ~ "up",
                            TRUE ~ "no"),
             contrast = coef)
    x <- bind_rows(x, res)
  }
  return(x)
}

# de results
de <- NULL

#
# DE analysis of INV lesions
#
s <- filter(fds$samples, path_specimen == "ipmn_inv")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
y <- log2_norm_cpm[, s$aoi]

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(invinv_vs_invnl = path_aoiinv - path_aoinl,
                           invhgd_vs_invnl = path_aoihgd - path_aoinl,
                           invlgd_vs_invnl = path_aoilgd - path_aoinl,
                           invinv_vs_invlgd = path_aoiinv - path_aoilgd,
                           invhgd_vs_invlgd = path_aoihgd - path_aoilgd,
                           invinv_vs_invhgd = path_aoiinv - path_aoihgd,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de <- bind_rows(de, res)


#
# DE analysis of LGD lesions
#
s <- filter(fds$samples, path_specimen == "ipmn_lgd")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd"))
y <- log2_norm_cpm[, s$aoi]

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(lgdlgd_vs_lgdnl = path_aoilgd - path_aoinl,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de <- bind_rows(de, res)


#
# setup cohort of all samples 
#
s <- fds$samples
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
s$path_specimen <- factor(s$path_specimen, levels=c("ipmn_lgd", "ipmn_inv"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, 
                         levels=c("ipmn_lgd_nl", "ipmn_lgd_lgd", 
                                  "ipmn_inv_nl", "ipmn_inv_lgd", 
                                  "ipmn_inv_hgd", "ipmn_inv_inv"))
s$cluster <- factor(s$cluster_name, levels=c("cNL", "cLR", "cHR"))
y <- log2_norm_cpm[, s$aoi]


#
# DE analysis INV vs LGD specimens
#
design <- model.matrix(~0 + path_patient, s)
colnames(design) <- levels(s$path_patient)
colnames(design)
contrasts <- makeContrasts(
  invinv_vs_lgdnl = ipmn_inv_inv - ipmn_lgd_nl,
  invhgd_vs_lgdnl = ipmn_inv_hgd - ipmn_lgd_nl,
  invlgd_vs_lgdnl = ipmn_inv_lgd - ipmn_lgd_nl,
  invnl_vs_lgdnl = ipmn_inv_nl - ipmn_lgd_nl,
  invinv_vs_lgdlgd = ipmn_inv_inv - ipmn_lgd_lgd,
  invhgd_vs_lgdlgd = ipmn_inv_hgd - ipmn_lgd_lgd,
  invlgd_vs_lgdlgd = ipmn_inv_lgd - ipmn_lgd_lgd,
  levels=design
)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de <- bind_rows(de, res)


#
# DE analysis of clusters
#
design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(clusterHRvLR = clustercHR - clustercLR,
                           clusterHRvNL = clustercHR - clustercNL,
                           clusterLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de <- bind_rows(de, res)


#
# DE analysis of low-grade / normal regions within clusters
#
s <- fds$samples %>% filter(path_aoi %in% c("nl","lgd"))
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
s$path_specimen <- factor(s$path_specimen, levels=c("ipmn_lgd", "ipmn_inv"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, 
                         levels=c("ipmn_lgd_nl", "ipmn_lgd_lgd", 
                                  "ipmn_inv_nl", "ipmn_inv_lgd", 
                                  "ipmn_inv_hgd", "ipmn_inv_inv"))
s$cluster <- factor(s$cluster_name, levels=c("cNL", "cLR", "cHR"))
y <- log2_norm_cpm[, s$aoi]

design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(nllgd_cHRvLR = clustercHR - clustercLR,
                           nllgd_cHRvNL = clustercHR - clustercNL,
                           nllgd_cLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de <- bind_rows(de, res)

# write all results
table(de$contrast)
write_xlsx(de, file.path(results_dir, "de_path.xlsx"))

# write cluster results
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")
x <- de %>% filter(contrast %in% contrasts)
write_xlsx(x, file.path(results_dir, "de_clusters.xlsx"))


```

## DE Volcano plots

```{r de volcano plots}

plot_de_volcano <- function(res, nlabel=50) {
  # label the top n up/down genes
  label_data <- bind_rows(filter(res, de == "up") %>% slice_max(log2fc, n=nlabel),
                          filter(res, de == "dn") %>% slice_min(log2fc, n=nlabel))
  label_data <- distinct(label_data)

  p <- ggplot(res, aes(x=log2fc, y=-log10(padj), color=de, size=de)) +
    geom_point(alpha=0.7) +
    geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +
    scale_color_manual(values=c("no"="grey", "dn"="blue", "up"="red")) +
    scale_size_manual(values=c("no"=1, "dn"=2, "up"=2)) +
    theme_minimal() +
    theme(legend.position="bottom") +
    theme(axis.line = element_line(color = "black")) +
    labs(x="log2fc", y="-log10(padj)")
  return(p)
}

# volcano plots
for (mycontrast in unique(de$contrast)) {
  print(mycontrast)
  x <- filter(de, contrast == mycontrast)
  p <- plot_de_volcano(x) + 
    labs(title="Volcano Plot", subtitle=mycontrast)
  f <- paste0("volcano_de_", mycontrast)
  ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=8)
  ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=8)
}

```

## DE Heatmaps

```{r de heatmaps, include=FALSE}

# heatmap functions
plot_pheatmap <- function(m, 
                          annot_col, 
                          annot_row, 
                          annot_colors,
                          filename, width, height,
                          xmin=-6, xmax=6) {
  pheatmap(m,
           scale = "none",
           color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
           breaks = seq(xmin, xmax, length.out=51),
           annotation_col = annot_col,
           annotation_row = annot_row,
           annotation_colors = annot_colors,
           border_color = NA,
           cluster_rows = TRUE,
           cluster_cols = TRUE,
           clustering_distance_rows = "euclidean",
           clustering_distance_cols = "euclidean",
           clustering_method = "ward.D2",
#           cellwidth = 6,
#           cellheight = 6,
           fontsize = 6,
           fontsize_row = 6,
           fontsize_col = 6,
           filename = filename, 
           width = width,
           height = height)
}

# setup heatmap annotations
gene_annot <- fds$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot) %>%
  column_to_rownames("gene") %>%
  as.data.frame()
annot_col <- column_to_rownames(fds$samples, "aoi") %>% 
  select(patient, path_aoi, path_specimen, cluster_name) %>%
  as.data.frame()

#
# IPMN clusters
#
# de gene params
padj_cutoff <- 0.01
log2fc_cutoff <- 3
ntop <- 10
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

g <- filter(de, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)
cluster_marker_genes <- g

# centered gene expression (subtract median)
s <- fds$samples
x <- log2_norm_cpm[g, s$aoi]
u <- apply(x, 1, mean)
x <- sweep(x, 1, u, FUN="-")
xmin <- -6
xmax <- 6

# heatmap
f <- file.path(plot_dir,  paste0("heatmap_de_cluster_markers_", ngenes, "_genes"))
h <- 6
w <- 6
p <- plot_pheatmap(x, 
                   annot_col[s$aoi,], 
                   annot_row[g,], 
                   annot_colors=color_scales,
                   filename=paste0(f, ".png"), 
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


#
# LGD specimens
#
# de gene params
padj_cutoff <- 1e-5
log2fc_cutoff <- 3
ntop <- 50
contrasts <- c("lgdlgd_vs_lgdnl")
path_types <- c("ipmn_lgd_nl_epithelial", "ipmn_lgd_lgd_epithelial")

g <- filter(de, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)

# standardize gene expression
s <- fds$samples %>% filter(path %in% path_types)
snl <- fds$samples %>% filter(path %in% "ipmn_lgd_nl_epithelial")

x <- log2_norm_cpm[g, s$aoi]
u <- apply(x[, snl$aoi], 1, median)
x <- sweep(x, 1, u, FUN="-")
v <- pmax(apply(x, 1, sd), 1)
x <- sweep(x, 1, v, FUN="/")
xmin <- -3
xmax <- 3

# x <- log2_norm_cpm[g, s$aoi]
# u <- apply(x[, snl$aoi], 1, median)
# x <- sweep(x, 1, u, FUN="-")
# v <- pmax(apply(x[, snl$aoi], 1, sd), 1)
# x <- sweep(x, 1, v, FUN="/")


# heatmap
f <- file.path(plot_dir,  paste0("heatmap_de_ipmn_lgd_", ngenes, "_genes"))
h <- round(3 + (length(g) / 10))
w <- 10
p <- plot_pheatmap(x, 
                   annot_col[s$aoi,], 
                   annot_row[g,], 
                   annot_colors=color_scales,
                   filename=paste0(f, ".png"), 
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)



#
# HGD/INV specimens
#
# de gene params
padj_cutoff <- 1e-5
log2fc_cutoff <- 3
ntop <- 50
contrasts <- c("invhgd_vs_invlgd",
               "invhgd_vs_invnl", 
               "invlgd_vs_invnl", 
               "invinv_vs_invhgd",
               "invinv_vs_invlgd",
               "invinv_vs_invnl")
path_types <- c("ipmn_inv_hgd_epithelial", 
                "ipmn_inv_inv_epithelial",
                "ipmn_inv_lgd_epithelial", 
                "ipmn_lgd_lgd_epithelial",
                "ipmn_inv_nl_epithelial")
nl_types <- c("ipmn_inv_nl_epithelial")

g <- filter(de, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)

# standardize gene expression
s <- fds$samples %>% filter(path %in% path_types)
snl <- fds$samples %>% filter(path %in% nl_types)
x <- log2_norm_cpm[g, s$aoi]
u <- apply(x[, snl$aoi], 1, median)
x <- sweep(x, 1, u, FUN="-")
v <- pmax(apply(x, 1, sd), 1)
x <- sweep(x, 1, v, FUN="/")
xmin <- -3
xmax <- 3

# x <- log2_norm_cpm[g, s$aoi]
# u <- apply(x[, snl$aoi], 1, median)
# x <- sweep(x, 1, u, FUN="-")
# v <- pmax(apply(x[, snl$aoi], 1, sd), 1)
# x <- sweep(x, 1, v, FUN="/")
# xmin <- -4
# xmax <- 4

# heatmap
f <- file.path(plot_dir,  paste0("heatmap_de_inv_ipmn_", ngenes, "_genes"))
h <- round(3 + (length(g) / 10))
w <- 10
p <- plot_pheatmap(x, 
                   annot_col[s$aoi,], 
                   annot_row[g,], 
                   annot_colors=color_scales,
                   filename=paste0(f, ".png"), 
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


#
# Full IPMN Dataset
#
padj_cutoff <- 1e-5
log2fc_cutoff <- 5
ntop <- 50

nl_types <- c("ipmn_lgd_nl_epithelial",
              "ipmn_inv_nl_epithelial")

g <- filter(de, abs(log2fc) > log2fc_cutoff, padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)

# standardize gene expression
s <- fds$samples
x <- log2_norm_cpm[g, s$aoi]
x <- t(scale(t(x)))
xmin <- -2
xmax <- 2

# heatmap
f <- file.path(plot_dir,  paste0("heatmap_de_ipmn_", ngenes, "_genes"))
h <- round(3 + (length(g) / 10))
w <- 10

p <- plot_pheatmap(x,
                   annot_col[s$aoi,],
                   annot_row[g,],
                   annot_colors=color_scales,
                   filename=paste0(f, ".png"),
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)

```


## PCA feature plots for marker genes

```{r pca feature plot for each marker gene, include=FALSE}

s <- fds$samples
y <- log2_norm_cpm

for (gene in cluster_marker_genes) {
  print(gene)
  x <- bind_cols(s, as_tibble(pca_res$x), gene=unlist(y[gene,]))
  p <- ggplot(x, aes(x=PC1, y=PC2, color=gene, shape=path_specimen)) +
    geom_point(alpha = 0.8, size=3) +
    scale_color_viridis_c() +
    theme_minimal() +
    theme(aspect.ratio=1) +
    labs(x="PC1", y="PC2", title=gene)
  f <- file.path(plot_dir, paste0("pca_featureplot_", gene))
  ggsave(paste0(f, ".pdf"), plot=p, width=4, height=3)
  ggsave(paste0(f, ".png"), plot=p, width=4, height=3)
}

```


## DE 2D scatter plots

```{r ipmn de 2d scatter plots}

plot_de_scatter2d <- function(tbl, xcontrast, ycontrast, nlabel=10, 
                              highlight_genes=NULL) {
  if (is.null(highlight_genes)) {
    highlight_genes <- c()
  }
  xde <- sym(paste0("de_", xcontrast))
  yde <- sym(paste0("de_", ycontrast))
  xlog2fc <- sym(paste0("log2fc_", xcontrast))
  ylog2fc <- sym(paste0("log2fc_", ycontrast))

  x <- tbl %>% 
    mutate(subtype = paste0("x", {{xde}}, "_", "y", {{yde}}),
           log2fc_abs_x = case_when({{xde}} == "no" ~ 0,
                                    TRUE ~ abs({{xlog2fc}})),
           log2fc_abs_y = case_when({{yde}} == "no" ~ 0,
                                    TRUE ~ abs({{ylog2fc}})),
           log2fc_abs_max = pmax(log2fc_abs_x, log2fc_abs_y))
  
  color_scale = c(xno_yno = "#888888",
                  xup_yup = "#FF0000",
                  xdn_ydn = "#0000aa",
                  xup_ydn = "#eeee00",
                  xdn_yup = "#ff00ff",
                  xup_yno = "#32cd32",
                  xdn_yno = "#670092",
                  xno_yup = "#ff6700",
                  xno_ydn = "#16ccc4")
  size_scale = c(xno_yno = 1, 
                 xup_yup = 2,
                 xdn_ydn = 2,
                 xup_ydn = 2,
                 xdn_yup = 2,
                 xup_yno = 2,
                 xdn_yno = 2,
                 xno_yup = 2,
                 xno_ydn = 2)
  
  annot_data <- filter(x, subtype != "xno_yno")
  none_data <- filter(x, subtype == "xno_yno")
  label_data <- annot_data %>% 
    group_by(subtype) %>%
    slice_max(log2fc_abs_max, n=nlabel) %>%
    ungroup()
  highlight_data <- filter(x, gene %in% highlight_genes)

  p <- ggplot(x, aes(x={{xlog2fc}}, 
                     y={{ylog2fc}},
                     color=subtype, 
                     size=subtype)) +
    geom_point(data=none_data, alpha=0.4) + 
    geom_point(data=annot_data, alpha=0.8) +
    geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) + 
    geom_point(data=highlight_data, pch=21, fill=NA, size=5, colour="#ffff00", stroke=1) + 
    geom_point(data=highlight_data, pch=21, fill=NA, size=4, colour="#222222", stroke=1, alpha=0.5) +
    geom_text_repel(data=highlight_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +     
    scale_color_manual(values = color_scale) +
    scale_size_manual(values = size_scale, guide = "none") + 
    theme_minimal() +
    xlab(as_string(xlog2fc)) +
    ylab(as_string(ylog2fc)) +
    labs(color = "DE Genes")
  return(p)
}


# merge de analysis by gene
de_merged <- de %>% select(gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))
nlabel <- 10
width <- 8
height <- 8


#
# IPMN Clusters
#
p <- plot_de_scatter2d(de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterHRvLR", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_clusterHRvNL_y_clusterHRvLR")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterLRvNL", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterLRvNL (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_clusterHRvNL_y_clusterLRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


#
# IPMN Clusters subset with NL and LGD
#
p <- plot_de_scatter2d(de_merged, 
                       xcontrast="nllgd_cHRvLR", 
                       ycontrast="nllgd_cHRvNL", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_nllgd_cHRvLRL_y_nllgd_cHRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


#
# IPMN INV Specimens
#
p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invinv_vs_invnl", 
                       ycontrast="invinv_vs_invlgd", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="INV vs NL (x-axis) INV vs LGD (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invinv_vs_invnl_y_invinv_vs_invlgd")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invinv_vs_invnl", 
                       ycontrast="invhgd_vs_invnl", 
                       nlabel=nlabel) +
  labs(title="IPMN INV Specimens", 
       subtitle="INV-INV vs INV-NL (x-axis) INV-HGD vs INV-NL (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invinv_vs_invnl_y_invhgd_vs_invnl")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invinv_vs_invnl", 
                       ycontrast="invinv_vs_invhgd", 
                       nlabel=nlabel) +
  labs(title="IPMN INV Specimens", 
       subtitle="INV-INV vs INV-NL (x-axis) INV-INV vs INV-HGD (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invinv_vs_invnl_y_invinv_vs_invhgd")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invhgd_vs_invnl", 
                       ycontrast="invlgd_vs_invnl", 
                       nlabel=nlabel) +
  labs(title="IPMN INV Specimens", 
       subtitle="INV-HGD vs INV-NL (x-axis) INV-LGD vs INV-NL (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invhgd_vs_invnl_y_invlgd_vs_invnl")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)



#
# IPMN: LGD vs NL in INV and LGD specimens
#
p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invlgd_vs_invnl", 
                       ycontrast="lgdlgd_vs_lgdnl", 
                       nlabel=nlabel) +
  labs(title="IPMN: INV vs LGD Specimens", 
       subtitle="INV-LGD vs INV-NL (x-axis) LGD-LGD vs LGD-NL (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invlgd_vs_invnl_y_lgdlgd_vs_lgdnl")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invlgd_vs_lgdnl", 
                       ycontrast="invlgd_vs_lgdlgd", 
                       nlabel=nlabel) +
  labs(title="IPMN INV vs LGD specimens ", 
       subtitle="INV-LGD vs LGD-NL (x-axis) INV-LGD vs LGD-LGD (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invlgd_vs_lgdnl_y_invlgd_vs_lgdlgd")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_merged, 
                       xcontrast="invlgd_vs_invnl", 
                       ycontrast="invlgd_vs_lgdlgd", 
                       nlabel=nlabel) +
  labs(title="IPMN INV vs LGD specimens ", 
       subtitle="INV-LGD vs INV-NL (x-axis) INV-LGD vs LGD-LGD (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_invlgd_vs_invnl_y_invlgd_vs_lgdlgd")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


```


# GSEA

## GSEA analysis

```{r gsea, include=FALSE}

get_de_ranks <- function(de, a) {
  ranks <- de %>% 
    dplyr::filter(contrast == a) %>%
    select(gene, log2fc, padj) %>%
    mutate(rank = log2fc)
  ranks = sort(setNames(ranks$rank, ranks$gene), decreasing = TRUE)
  return(ranks)
}

plot_gsea_enrichment <- function(x, ranks, a, gs) {
  txt_stat <- paste0("ES=", round(x$ES,2), " NES=", round(x$NES, 2), " padj=", format(x$padj, scientific=TRUE, digits=3))
  txt_title <- paste0("ranks: ", a, " gs: ", x$pathway)
  p <- plotEnrichment(gs[[x$pathway]], ranks) +
    annotate(geom="text", x=150, y=0.1, label=txt_stat, hjust=0) +
    labs(title = txt_title, 
         xlab = "Rank",
         ylab = "Enrichment Score") +
    theme(plot.title = element_text(size=8))
  return(p)
}

run_batch_fgsea <- function(my_analyses, my_de, my_gs, my_prefix, 
                            my_plot_dir, 
                            my_padj_cutoff=0.01,
                            do_plots=TRUE) {
  gsea <- NULL
  for (a in my_analyses) {
    ranks <- get_de_ranks(my_de, a)
    #res <- fgseaSimple(pathways=my_gs, stats=ranks, minSize=10, nperm=10000)
    res <- fgsea(pathways=my_gs, stats=ranks, minSize=10, eps=0, nPermSimple=10000)
    res <- mutate(res, analysis = a)
    gsea <- bind_rows(gsea, res)
    print(a)
    if (do_plots) {
      for (i in 1:nrow(res)) {
        x <- res[i,]
        if (is.na(x$padj) | (x$padj >= my_padj_cutoff)) { next }
        print(x$pathway)
        p <- plot_gsea_enrichment(x, ranks, a, my_gs)
        f <- file.path(my_plot_dir, paste0(my_prefix, "_", a, "_gs_", x$pathway, ".pdf"))
        ggsave(f, plot=p, device="pdf", width=5, height=3)
      }
    }
  }
  return(gsea)
}


#
# Setup analysis
#
table(de$contrast)
analyses <- unique(de$contrast)


#
# PDAC gene sets
#
# gs <- c(gs_cptac, gs_hpa_panc, gs_pdacr)
# prefix <- "gsea_pdac"
# padj_cutoff <- 0.01
# gsea <- run_batch_fgsea(analyses, de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
# gsea_pdac <- as_tibble(gsea)
# 
# # write gsea results
# f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
# data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
# f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
# write_xlsx(gsea, f)

#
# CPTAC GSEA
#
prefix <- "gsea_cptac"
padj_cutoff <- 0.01
gs <- gs_cptac
gsea <- run_batch_fgsea(analyses, de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
gsea_cptac <- as_tibble(gsea)
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)


#
# HPA (TCGA) GSEA
#
prefix <- "gsea_hpa_panc"
padj_cutoff <- 0.01
gs <- gs_hpa_panc
gsea <- run_batch_fgsea(analyses, de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
gsea_hpa_panc <- as_tibble(gsea)
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# PDAC subtypes GSEA
#
prefix <- "gsea_pdacr"
padj_cutoff <- 0.01
gs <- gs_pdacr
gsea <- run_batch_fgsea(analyses, de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
gsea_pdacr <- as_tibble(gsea)
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# Hallmark gene set analysis
#
prefix <- "gsea_hallmark"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, de, gs_hallmark, prefix, gsea_plot_dir, 
                        padj_cutoff, do_plots=FALSE)
gsea_hallmark <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_name) %>% distinct(), 
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# 3CA gene set analysis
#
prefix <- "gsea_3ca"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, de, gs_3ca, prefix, gsea_plot_dir, 
                        padj_cutoff, do_plots=FALSE)
gsea_3ca <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_name) %>% distinct(), 
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)


#
# Gene Ontology analysis
#
prefix <- "gsea_gobp"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, de, gs_gobp, prefix, gsea_plot_dir, 
                        padj_cutoff, do_plots=FALSE)
gsea_gobp <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_name) %>% distinct(),
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)


```

## GSEA plots clusters CPTAC and HPA

```{r gsea plots clusters}

# CPTAC and HPA
pathways <- c("CPTAC_PROT_ALL_DN", "CPTAC_PROT_ALL_UP",
              "CPTAC_RNA_ALL_DN", "CPTAC_RNA_ALL_UP",
              "HPA_PANCREATIC_CANCER_FAV_PROGNOSIS",
              "HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS")
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")
padj_cutoff <- 0.01

gsea <- bind_rows(gsea_cptac, gsea_hpa_panc)
gsea <- gsea %>%
  filter(pathway %in% pathways, analysis %in% contrasts)
gsea <- mutate(gsea, neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))
gsea$analysis <- factor(gsea$analysis, levels=contrasts)

p <- ggplot(gsea, aes(x=analysis, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_barplot_clusters_pdac")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=4)


#
# just NL and LGD disease
#
contrasts <- c("nllgd_cHRvLR", "nllgd_cHRvNL", "nllgd_cLRvNL")
padj_cutoff <- 0.01

gsea <- bind_rows(gsea_cptac, gsea_hpa_panc)
gsea <- gsea %>%
  filter(pathway %in% pathways, analysis %in% contrasts)
gsea <- mutate(gsea, neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))

p <- ggplot(gsea, aes(x=reorder(analysis, NES), y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_barplot_nllgd_clusters_pdac")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=4)


```


## GSEA plots clusters Hallmark

```{r gsea clusters hallmark}

# Hallmark
padj_cutoff <- 0.01

# all regions
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

gsea <- gsea_hallmark %>%
  filter(analysis %in% contrasts)
gsea <- mutate(gsea, pathway = gsub("HALLMARK_", "", pathway))
gsea <- gsea %>% mutate(
  neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  sig = padj < padj_cutoff)

# only keep significant results (at least one analysis)
gsea <- gsea %>% group_by(pathway) %>%
  filter(any(padj < padj_cutoff)) %>%
  ungroup()

# arrange by cluster 1v2 for readability
gsea$pathway <- factor(gsea$pathway, levels=gsea %>% filter(analysis == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway))

p <- ggplot(gsea, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  xlim(-2.5,2.5) +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=7)) +
#  theme(axis.text.x = element_text(color="black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ analysis, nrow=1) + 
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")

f <- paste0("gsea_barplot_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=3)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=3)


# just NL and LGD regions

contrasts <- c("nllgd_cHRvLR", "nllgd_cHRvNL", "nllgd_cLRvNL")
gsea <- gsea_hallmark %>%
  filter(analysis %in% contrasts)
gsea <- mutate(gsea, pathway = gsub("HALLMARK_", "", pathway))
gsea <- gsea %>% mutate(
  neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  sig = padj < padj_cutoff)

# only keep significant results (at least one analysis)
gsea <- gsea %>% group_by(pathway) %>%
  filter(any(padj < padj_cutoff)) %>%
  ungroup()

# arrange by cluster 1v2 for readability
gsea$pathway <- factor(gsea$pathway, levels=gsea %>% filter(analysis == "nllgd_cHRvLR") %>% arrange(NES) %>% pull(pathway))

p <- ggplot(gsea, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  xlim(-2.5,2.5) +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=7)) +
#  theme(axis.text.x = element_text(color="black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ analysis, nrow=1) + 
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")

f <- paste0("gsea_barplot_nl_lgd_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=3)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=3)

```

## GSEA plots clusters 3CA

```{r gsea clusters 3CA}

contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")
padj_cutoff <- 0.01

gsea <- gsea_3ca %>%
  filter(analysis %in% contrasts)
gsea <- mutate(gsea, pathway = gsub("GAVISH_3CA_", "", pathway))
gsea <- mutate(gsea, pathway = gsub("METAPROGRAM_", "", pathway))
gsea <- gsea %>% mutate(
  neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  sig = padj < padj_cutoff)

# only keep significant results (at least one analysis)
gsea <- gsea %>% group_by(pathway) %>%
  filter(any(padj < padj_cutoff)) %>%
  ungroup()

# arrange by clusterHR for readability
gsea$pathway <- factor(gsea$pathway, levels=gsea %>% filter(analysis == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway))

p <- ggplot(gsea, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
#  xlim(-3,3) +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=7)) +
#  theme(axis.text.x = element_text(color="black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ analysis, nrow=1) + 
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")

f <- paste0("gsea_barplot_clusters_3ca")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=6)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=6)

```

## GSEA plots clusters pdacr

```{r}

padj_cutoff <- 0.01
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")


#
# Moffitt (main figure)
#
pathways <- c("Moffitt.F6_BasalLike.top250", 
              "Moffitt.F8_Classical.top250",
              "Moffitt.F3_Exocrine.top250")

gsea <- gsea_pdacr %>%
  filter(analysis %in% contrasts, pathway %in% pathways)
gsea <- mutate(gsea, neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))

# arrange for readability
gsea$pathway <- factor(gsea$pathway, levels=gsea %>% filter(analysis == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway))
gsea$analysis <- factor(gsea$analysis, levels=c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR"))

p <- ggplot(gsea, aes(x=analysis, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=4)



#
# Other pathways (supplemental figure)
#
moffitt_pathways <- c("Moffitt.F6_BasalLike.top250", 
                      "Moffitt.F8_Classical.top250",
                      "Moffitt.F3_Exocrine.top250")
collison_pathways <- c("Collisson.Classical", 
                       "Collisson.Exocrine",
                       "Collisson.QM")
icgc_pathways <- c("ICGC.ADEX.Down", 
                   "ICGC.ADEX.Up",
                   "ICGC.Immunogenic.Down",
                   "ICGC.Immunogenic.Up",
                   "ICGC.Progenitor.Down",
                   "ICGC.Progenitor.Up",
                   "ICGC.Squamous.Down",
                   "ICGC.Squamous.Up")
chan_seng_yue_pathways <- c("Chan_Seng_Yue.BasalA", 
                            "Chan_Seng_Yue.BasalB",
                            "Chan_Seng_Yue.ClassicalA",
                            "Chan_Seng_Yue.ClassicalB")
pathways <- c(moffitt_pathways, collison_pathways, icgc_pathways, chan_seng_yue_pathways)

gsea <- gsea_pdacr %>%
  filter(analysis %in% contrasts, pathway %in% pathways)
gsea <- mutate(gsea, neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))

# add author
gsea <- gsea %>% mutate(
  author = case_when(pathway %in% moffitt_pathways ~ "Moffitt",
                     pathway %in% collison_pathways ~ "Collison",
                     pathway %in% icgc_pathways ~ "ICGC",
                     pathway %in% chan_seng_yue_pathways ~ "Chan Seng Yue")
)

# arrange for readability
gsea$pathway <- factor(gsea$pathway, levels=gsea %>% filter(analysis == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway))
gsea$analysis <- factor(gsea$analysis, levels=c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR"))

p <- ggplot(gsea, aes(x=analysis, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)") +
  facet_wrap(author ~ pathway, nrow=3)
f <- paste0("gsea_barplot_clusters_pdac_subtypes")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=10)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=10)


```


## GSEA LE (leading edge) pdac subtypes

```{r gsea leading edge}

gsea_leading_edge_matrix <- function(x, my_padj_cutoff=0.01, min_gene_sets=2) {
  # get significant pathways
  sig_pathways <- x %>%
    filter(padj < my_padj_cutoff) %>% 
    distinct(pathway) %>% 
    pull()
  # filter results for significant pathways
  x <- x %>%
    filter(pathway %in% sig_pathways) %>%
    arrange(desc(NES))
  sig_pathways <- factor(unique(x$pathway))
  
  # get leading edge genes
  x <- x %>% unnest_longer(leadingEdge, values_to="gene")
  leading_edge_genes <- x %>% 
    count(gene) %>% 
    filter(n >= min_gene_sets) %>% 
    arrange(desc(n)) %>% 
    distinct(gene) %>% 
    pull()
  leading_edge_genes <- factor(leading_edge_genes)
  # filter results for leading edge genes
  x <- x %>% filter(gene %in% leading_edge_genes)
  
  # table with genes (rows) vs pathways (columns)
  x <- x %>% select(pathway, gene, NES) %>% 
    pivot_wider(id_cols=gene, names_from=pathway, values_from=NES)
  m <- x %>% column_to_rownames(var="gene")
  m[is.na(m)] <- 0  

  # filter empty columns
  # m <- m[, abs(colSums(m)) > 1]
  # order by number of pathways sharing the gene
  #m <- m[order(rowSums(m), decreasing=TRUE),]
  #m <- m[,order(colSums(m), decreasing=TRUE)]
  return(m)
}


gsea_leading_edge_matrix2 <- function(x, my_padj_cutoff=0.01, min_gene_sets=2) {
  # get significant pathways
  sig_pathways <- x %>%
    filter(padj < my_padj_cutoff) %>% 
    distinct(pathway) %>% 
    pull()
  # filter results for significant pathways
  x <- x %>%
    filter(pathway %in% sig_pathways) %>%
    arrange(desc(NES))
  sig_pathways <- factor(unique(x$pathway))
  
  # get leading edge genes
  x <- x %>% unnest_longer(leadingEdge, values_to="gene")

  # genes must be in leading edge of all analyses (intersection)
  # and enrichment in the same direction
  num_analyses <- length(unique(x$analysis))
  x <- x %>% group_by(gene, pathway) %>% 
    filter(length(unique(sign(NES))) == 1, n() == num_analyses) %>%
    ungroup() %>%
    mutate(pathway = paste0(analysis, ".", pathway))  
  
  leading_edge_genes <- x %>% 
    count(gene) %>% 
    filter(n >= min_gene_sets) %>% 
    arrange(desc(n)) %>% 
    distinct(gene) %>% 
    pull()
  leading_edge_genes <- factor(leading_edge_genes)
  # filter results for leading edge genes
  x <- x %>% filter(gene %in% leading_edge_genes)
  
  # table with genes (rows) vs pathways (columns)
  x <- x %>% select(pathway, gene, NES) %>% 
    pivot_wider(id_cols=gene, names_from=pathway, values_from=NES)
  m <- x %>% column_to_rownames(var="gene")
  m[is.na(m)] <- 0  

  # filter empty columns
  # m <- m[, abs(colSums(m)) > 1]
  # order by number of pathways sharing the gene
  #m <- m[order(rowSums(m), decreasing=TRUE),]
  #m <- m[,order(colSums(m), decreasing=TRUE)]
  return(m)
}


gsea_leading_edge_matrix3 <- function(x, 
                                      my_padj_cutoff=0.01, 
                                      min_gene_sets=2,
                                      min_analyses=2) {
  # get significant pathways
  sig_pathways <- x %>%
    filter(padj < my_padj_cutoff) %>% 
    distinct(pathway) %>% 
    pull()
  # filter results for significant pathways
  x <- x %>%
    filter(pathway %in% sig_pathways) %>%
    arrange(desc(NES))
  sig_pathways <- factor(unique(x$pathway))
  # assign direction to enrichment score
  x$nes_sign <- factor(sign(x$NES))

  # get leading edge genes
  x <- x %>% unnest_longer(leadingEdge, values_to="gene")

  # genes must be in leading edge of at least 'min_analyses' analysis
  # and enrichment in the same direction
  x <- x %>% 
    group_by(gene, pathway, nes_sign) %>% 
    filter(n() >= min_analyses) %>%
    ungroup() %>%
    mutate(analysis_pathway = paste0(analysis, ".", pathway))  

  leading_edge_genes <- x %>% 
    group_by(analysis) %>%
    count(gene) %>% 
    filter(n >= min_gene_sets) %>% 
    arrange(desc(n)) %>% 
    distinct(gene) %>% 
    pull()
  leading_edge_genes <- factor(leading_edge_genes)
  
  # filter results for leading edge genes
  x <- x %>% filter(gene %in% leading_edge_genes)
  
  # table with genes (rows) vs pathways (columns)
  x <- x %>% select(analysis_pathway, gene, NES) %>% 
    pivot_wider(id_cols=gene, names_from=analysis_pathway, values_from=NES)
  m <- x %>% column_to_rownames(var="gene")
  m[is.na(m)] <- 0
  return(m)
}


gsea_leading_edge_matrix4 <- function(gsea_tbl,
                                      de_tbl,
                                      gsea_padj_cutoff=0.01,
                                      de_padj_cutoff=0.01,
                                      min_gene_sets=2,
                                      min_analyses=2) {
  
  # get leading edge from significant gsea results
  x <- gsea_tbl %>% 
    filter(padj < gsea_padj_cutoff) %>%
    unnest_longer(leadingEdge, values_to="gene")
  # join de table
  x <- left_join(x, de_tbl, by=join_by(gene==gene, analysis==contrast), suffix=c("",".de"))
  sig_de_genes <- x %>% 
    filter(padj.de < de_padj_cutoff) %>%
    distinct(gene) %>%
    pull()
  # keep genes that are significantly de
  x <- x %>% filter(gene %in% sig_de_genes)
  # get leading edge genes from significant gsea results
  x$nes_sign <- factor(sign(x$NES))

  # genes must be in leading edge of at least 'min_analyses' (same direction)
  x <- x %>% 
    group_by(gene, pathway, nes_sign) %>% 
    filter(n() >= min_analyses) %>%
    ungroup() %>%
    mutate(analysis_pathway = paste0(analysis, ".", pathway))  

  # gsea result must be significant in at least "min_gene_sets" pathways
  leading_edge_genes <- x %>% 
    group_by(analysis, nes_sign) %>%
    count(gene) %>% 
    filter(n >= min_gene_sets) %>% 
    arrange(desc(n)) %>% 
    distinct(gene) %>% 
    pull()
  leading_edge_genes <- factor(leading_edge_genes)
  # filter results for leading edge genes
  x <- x %>% filter(gene %in% leading_edge_genes)
  
  # table with genes (rows) vs analysis-pathways (columns)
  x <- x %>% select(analysis_pathway, gene, log2fc) %>% 
    pivot_wider(id_cols=gene, names_from=analysis_pathway, values_from=log2fc)
  m <- x %>% column_to_rownames(var="gene")
  m[is.na(m)] <- 0
  return(m)
}

# setup heatmap annotations
gene_annot <- fds$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot) %>%
  column_to_rownames("gene") %>%
  as.data.frame()


#
# basallike genes
#
padj_cutoff <- 0.01
pathways <- c("Moffitt.F6_BasalLike.top250",
              "Collisson.QM",
              "ICGC.Squamous.Up",
              "Chan_Seng_Yue.BasalA",
              "Chan_Seng_Yue.BasalB")
contrasts <- c("clusterHRvNL", "clusterHRvLR", "clusterLRvNL")

x <- filter(gsea_pdacr, analysis %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix4(x, de, gsea_padj_cutoff=padj_cutoff, min_gene_sets=2, min_analyses=1)

# show intersection with CPTAC
genes <- rownames(m)
length(intersect(genes, gs_cptac$CPTAC_RNA_ALL_UP))
length(intersect(genes, gs_cptac$CPTAC_PROT_ALL_UP))
length(intersect(genes, union(gs_cptac$CPTAC_RNA_ALL_UP, gs_cptac$CPTAC_PROT_ALL_UP)))
length(intersect(genes, gs_hpa_panc$HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS))
# agreement across studies
table(rowSums(sign(m)))

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 8
w <- 6
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_basal.pdf"), width=w, height=h)


#
# classical genes
#
padj_cutoff <- 0.01
pathways <- c(
  "Chan_Seng_Yue.ClassicalA",
  "Chan_Seng_Yue.ClassicalB",
  "Collisson.Classical",
  "ICGC.Immunogenic.Up",
  "Moffitt.F8_Classical.top250"
)
contrasts <- c("clusterHRvNL", "clusterHRvLR", "clusterLRvNL")

x <- filter(gsea_pdacr, analysis %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix4(x, de, gsea_padj_cutoff=padj_cutoff, min_gene_sets=2, min_analyses=1)

# show intersection with CPTAC
genes <- rownames(m)
length(intersect(genes, gs_cptac$CPTAC_RNA_ALL_UP))
length(intersect(genes, gs_cptac$CPTAC_PROT_ALL_UP))
length(intersect(genes, union(gs_cptac$CPTAC_RNA_ALL_UP, gs_cptac$CPTAC_PROT_ALL_UP)))
length(intersect(genes, gs_hpa_panc$HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS))
# agreement across studies
table(rowSums(sign(m)))

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 8
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_classical.pdf"), width=w, height=h)


#
# exocrine genes
#
padj_cutoff <- 0.01
pathways <- c(
  "Collisson.Exocrine",
  "ICGC.ADEX.Up",
  "Moffitt.F3_Exocrine.top250"
)
contrasts <- c("clusterHRvNL", "clusterHRvLR", "clusterLRvNL")

x <- filter(gsea_pdacr, analysis %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix4(x, de, gsea_padj_cutoff=padj_cutoff, min_gene_sets=2, min_analyses=1)

# show intersection with CPTAC/prognosis
genes <- rownames(m)
length(intersect(genes, gs_cptac$CPTAC_RNA_ALL_DN))
length(intersect(genes, gs_cptac$CPTAC_PROT_ALL_UP))
length(intersect(genes, union(gs_cptac$CPTAC_RNA_ALL_DN, gs_cptac$CPTAC_PROT_ALL_DN)))
length(intersect(genes, gs_hpa_panc$HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS))
length(intersect(genes, gs_hpa_panc$HPA_PANCREATIC_CANCER_FAV_PROGNOSIS))
# agreement across studies
table(rowSums(sign(m)))

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 7
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_exocrine.pdf"), width=w, height=h)

```

## GSEA LE hallmark

```{r gsea leading edge hallmark}

# setup heatmap annotations
gene_annot <- fds$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na")) %>% 
  mutate(moffitt_basallike = case_when(gene %in% gs_pdacr$Moffitt.F6_BasalLike.top250 ~ "up",
                                       TRUE ~ "no"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot, moffitt_basallike) %>%
  column_to_rownames("gene") %>%
  as.data.frame()

pathways <- c(
  "HALLMARK_INTERFERON_ALPHA_RESPONSE",
  "HALLMARK_INTERFERON_GAMMA_RESPONSE",
  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION"
)
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

x <- filter(gsea_hallmark, analysis %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix4(x, de, gsea_padj_cutoff=0.01, de_padj_cutoff=0.01, 
                               min_gene_sets=1, min_analyses=2)

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 10
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_hallmark.pdf"), width=w, height=h)

```

# GSVA

## GSVA hallmark pathways

```{r gsva hallmark}

# hallmark pathways
gene_sets <- c("H_TNFA_SIGNALING_VIA_NFKB", 
               "H_INTERFERON_GAMMA_RESPONSE",
               "H_INTERFERON_ALPHA_RESPONSE",
               "H_EPITHELIAL_MESENCHYMAL_TRANSITION")

path_aoi_shapes = list(
  "nl"=21,
  "lgd"=25,
  "hgd"=24,
  "inv"=23
)

hallmark_color_scales = list(
  "H_TNFA_SIGNALING_VIA_NFKB" = "#ff0000",
  "H_INTERFERON_GAMMA_RESPONSE" = "#ffff00",
  "H_INTERFERON_ALPHA_RESPONSE" = "#ffaa00",
  "H_EPITHELIAL_MESENCHYMAL_TRANSITION" = "#0022ff"
)


s <- fds$samples
y <- log2_norm_cpm
x <- as_tibble(gsva_res[gene_sets,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, bind_cols(fds$samples, as_tibble(pca_res$x)), by=join_by(aoi == aoi))
x$path <- factor(x$path, levels=c("ipmn_lgd_nl_epithelial", "ipmn_lgd_lgd_epithelial",
                                  "ipmn_inv_nl_epithelial", "ipmn_inv_lgd_epithelial",
                                  "ipmn_inv_hgd_epithelial", "ipmn_inv_inv_epithelial"))
x$path_aoi <- factor(x$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))

x <- x %>% 
  group_by(patient, pathway, path_aoi) %>%
  mutate(medianvalue = median(value)) %>%
  ungroup()

# boxplot individual patients
p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0.05, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "boxplot_gsva_hallmark_inflammation")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# boxplots
#
p <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.05, dodge.width=0.75),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = hallmark_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"))
f <- file.path(plot_dir, "boxplot_gsva_hallmark_path")
ggsave(paste0(f, ".pdf"), p, width=7, height=5)

p <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.05, dodge.width=0.75),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = hallmark_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"))
f <- file.path(plot_dir, "boxplot_gsva_hallmark_cluster")
ggsave(paste0(f, ".pdf"), p, width=6, height=4)

```


## GSVA pdac subtypes

```{r}

gene_sets <- c("Moffitt.F3_Exocrine.top250",
               "Moffitt.F8_Classical.top250",
               "Moffitt.F6_BasalLike.top250")

moffitt_color_scales = list(
  "Moffitt.F3_Exocrine.top250" = "#4400bb",
  "Moffitt.F8_Classical.top250" = "#ffff00",
  "Moffitt.F6_BasalLike.top250" = "#ff00ff"
)

path_aoi_shapes = list(
  "nl"=21,
  "lgd"=25,
  "hgd"=24,
  "inv"=23
)

s <- fds$samples
y <- log2_norm_cpm
x <- as_tibble(gsva_res[gene_sets,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, bind_cols(fds$samples, as_tibble(pca_res$x)), by=join_by(aoi == aoi))
x$path <- factor(x$path, levels=c("ipmn_lgd_nl_epithelial", "ipmn_lgd_lgd_epithelial",
                                  "ipmn_inv_nl_epithelial", "ipmn_inv_lgd_epithelial",
                                  "ipmn_inv_hgd_epithelial", "ipmn_inv_inv_epithelial"))
x$path_aoi <- factor(x$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))
x$pathway <- factor(x$pathway, levels=gene_sets)

p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "boxplot_patient_gsva_pdacr_subtypes")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# PCA plot
#
p <- ggplot(x, aes(x=PC1, y=PC2, color=value, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  theme_minimal() +
  theme(legend.text = element_text(size=7)) +
  labs(x="PC1", y="PC2", title="Moffitt Subtypes") + 
  facet_wrap(~ pathway, nrow=1)
f <- file.path(plot_dir, "pca_featureplot_moffitt")
ggsave(paste0(f, ".pdf"), plot=p, width=9, height=3.5)
ggsave(paste0(f, ".png"), plot=p, width=9, height=3.5)


#
# lineplot
#
x <- x %>% 
  group_by(patient, pathway, path_aoi) %>%
  mutate(medianvalue = median(value)) %>%
  ungroup()

p <- ggplot(x, aes(color=patient, shape=path_specimen, group=patient)) + 
  geom_point(aes(x=path_aoi, y=value)) +
  geom_line(aes(x=path_aoi, y=medianvalue, group=patient)) +
  geom_hline(yintercept=0) +
  scale_color_manual(values = color_scales$patient) +
  facet_grid(pathway ~ path_specimen, scales="free_x") + 
  theme_minimal()
f <- file.path(plot_dir, "lineplot_gsva_moffitt_path")
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

#
# boxplots
#
p <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75, linewidth=0.5) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_path")
ggsave(paste0(f, ".pdf"), p, width=5, height=4)

p <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_cluster")
ggsave(paste0(f, ".pdf"), p, width=4, height=4)


p <- ggplot(filter(x, path_aoi %in% c("nl", "lgd")), aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_cluster_nllgd")
ggsave(paste0(f, ".png"), p, width=5, height=4)
ggsave(paste0(f, ".pdf"), p, width=5, height=4)


```

## GSVA PCA correlations

```{r gsva pca correlation}

gene_sets <- c("Moffitt.F3_Exocrine.top250",
               "Moffitt.F8_Classical.top250",
               "Moffitt.F6_BasalLike.top250")
pc_names <- c("PC1", "PC2")

x <- bind_cols(fds$samples, as_tibble(pca_res$x), t(gsva_res))
x <- x %>% select(aoi, path_specimen, path_aoi, cluster_name, all_of(pc_names), all_of(gene_sets))

coefs <- matrix(nrow=length(pc_names), ncol=length(gene_sets), dimnames=list(pc_names, gene_sets))
pvals <- matrix(nrow=length(pc_names), ncol=length(gene_sets), dimnames=list(pc_names, gene_sets))
for (i in pc_names) {
  for (j in gene_sets) {
    res <- cor.test(pull(x, i), pull(x, j), method="pearson", exact=TRUE)
    coefs[i, j] <- res$estimate
    pvals[i, j] <- res$p.value
  }
}
coefs <- as_tibble(coefs, rownames="pc_name") %>% 
  pivot_longer(cols=gene_sets, names_to="gs_name", values_to="coef")
pvals <- as_tibble(pvals, rownames="pc_name") %>% 
  pivot_longer(cols=gene_sets, names_to="gs_name", values_to="pvals")


x <- x %>% pivot_longer(cols=all_of(gene_sets), names_to="gs_name", values_to="gsva")
x <- x %>% pivot_longer(cols=all_of(pc_names), names_to="pc_name", values_to="pc_value")
x$gs_name <- factor(x$gs_name, levels=gene_sets)
x$pc_name <- factor(x$pc_name, levels=pc_names)
x <- left_join(x, coefs, by=join_by(pc_name, gs_name))
x <- left_join(x, pvals, by=join_by(pc_name, gs_name))
x <- x %>% mutate(label=paste0("r=", round(coef, 3), " p=", signif(pvals, 3)))


p <- ggplot(x, aes(x=pc_value, y=gsva)) + 
  geom_point(aes(color=path_aoi, shape=path_specimen)) + 
  geom_smooth(color="black", linetype="dashed", method='lm', formula=y~x) +
  scale_color_manual(values = color_scales$path_aoi) +
  geom_text(data=distinct(x, gs_name, pc_name, .keep_all=TRUE), 
            aes(label=label), x=Inf, y=Inf, size=3, vjust="inward", hjust="inward") + 
  theme_minimal() + 
  theme(aspect.ratio=1) + 
  facet_grid(pc_name ~ gs_name)
f <- file.path(plot_dir, "scatter_cor_pca_moffitt")
ggsave(paste0(f, ".png"), p, width=7, height=5)
ggsave(paste0(f, ".pdf"), p, width=7, height=5)


```

# NL-LGD analysis

## NL-LGD PCA

```{r nl_lgd cohort}

# make nl_lgd cohort
s <- fds$samples %>% filter(path_aoi %in% c("nl","lgd"))
y <- log2_norm_cpm[, s$aoi]

# subset most variable genes
ntop <- 250
myspan <- 0.5
y <- most_var_genes(y, ntop=ntop, span=myspan)

# plot dimensions
width <- 4
height <- 3

# PCA
nllgd.pca <- prcomp(t(y), center=TRUE, scale.=TRUE)
nllgd.pca_var <- round(100 * nllgd.pca$sdev^2 / sum( nllgd.pca$sdev^2 ), 2)
nllgd.pca_var_txt <- paste(colnames(nllgd.pca$x), " (", paste(as.character(nllgd.pca_var), "%", ")", sep=""), sep="")

# percent variance of each PC
p <- fviz_eig(nllgd.pca, col.var="blue", addlabels=TRUE)
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_eig.pdf"))
ggsave(f, p)

# contributions of variables to PC1 and PC2
p <- fviz_contrib(nllgd.pca, choice="var", axes = 1:2, top=fviz_ntop) +
  theme(axis.text.x = element_text(angle=90))
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_contrib.pdf"))
ggsave(f, p)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  nllgd.pca,
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_pca_var.pdf"))
ggsave(f, p, width=7, height=7)

#
# PCA plots
#
x <- bind_cols(s, as_tibble(nllgd.pca$x))

# bg_ks_dist
p <- ggplot(x, aes(x=PC1, y=PC2, color=bg_auc)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_bg_auc.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# patient
p <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = pals::cols25()) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_patient.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path specimen
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_path_specimen.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path aoi
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_path_aoi.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path
p <- ggplot(x, aes(x=PC1, y=PC2, color=path, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_path.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)


```


## NL-LGD PCA Clustering

```{r nllgd cluster}

# make nl_lgd cohort
s <- fds$samples %>% filter(path_aoi %in% c("nl","lgd"))
y <- log2_norm_cpm[, s$aoi]

# choose number of clusters
p <- fviz_nbclust(nllgd.pca$x, FUNcluster=hcut, method="silhouette")
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_nbclust.pdf"))
ggsave(f, plot=p, device="pdf", width=4, height=4)

# set number of clusters
nclusters <- 3

# hierarchical clustering based on first two PCs
# x <- nllgd.pca$x[,1:2]
# d <- dist(x)
# hc <- hclust(d, method="ward.D2")
# memb <- cutree(hc, k=nclusters)
# memb <- paste0("c", memb)

# commented out: decided to use hierarchical clustering
# kmeans clustering based on PCs
# km <- kmeans(x, centers=nclusters)
# memb <- paste0("c", km$cluster)

# pam
x <- nllgd.pca$x[,1:2]
d <- dist(x)
pamx <- cluster::pam(d, k=3)
memb <- paste0("c", pamx$clustering)

# add cluster membership to sample properties
s$cluster_nllgd <- memb
# save clustering
nllgd_samples <- s

```

## NL-LGD PCA cluster plots and naming

```{r}

# plot clustering
x <- bind_cols(nllgd_samples, as_tibble(nllgd.pca$x))

# cluster
p <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_nllgd, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster) +
  theme_bw() +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
p

# identify clusters by plot
nllgd_samples <- nllgd_samples %>% mutate(
  cluster_nllgd_name = case_when(cluster_nllgd == "c1" ~ "cHR",
                                 cluster_nllgd == "c2" ~ "cLR",
                                 cluster_nllgd == "c3" ~ "cNL")
)
x <- bind_cols(nllgd_samples, as_tibble(nllgd.pca$x))

# path specimen
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# path aoi
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# patient
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = pals::cols25()) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# cluster (original)
p4 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster_name) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# cluster (nl-lgd)
p5 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_nllgd_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster_name) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

p <- p1 + p2 + p3 + p4 + theme(legend.position="none") + p5 + plot_layout(nrow=1)
f <- file.path(plot_dir, paste0("pca_nllgd_pca_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=22, height=6)


```


## NL-LGD DE analysis

```{r nl_lgd de}

nllgd_de <- NULL

#
# DE analysis of INV specimens
#
s <- filter(nllgd_samples, path_specimen == "ipmn_inv")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd"))
y <- log2_norm_cpm[, s$aoi]

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(invlgd_vs_invnl = path_aoilgd - path_aoinl,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
nllgd_de <- bind_rows(nllgd_de, res)


#
# DE analysis of LGD lesions
#
s <- filter(nllgd_samples, path_specimen == "ipmn_lgd")
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd"))
y <- log2_norm_cpm[, s$aoi]

design <- model.matrix(~0 + path_aoi + patient, s)
contrasts <- makeContrasts(lgdlgd_vs_lgdnl = path_aoilgd - path_aoinl,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
nllgd_de <- bind_rows(nllgd_de, res)


#
# setup cohort of all samples 
#
s <- nllgd_samples
s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd"))
s$path_specimen <- factor(s$path_specimen, levels=c("ipmn_lgd", "ipmn_inv"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, 
                         levels=c("ipmn_lgd_nl", "ipmn_lgd_lgd", 
                                  "ipmn_inv_nl", "ipmn_inv_lgd"))
s$cluster <- factor(s$cluster_name, levels=c("cNL", "cLR", "cHR"))
y <- log2_norm_cpm[, s$aoi]


#
# DE analysis INV vs LGD specimens
#
design <- model.matrix(~0 + path_patient, s)
colnames(design) <- levels(s$path_patient)
colnames(design)
contrasts <- makeContrasts(
  invnl_vs_lgdnl = ipmn_inv_nl - ipmn_lgd_nl,
  invlgd_vs_lgdlgd = ipmn_inv_lgd - ipmn_lgd_lgd,
  levels=design
)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
nllgd_de <- bind_rows(nllgd_de, res)

#
# DE analysis of clusters
#
design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(clusterHRvLR = clustercHR - clustercLR,
                           clusterHRvNL = clustercHR - clustercNL,
                           clusterLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
nllgd_de <- bind_rows(nllgd_de, res)


# write results
table(nllgd_de$contrast)
write_xlsx(de, file.path(results_dir, "de_nllgd.xlsx"))

```

## NL-LGD DE correlations

```{r nllgd de correlations}

# merge de analysis by gene
nllgd_de_merged <- nllgd_de %>% select(gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))

cor.test(nllgd_de_merged$log2fc_clusterHRvNL, nllgd_de_merged$log2fc_clusterLRvNL)
cor.test(nllgd_de_merged$log2fc_clusterHRvNL, nllgd_de_merged$log2fc_clusterHRvLR)
cor.test(nllgd_de_merged$log2fc_clusterLRvNL, nllgd_de_merged$log2fc_clusterHRvLR)


```

## NL-LGD DE plots

```{r nllgd de plots}

# volcano plots
for (mycontrast in unique(nllgd_de$contrast)) {
  print(mycontrast)
  x <- filter(de, contrast == mycontrast)
  p <- plot_de_volcano(x) + 
    labs(title="Volcano Plot", subtitle=mycontrast)
  f <- paste0("volcano_nllgd_de_", mycontrast)
  ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=8)
  ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=8)
}


#
# scatter plots
#
nlabel <- 10
width <- 8
height <- 8


# clusters
p <- plot_de_scatter2d(nllgd_de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterHRvLR", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "scatter2d_nllgd_de_x_clusterHRvNL_y_clusterHRvLR")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(nllgd_de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterLRvNL", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterLRvNL (y-axis)")
f <- file.path(plot_dir, "scatter2d_nllgd_de_x_clusterHRvNL_y_clusterLRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)



# path
p <- plot_de_scatter2d(nllgd_de_merged, 
                       xcontrast="invlgd_vs_invnl", 
                       ycontrast="lgdlgd_vs_lgdnl", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="invlgd_vs_invnl (x-axis) lgdlgd_vs_lgdnl (y-axis)")
f <- file.path(plot_dir, "scatter2d_nllgd_de_x_invlgd_vs_invnl_y_lgdlgd_vs_lgdnl")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(nllgd_de_merged, 
                       xcontrast="invnl_vs_lgdnl", 
                       ycontrast="invlgd_vs_lgdlgd", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="invnl_vs_lgdnl (x-axis) invlgd_vs_lgdlgd (y-axis)")
f <- file.path(plot_dir, "scatter2d_nllgd_de_x_invnl_vs_lgdnl_y_invlgd_vs_lgdlgd")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)



#
# 3D scatter plot
#
#https://plotly.com/r/reference/#scatter-mode

tbl <- nllgd_de_merged
xcontrast <- "clusterHRvNL"
ycontrast <- "clusterLRvNL"
zcontrast <- "clusterHRvLR"

xlog2fc_formula <- as.formula(paste0("~log2fc_", xcontrast))
ylog2fc_formula <- as.formula(paste0("~log2fc_", ycontrast))
zlog2fc_formula <- as.formula(paste0("~log2fc_", zcontrast))
xlog2fc <- sym(paste0("log2fc_", xcontrast))
ylog2fc <- sym(paste0("log2fc_", ycontrast))
zlog2fc <- sym(paste0("log2fc_", zcontrast))
xde <- sym(paste0("de_", xcontrast))
yde <- sym(paste0("de_", ycontrast))
zde <- sym(paste0("de_", zcontrast))

tbl <- tbl %>% mutate(
  subtype = paste0("x", {{xde}}, "_y", {{yde}}, "_z", {{zde}}),
  marker_size = case_when(subtype == "xno_yno_zno" ~ 0.001,
                          TRUE ~ 2),
  log2fc_abs_x = case_when({{xde}} == "no" ~ 0,
                           TRUE ~ abs({{xlog2fc}})),
  log2fc_abs_y = case_when({{yde}} == "no" ~ 0,
                           TRUE ~ abs({{ylog2fc}})),
  log2fc_abs_z = case_when({{yde}} == "no" ~ 0,
                           TRUE ~ abs({{zlog2fc}})),
  log2fc_abs_max = pmax(log2fc_abs_x, log2fc_abs_y, log2fc_abs_z)
)

marker_colors = c(
  xup_yup_zup = "#ffffff",
  xup_yup_zno = "#ffff88",
  xup_yup_zdn = "#ffff00",
  xup_yno_zup = "#ff88ff",
  xup_yno_zno = "#ff8888",
  xup_yno_zdn = "#ff8800",
  xup_ydn_zup = "#ff00ff",
  xup_ydn_zno = "#ff0088",
  xup_ydn_zdn = "#ff0000",
  xno_yup_zup = "#88ffff",
  xno_yup_zno = "#88ff88",
  xno_yup_zdn = "#88ff00",
  xno_yno_zup = "#8888ff",
  xno_yno_zno = "#cccccccc",
  xno_yno_zdn = "#888800",
  xno_ydn_zup = "#8800ff",
  xno_ydn_zno = "#880088",
  xno_ydn_zdn = "#880000",
  xdn_yup_zup = "#88ffff",
  xdn_yup_zno = "#00ff88",
  xdn_yup_zdn = "#00ff00",
  xdn_yno_zup = "#0088ff",
  xdn_yno_zno = "#008888",
  xdn_yno_zdn = "#008800",
  xdn_ydn_zup = "#0000ff",
  xdn_ydn_zno = "#000088",
  xdn_ydn_zdn = "#000000"
)

label_data <- filter(tbl, subtype != "xno_yno_zno", log2fc_abs_max > 0) %>% 
  group_by(subtype) %>%
  slice_max(log2fc_abs_max, n=5) %>%
  ungroup()

label_text <- list(
  family = "helvetica",
  size = 6,
  color = toRGB("grey50"))

p <- plot_ly(filter(tbl, subtype == "xno_yno_zno"), 
             type="scatter3d",
             mode="markers",
             x=xlog2fc_formula,
             y=ylog2fc_formula,
             z=zlog2fc_formula,
             color=~subtype,
             size=~marker_size,
             colors=marker_colors) %>%
  add_markers(data = filter(tbl, subtype != "xno_yno_zno"),
              marker = list(line = list(color = '#000000', width=1)),
              text=~gene) %>%
  add_text(data = label_data,
           x=xlog2fc_formula,
           y=ylog2fc_formula,
           z=zlog2fc_formula,
           text=~gene,
           textfont=label_text)


#
# Heatmap
#
# setup heatmap annotations
gene_annot <- fds$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na")) %>% 
  mutate(moffitt_basallike = case_when(gene %in% gs_pdacr$Moffitt.F6_BasalLike.top250 ~ "up",
                                       TRUE ~ "no"),
         moffitt_classical = case_when(gene %in% gs_pdacr$Moffitt.F8_Classical.top250 ~ "up",
                                       TRUE ~ "no"),
         moffitt_exocrine = case_when(gene %in% gs_pdacr$Moffitt.F3_Exocrine.top250 ~ "up",
                                       TRUE ~ "no"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot, 
         moffitt_basallike, moffitt_classical, moffitt_exocrine) %>%
  column_to_rownames("gene") %>%
  as.data.frame()

annot_col <- column_to_rownames(nllgd_samples, "aoi") %>% 
  select(patient, path_aoi, path_specimen, cluster_name, cluster_nllgd_name) %>%
  as.data.frame()


# de gene params
padj_cutoff <- 0.01
log2fc_cutoff <- 3
ntop <- 5
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

g <- filter(nllgd_de, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)
cluster_marker_genes <- g

g <- c("OLFM4", "CRISP3")
#g <- label_data %>% pull(gene)

# centered gene expression (subtract mean)
s <- nllgd_samples
x <- log2_norm_cpm[g, s$aoi]
u <- apply(x, 1, mean)
x <- sweep(x, 1, u, FUN="-")
xmin <- -4
xmax <- 4

# standardize (divide by standard deviation)
# s <- fds$samples
# snl <- fds$samples %>% filter(cluster == "c3")
# x <- log2_norm_cpm[g, s$aoi]
# u <- apply(x[, snl$aoi], 1, median)
# x <- sweep(x, 1, u, FUN="-")
# v <- pmax(apply(x, 1, sd), 1)
# x <- sweep(x, 1, v, FUN="/")
# xmin <- -3
# xmax <- 3

# heatmap
f <- file.path(plot_dir,  paste0("heatmap_nllgd_de_cluster_markers_", ngenes, "_genes"))
h <- 10
w <- 10

p <- pheatmap(x,
              scale = "none",
              color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
              breaks = seq(xmin, xmax, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              annotation_col = annot_col[s$aoi,],
              annotation_row = annot_row[g,],
              annotation_colors = color_scales,
              border_color = NA,
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "complete",
              fontsize = 6,
              fontsize_row = 6,
              fontsize_col = 6,
              filename = paste0(f, ".png"), 
              width = w,
              height = h)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)



```

## NL-LGD GSEA

```{r nllgd gsea}

#
# Setup analysis
#
table(nllgd_de$contrast)
analyses <- unique(nllgd_de$contrast)

#
# CPTAC GSEA
#
prefix <- "gsea_nllgd_cptac"
padj_cutoff <- 0.01
gs <- gs_cptac
gsea <- run_batch_fgsea(analyses, nllgd_de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
gsea_nllgd_cptac <- as_tibble(gsea)
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# HPA (TCGA) GSEA
#
prefix <- "gsea_nllgd_hpa_panc"
padj_cutoff <- 0.01
gs <- gs_hpa_panc
gsea <- run_batch_fgsea(analyses, nllgd_de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
gsea_nllgd_hpa_panc <- as_tibble(gsea)
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# PDAC subtypes GSEA
#
prefix <- "gsea_nllgd_pdacr"
padj_cutoff <- 0.01
gs <- gs_pdacr
gsea <- run_batch_fgsea(analyses, nllgd_de, gs, prefix, gsea_plot_dir, padj_cutoff, do_plots=FALSE)
gsea_nllgd_pdacr <- as_tibble(gsea)
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# Hallmark gene set analysis
#
prefix <- "gsea_nllgd_hallmark"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, nllgd_de, gs_hallmark, prefix, gsea_plot_dir, 
                        padj_cutoff, do_plots=FALSE)
gsea_nllgd_hallmark <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_name) %>% distinct(), 
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

#
# 3CA gene set analysis
#
prefix <- "gsea_nllgd_3ca"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, nllgd_de, gs_3ca, prefix, gsea_plot_dir, 
                        padj_cutoff, do_plots=FALSE)
gsea_nllgd_3ca <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_name) %>% distinct(), 
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)


#
# Gene Ontology analysis
#
prefix <- "gsea_nllgd_gobp"
padj_cutoff <- 0.01
gsea <- run_batch_fgsea(analyses, nllgd_de, gs_gobp, prefix, gsea_plot_dir, 
                        padj_cutoff, do_plots=FALSE)
gsea_nllgd_gobp <- as_tibble(gsea) %>%
  left_join(select(msigdb_gene_sets, gs_cat, gs_name) %>% distinct(),
            by=c("pathway"="gs_name"))
# write gsea results
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(gsea, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(gsea, f)

```

## NL-LGD GSEA plots

```{r nllgd gsea plots}

#
# CPTAC and HPA
#
pathways <- c("CPTAC_PROT_ALL_DN", "CPTAC_PROT_ALL_UP",
              "CPTAC_RNA_ALL_DN", "CPTAC_RNA_ALL_UP",
              "HPA_PANCREATIC_CANCER_FAV_PROGNOSIS",
              "HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS")
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")
padj_cutoff <- 0.01

x <- bind_rows(gsea_nllgd_cptac, gsea_nllgd_hpa_panc) %>% 
  filter(analysis %in% contrasts, pathway %in% pathways) %>%
#  mutate(group="nllgd", analysis = paste0("nllgd.", sub("cluster", "", analysis)))
  mutate(group="nllgd", analysis = sub("cluster", "", analysis))
y <- bind_rows(gsea_cptac, gsea_hpa_panc) %>% 
  filter(analysis %in% contrasts, pathway %in% pathways) %>%
#  mutate(group="all", analysis = paste0("all.", sub("cluster", "", analysis)))
  mutate(group="all", analysis = sub("cluster", "", analysis))

gsea <- bind_rows(x, y)
#gsea <- bind_rows(gsea_cptac, gsea_hpa_panc)
#gsea <- gsea %>%
#  filter(pathway %in% pathways, analysis %in% contrasts)
#gsea <- gsea %>% filter(analysis %in% contrasts)
gsea <- mutate(gsea, neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))

p <- ggplot(gsea, aes(x=reorder(analysis, NES), y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_grid(pathway ~ group) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_nllgd_barplot_clusters_pdac")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=9)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=9)


#
# Hallmark
#
padj_cutoff <- 0.01
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

x <- bind_rows(gsea_nllgd_hallmark) %>% 
  filter(analysis %in% contrasts) %>%
  mutate(group="nllgd", analysis = sub("cluster", "", analysis))
y <- bind_rows(gsea_hallmark) %>% 
  filter(analysis %in% contrasts) %>%
  mutate(group="all", analysis = sub("cluster", "", analysis))
gsea <- bind_rows(x, y)
gsea <- mutate(gsea, pathway = gsub("HALLMARK_", "", pathway))
gsea <- gsea %>% mutate(
  neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  sig = padj < padj_cutoff)

# only keep significant results (at least one analysis)
gsea <- gsea %>% group_by(pathway) %>%
  filter(any(padj < padj_cutoff)) %>%
  ungroup()

# arrange by cluster 1v2 for readability
pathway_levels <- gsea %>% 
  filter(analysis == "HRvLR") %>%
  arrange(NES) %>%
  distinct(pathway) %>%
  pull(pathway)
gsea$pathway <- factor(gsea$pathway, levels=pathway_levels)

p <- ggplot(gsea, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  xlim(-2.5,2.5) +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=6)) +
#  theme(axis.text.x = element_text(color="black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_grid(analysis ~ group) +
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")
f <- paste0("gsea_nllgd_barplot_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=8)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=8)


#
# Moffitt
# 
padj_cutoff <- 0.01
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")
pathways <- c("Moffitt.F6_BasalLike.top250", 
              "Moffitt.F8_Classical.top250",
              "Moffitt.F3_Exocrine.top250")

x <- gsea_nllgd_pdacr %>% 
  filter(analysis %in% contrasts, pathway %in% pathways) %>%
  mutate(group="nllgd", analysis = sub("cluster", "", analysis))
y <- gsea_pdacr %>% 
  filter(analysis %in% contrasts, pathway %in% pathways) %>%
  mutate(group="all", analysis = sub("cluster", "", analysis))
gsea <- bind_rows(x, y)
gsea <- mutate(gsea, neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))

p <- ggplot(gsea, aes(x=reorder(analysis, NES), y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_grid(pathway ~ group) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_nllgd_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=6)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=6)



```


# End of working analysis



# gene plots

```{r nllgd gene plots}


plot_gene <- function(s, x, gene, aes_x, aes_color) {
  exprs <- unlist(x[gene, ])
  x <- s %>% add_column(gene = exprs)
  p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
    geom_boxplot(outlier.shape = NA, width=0.75) +
#    geom_violin(scale="width", width=0.5) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.5)) +
    ggtitle(paste0("Gene: ", gene)) +
    theme_minimal() + 
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
    theme(axis.line = element_line(color = "black"))
  return(p)
}

plot_gene2 <- function(s, x, gene, aes_x, aes_color) {
  exprs <- unlist(x[gene, ])
  x <- s %>% add_column(gene = exprs)
  p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
    geom_boxplot(outlier.shape=NA, width=0.75) +
    geom_point(position=position_jitterdodge(jitter.width=0.1), 
               aes(group={{aes_color}}), pch=21) +
    ggtitle(paste0("Gene: ", gene)) +
    theme_bw() + 
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
    theme(axis.line = element_line(color = "black"))
  return(p)
}

s <- fds$samples %>% filter(path_aoi %in% c("nl","lgd"))
y <- log2_norm_cpm[, s$aoi]


s <- fds$samples
y <- log2_norm_cpm

genes <- c("OLFM4", "MALL", "TNS4", "CLDN4", "DUOX2")

w <- 5
h <- 4


g <- "PIKFYVE"

x <- s %>% add_column(gene = y[g,])
p <- ggplot(x, aes(x=cluster_name, y=gene, fill=cluster_name)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(aes(group=cluster_name),
             position=position_jitterdodge(jitter.width=0.1))


p <- ggplot(x, aes(x={{ aes_x }}, y=gene, fill={{aes_color}})) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1), 
             aes(group={{aes_color}}), pch=21) +
  ggtitle(paste0("Gene: ", gene)) +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(axis.line = element_line(color = "black"))
return(p)
  

p <- plot_gene(s, y, "OLFM4", path_aoi, path_specimen) + 
  scale_fill_manual(values=color_scales$path_aoi)


for (g in genes) {
  if (!(g %in% rownames(y))) { next }
  p <- plot_gene(s, y, g, path_patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
}

#s <- s %>% filter(study == "ipmn")
#y <- y[, s$aoi]
genes <- c("CLU", "TNFAIP2", "MUC3A", "COL1A1", "MALL")
for (g in genes) {
  p <- plot_gene2(s, y, g, path_patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
  p <- plot_gene2(s, y, g, patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_by_patient_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
  p <- plot_gene2(s, y, g, patient_slide, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_by_patient_slide_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
}


```




# End of working analysis

```{r nllgd plots}

#
# plots with just normal and lgd regions
#
alpha_aoi <- c("nl"=1, "lgd"=1, "hgd"=0.1, "inv"=0.1)
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen, alpha=path_aoi)) +
  geom_point(size=3) +
  scale_alpha_manual(values=alpha_aoi) +
  scale_color_manual(values=color_scales$path_aoi) +
  theme_bw() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_path_aoi_nllgd.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

p <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen, alpha=path_aoi)) +
  geom_point(size=3) +
  scale_alpha_manual(values=alpha_aoi) +
  scale_color_manual(values=color_scales$cluster_name) +
  theme_bw() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_path_cluster_nllgd.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)



```



# Maitra NKX6-2 paper

```{r maitra nkx6-2}


y <- log2_norm_cpm
s <- fds$samples
s$patient <- factor(s$patient)
s$patient_slide <- factor(paste0(s$patient, "_", s$slide))
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "adm", "panin", "lgd", "hgd", "inv", "pdac"))
s$path_specimen <- factor(s$path_specimen, levels=c("nl", "ipmn_lgd", "ipmn_inv", "pnet", "pdac"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, 
                         levels=c("nl_nl", "ipmn_lgd_nl", "ipmn_inv_nl", 
                                  "pnet_nl", "pdac_nl", "nl_adm", "pdac_adm", 
                                  "nl_panin", "pdac_panin", 
                                  "ipmn_lgd_lgd", "ipmn_inv_lgd", 
                                  "ipmn_inv_hgd", "ipmn_inv_inv", 
                                  "pdac_pdac"))

width = 8
height = 5
genes <- c("MUC5AC", "S100P", "SYTL2", "LRRTM4", "ST6GALNAC2", "TMEM67",
           "EIF4A3", "CLDN18", "MSMB", 
           "KRT17", "LAMB3", "LAMC2", "PNLIP", "AMY1A", "CLU",
           "TNFAIP2", "MMP7", "MALL", "MUC6", "CEACAM5", "CEACAM1",
           "PGC", "GALNT6", "CELA3A", "BPIFB1", "EPPK1", "PITX1", "LIF", 
           "CTRB1", "NKX6-2", "OLFM4", "CD55", "POSTN", "RBP4", "POU5F1")
#genes <- c("CD72", "DOK5", "ENO1", "POU2F2", "SUPT6H", "DNAJC17")
for (g in genes) {
  if (!(g %in% rownames(y))) { next }
  p <- plot_gene(s, y, g, path_patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
}

#s <- s %>% filter(study == "ipmn")
#y <- y[, s$aoi]
genes <- c("CLU", "TNFAIP2", "MUC3A", "COL1A1", "MALL")
for (g in genes) {
  p <- plot_gene2(s, y, g, path_patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
  p <- plot_gene2(s, y, g, patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_by_patient_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
  p <- plot_gene2(s, y, g, patient_slide, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_by_patient_slide_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
}




```


# GSVA (ssGSEA)

```{r gsva analysis}


for (i in 1:nrow(res)) {
  gs_name = rownames(res)[i]
  x <- s %>% add_column(gs_score = unlist(res[i,]))
  p <- ggplot(x, aes(x=cluster, y=gs_score, fill=path)) +
    geom_boxplot(outlier.shape=NA, width=0.75) +
    geom_point(position=position_jitterdodge(jitter.width=0.1),
               aes(group=path), pch=21, size=2, alpha=0.8) +
    ggtitle(paste0("GSVA: ", gs_name)) +
    theme_bw() + 
    theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ study, scale="free_x")
  f <- file.path(plot_dir, paste0("gsva_boxplot_cluster_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)
  
  p <- ggplot(x, aes(x=patient, y=gs_score, fill=path)) +
    geom_boxplot(outlier.shape=NA, width=0.75) +
    geom_point(position=position_jitterdodge(jitter.width=0.1),
               aes(group=path), pch=21, size=2, alpha=0.8) +
    ggtitle(paste0("GSVA: ", gs_name)) +
    theme_bw() + 
    theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ study, scale="free_x")
  f <- file.path(plot_dir, paste0("gsva_boxplot_patient_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)

  p <- ggplot(x, aes(x=fct_reorder(aoi, gs_score), y=gs_score, fill=path)) +
    geom_col() +
    ggtitle(paste0("GSVA: ", gs_name)) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ patient, scale="free_x") +
    theme_bw() + 
    theme(axis.text.x = element_text(size=4, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    theme(strip.text.x = element_text(angle = 90),
          strip.clip = "off")
  f <- file.path(plot_dir, paste0("gsva_aoi_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)
}






x <- bind_cols(s, t(res))
p <- ggplot(x, aes(x=fct_reorder(aoi, `Moffitt.F6_BasalLike.top100`))) +
  geom_point(aes(y=`Moffitt.F6_BasalLike.top100`), fill="red", size=2, pch=21) +
  geom_point(aes(y=`Moffitt.F8_Classical.top100`), fill="cyan", size=2, pch=22) +
  geom_point(aes(y=`Moffitt.F3_Exocrine.top100`), fill="blue", size=2, pch=22) +
  theme_bw() +
  theme(axis.text.x = element_blank()) +
  facet_grid(~ path, scales="free_x")



p <- ggplot() +
    geom_errorbar(data=data1, aes(x=.data$x, ymin=.data$q25+1, ymax=.data$q75+1, color=.data$keep), alpha=0.75) +
    #geom_segment(data=data1, aes(x=x, xend=x, y=q25, yend=q75), color="grey", alpha=0.75) +
    geom_point(data=data1, aes(x=.data$x, y=.data$bg_lod+1), color="black", shape=4, alpha=0.75) +
    geom_point(data=data2, aes(x=.data$x, y=.data$value+1, fill=.data$quant), shape=21, alpha=0.75) +
    scale_color_manual(values=pals::cols25()) +
    scale_fill_viridis_d() +
    scale_y_log10() +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank()) +
    facet_grid(~ slide, scales="free_x") +
    labs(x = "AOI", y="Gene Expression")
  
  

  # quantiles
  quantiles <- c(0.25, 0.5, 0.75, 0.90, 0.95, 0.99)
  quant_names <- paste0("q", round(100 * quantiles))

  q <- as_tibble(x) %>%
    filter(!bg) %>%
    reframe(across(everything(), ~ stats::quantile(.x, quantiles))) %>%
    mutate(quant=quant_names) %>%
    tidyr::pivot_longer(colnames(x), names_to="aoi", values_to="value") %>%
    tidyr::pivot_wider(names_from="quant", values_from="value", names_prefix="") %>%
    inner_join(s, by=c("aoi"="aoi"), suffix=c("_gene", "_aoi"))

  data1 <- q %>%
    rowwise() %>%
    arrange(.data$keep, .data$q50) %>%
    mutate(x=factor(.data$aoi, .data$aoi)) %>%
    ungroup()

  data2 <- data1 %>%
    select("x", "slide", "keep", "q50", "q90", "q95", "q99") %>%
    tidyr::pivot_longer(cols = c("q50", "q90", "q95", "q99"),
                        names_to="quant", values_to="value")








p <- ggplot(x, aes(x=cluster, y=`Moffitt.F6_BasalLike.top100`, fill=path)) +
  geom_boxplot(outlier.shape=NA, width=0.75) + 
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=path), pch=21) +
  scale_fill_manual(values = color_scales$path) + 
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(axis.line = element_line(color = "black"))



p <- ggplot(x, aes(x=cluster, y=`Moffitt.F8_Classical.top100`, fill=path)) +
  geom_boxplot(outlier.shape=NA, width=0.75) + 
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=path), pch=21) +
  scale_fill_manual(values = color_scales$path) + 
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(axis.line = element_line(color = "black"))

p <- ggplot(x, aes(x=cluster, y=`Moffitt.F3_Exocrine.top100`, fill=path)) +
  geom_boxplot(outlier.shape=NA, width=0.75) + 
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=path), pch=21) +
  scale_fill_manual(values = color_scales$path) + 
  theme_bw() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(axis.line = element_line(color = "black"))








s <- fds$samples
s$path <- gsub("_epithelial", "", s$path)
s$path <- factor(s$path, levels=c("nl_nl", "pnet_nl", "ipmn_lgd_nl", 
                                  "ipmn_inv_nl", "pdac_nl", 
                                  "nl_adm", "nl_panin",
                                  "ipmn_lgd_lgd", "ipmn_inv_lgd", "pdac_adm",
                                  "pdac_panin", "ipmn_inv_hgd", "ipmn_inv_inv",
                                  "pdac_pdac"),
                 ordered=TRUE)

color_scale <- color_scales$path
names(color_scale) <- gsub("_epithelial", "", names(color_scale))


for (i in 1:nrow(res)) {
  gs_name = rownames(res)[i]
  x <- s %>% add_column(gs_score = unlist(res[i,]))
  p <- ggplot(x, aes(x=fct_reorder(patient, gs_score), y=gs_score, fill=path)) +
    geom_boxplot(outlier.shape=NA, width=0.75) +
    geom_point(position=position_jitterdodge(jitter.width=0.1),
               aes(group=path), pch=21, size=2, alpha=0.8) +
    ggtitle(paste0("GSVA: ", gs_name)) +
    theme_bw() + 
    theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ study, scale="free_x")
  f <- file.path(plot_dir, paste0("gsva_patient_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)
  
  p <- ggplot(x, aes(x=fct_reorder(aoi, gs_score), y=gs_score, fill=path)) +
    geom_col() +
    ggtitle(paste0("GSVA: ", gs_name)) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ patient, scale="free_x") +
    theme_bw() + 
    theme(axis.text.x = element_text(size=4, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    theme(strip.text.x = element_text(angle = 90),
          strip.clip = "off")
  f <- file.path(plot_dir, paste0("gsva_aoi_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)
}



# setup heatmap annotations
annot_col <- column_to_rownames(fds$samples, "aoi") %>% 
  select(study, patient, path_aoi, path_specimen) %>%
  as.data.frame()



# heatmap
f <- file.path(plot_dir,  paste0("heatmap_gsva_pdac_subtypes"))
h <- 6
w <- 10

p <- pheatmap(res,
              color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
              breaks = seq(-2, 2, length.out=51),
              border_color = NA,
              scale="row",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_col = annot_col,
              annotation_colors = color_scales,
              fontsize = 8,
              fontsize_row = 8,
              fontsize_col = 6,
              filename = paste0(f, ".png"),
              width = w,
              height = h)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


```





## GSVA / ssGSEA analysis

```{r gsva analysis}
library(GSVA)

# Appreciate and cite code and signatures obtained from here:
# https://github.com/PascaDiMagliano-Lab/Gift-of-Life-Public-Repository/blob/master/scripts/Spatial_Transcriptomics_Analysis/01_DGE_analysis.R

## Collison geneset (https://pubmed.ncbi.nlm.nih.gov/21460848/)
collison_gs <- list(
  Collison_Exocrine_Like = c('REG1B','REG3A','REG1A','PNLIPRP2','CEL','PNLIP','PLA2G1B','CELA3A','CPB1','CELA3B','CTRB2','CLPS','CELA2B','PRSS2','PRSS1','GP2','SLC3A1','CFTR','SLC4A4','SPINK1'),
  Collison_Basal = c('AIM2','FAM26F','GPM6B','S100A2','KRT14','CAV1','LOX','SLC2A3','TWIST1','PAPPA','NT5E','CKS2','HMMR','SLC5A3','PMAIP1','PHLDA1','SLC16A1','FERMT1','HK2','AHNAK2'),
  Collison_Classical = c('TMEM45B','SDR16C5','GPRC5A','AGR2','S100P','FXYD3','ST6GALNAC1','CEACAM5','CEACAM6','TFF1','TFF3','CAPN8','FOXQ1','ELF3','ERBB3','TSPAN8','TOX3','LGALS4','PLS1','GPX2','ATP10B','MUC13'))

## Moffit geneset (https://pubmed.ncbi.nlm.nih.gov/26343385/)
moffit_gs <- list(
  Moffit_Basal = c("VGLL1","UCA1","S100A2","LY6D","SPRR3","SPRR1B","LEMD1","KRT15","CTSL2","DHRS9","AREG","CST6","SERPINB3", "SERPINB4", "KRT6C", "KRT6A", "FAM83A", "SCEL", "FGFBP1", "KRT7", "KRT17", "GPR87", "TNS4", "SLC2A1", "ANXA8L2"),
  Moffit_Classical = c("BTNL8","FAM3D","ATAD4","AGR3","CTSE","LOC400573","LYZ","TFF2","TFF1","ANXA10","LGALS4","PLA2G10", "CEACAM6","VSIG2","TSPAN8","ST6GALNAC1","AGR2","TFF3","CYP3A7","MYO1A","CLRN3","KRT20","CDH17","SPINK4","REG4")
)

## Bailey geneset (https://www.nature.com/articles/nature16965)
bailey_gs <- list()
bailey_gs[['Bailey_Basal']] <- c('PGAM1','ANGPTL4','ARPC2','ITGA3','PTPN1','GSDMC','DRAP1','ANXA8','LDHA','ARHGAP23','HCAR2','ULBP2','KIAA1609','PTHLH','CEBPB','SLC16A3','EGFR','PLXNA1','ZBED2','IL2A','TPD5212','DUSP14','FSCN1','RND3','GPR87','PANX1','S100A2','ANXA1','KRT6A','ADAM17','PTRF','EMLIN1','FBXL7','CERCAM','DDR2','GXYLT2','CHSY3','VSTM4','COL8A1','PRRX1','COL5A2','SPARC','FBN1','FAP','COL6A3','COL1A2','BNC2','EFEMP2','FSTL1','COL3A1','GSG2','KIF18A','CDCA5','CCNA2','MCM10','DEPDC1','UBE2C','KIF4A','CCNB1','KIF2C','SPAG5','TPX2','HIST1H2B0','BUB1','CKAP2L','PTTG1','CDC20','ERCC6L','NCAPG','NCAPH','PSMA7','CCT6A','CCT7','DKC1','HSPE1','EIF2S2','MRPL11','MRPL17','LYAR','PSMA1','EIF4A3','PPAT','HAT1','PAICS','BRIX1','RUVBL1','CKS1B','TOMM40','CCT5','ALYREF')
bailey_gs[['Bailey_Classical']] <- c('LRRC66','GJB1','LGALS4','PLA2G10','PLS1','BCAS1','CAPN5','HNF4G','ARHGEF38','CAPN8','AP001187.9','ABP1','ERBB3','C9orf152','SLC44A4','IHH','HNF4A','FAM83E','PRR15L','EPS8L3')
bailey_gs[['Bailey_ADEX']] <- c('REG3G','SYCN','REG1P','SERPINI2','CPB1','RP11-331F4.4','AQP8','RBPJL','CUZD1','PLA2G1B','CLPS','CEL','CELA3B','CELA3A','CTRB1','CTRC','CTRB2','PNLIPRP1','CPA1','CPA2','UNC79','GCK','GJD2','SYT4','KCNK16','NOL4','SCGN','INS','CABP7','CHGB','BEX1','SVOP','MR7-3HG','ABCC8','HMGCLL1','SLC30A8','SST','CELF3','PCSK2','SCG3')
bailey_gs[['Bailey_Immunogenic']] <-  c('IGHV3-15','IGHV3-7','IGHV1-46','IGKV2-28','IGHV3-23','IGHV3-53','IGHV5-51','IGHA1','IGKV4-1','IGLV2-23','IGLV1-40','IGKV3-11','IGLL5','IGJ','IGLC3','IGKVLD-39','IGLV2-14','AC096579.7','IGKV3-20','IGKC','CSF1R','TYROBP','FCER1G','TLR8','C3AR1','CD86','WAS','SPI1','HAVCR2','SASH3','CYBB','MS4A4A','BTK','LAPTM5','PTPRC','MARCH1','CD53','AIF1','DOCK2','NCKAP1L','SIT1','GIMAP4','GPR174','SEPT1','CXCR6','ITK','TBC1D10C','GIMAP5','CD96','ZNF831','GZMK','PTPRCAP','SLAMF1','P2RY10','CD48','THEMIS','CD3E','CD3D','SH2D1A','CD2')

## purist genes (genes that are used to compute purist score)
purist_gs <- list(
  PurIST_basal = c("GPR87", "KRT6A", "KRT17", "KRT15", "DCBLD2", "BCAR3", "PTGES", "ITGA3", "C16orf74", "S100A2", "KRT5"),
  PurIST_classical = c("REG4", "ANXA10", "TFF1", "CDH17", "TSPAN8", "GATA6", "CLDN18", "LGALS4", "DDC", "SLC40A1", "CLRN3", "PLA2G10")
)


# filter genes measured by geomx platform
subtypes_gs <- c(collison_gs, bailey_gs, moffit_gs)
sapply(subtypes_gs, length)
for (gs_name in names(subtypes_gs)) {
  subtypes_gs[[gs_name]] <- intersect(subtypes_gs[[gs_name]], fds$meta$gene)
}
sapply(subtypes_gs, length)

# GSVA
y <- log2_norm_cpm
res <- gsva(y, subtypes_gs, method="gsva")

s <- fds$samples
s$path <- gsub("_epithelial", "", s$path)
s$path <- factor(s$path, levels=c("nl_nl", "pnet_nl", "ipmn_lgd_nl", 
                                  "ipmn_inv_nl", "pdac_nl", 
                                  "nl_adm", "nl_panin",
                                  "ipmn_lgd_lgd", "ipmn_inv_lgd", "pdac_adm",
                                  "pdac_panin", "ipmn_inv_hgd", "ipmn_inv_inv",
                                  "pdac_pdac"),
                 ordered=TRUE)

color_scale <- color_scales$path
names(color_scale) <- gsub("_epithelial", "", names(color_scale))


for (i in 1:nrow(res)) {
  gs_name = rownames(res)[i]
  x <- s %>% add_column(gs_score = unlist(res[i,]))
  p <- ggplot(x, aes(x=fct_reorder(patient, gs_score), y=gs_score, fill=path)) +
    geom_boxplot(outlier.shape=NA, width=0.75) +
    geom_point(position=position_jitterdodge(jitter.width=0.1),
               aes(group=path), pch=21, size=2, alpha=0.8) +
    ggtitle(paste0("GSVA: ", gs_name)) +
    theme_bw() + 
    theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ study, scale="free_x")
  f <- file.path(plot_dir, paste0("gsva_patient_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)
  
  p <- ggplot(x, aes(x=fct_reorder(aoi, gs_score), y=gs_score, fill=path)) +
    geom_col() +
    ggtitle(paste0("GSVA: ", gs_name)) +
    scale_fill_manual(values = color_scale) + 
    facet_grid(~ patient, scale="free_x") +
    theme_bw() + 
    theme(axis.text.x = element_text(size=4, color="black", angle=90, vjust=0.5, hjust=1),
          axis.line = element_line(color = "black")) +
    theme(strip.text.x = element_text(angle = 90),
          strip.clip = "off")
  f <- file.path(plot_dir, paste0("gsva_aoi_", gs_name))
  ggsave(paste0(f, ".pdf"), p, width=12, height=6)
}



# setup heatmap annotations
annot_col <- column_to_rownames(fds$samples, "aoi") %>% 
  select(study, patient, path_aoi, path_specimen) %>%
  as.data.frame()



# heatmap
f <- file.path(plot_dir,  paste0("heatmap_gsva_pdac_subtypes"))
h <- 6
w <- 10

p <- pheatmap(res,
              color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
              breaks = seq(-2, 2, length.out=51),
              border_color = NA,
              scale="row",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_col = annot_col,
              annotation_colors = color_scales,
              fontsize = 8,
              fontsize_row = 8,
              fontsize_col = 6,
              filename = paste0(f, ".png"),
              width = w,
              height = h)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


```

## PDAC purist score

```{r purist score}

calc_purist <- function(x, method="purist") {
  
  purist_baseline <- -6.815
  purist_tbl <- tribble(
    ~a, ~b, ~w,
    "GPR87", "REG4", 1.994,
    "KRT6A", "ANXA10", 2.031,
    "BCAR3", "GATA6", 1.618,
    "PTGES", "CLDN18", 0.922,
    "ITGA3", "LGALS4", 1.059,
    "C16orf74", "DDC", 0.929,
    "S100A2", "SLC40A1", 2.505,
    "KRT5", "CLRN3", 0.485
  )
  
  puristn_baseline <- -12.414
  puristn_tbl <- tribble(
    ~a, ~b, ~w,
    "GPR87", "REG4", 3.413,
    "KRT6A", "ANXA10", 3.437,
    "KRT17", "LGALS4", 2.078,
    "S100A2", "TFF1", 2.651,
    "C16orf74", "DDC", 0.901,
    "KRT15", "PLA2G10", 2.677,
    "PTGES", "CDH17", 2.911,
    "DCBLD2", "TSPAN8", 1.903
  )

  if (method == "purist") {
    tbl <- purist_tbl
    baseline <- purist_baseline
  } else {
    tbl <- puristn_tbl
    baseline <- puristn_baseline
  }
  
  calc_score <- function(x) {
    sum(tbl$w * as.integer(x[tbl$a] > x[tbl$b])) + baseline
  }

  scores <- apply(x, 2, calc_score)
  probs <- exp(scores) / (1 + exp(scores))
  subtypes <- if_else(scores > 0, "basal-like", "classical")
  as_tibble(bind_cols(sample=colnames(x),
                      score=scores, 
                      prob=probs, 
                      subtype=subtypes))
}

s <- fds$samples
s$path <- gsub("_epithelial", "", s$path)
s$path <- factor(s$path, levels=c("nl_nl", "pnet_nl", "ipmn_lgd_nl", 
                                  "ipmn_inv_nl", "pdac_nl", 
                                  "nl_adm", "nl_panin",
                                  "ipmn_lgd_lgd", "ipmn_inv_lgd", "pdac_adm",
                                  "pdac_panin", "ipmn_inv_hgd", "ipmn_inv_inv",
                                  "pdac_pdac"),
                 ordered=TRUE)

color_scale <- color_scales$path
names(color_scale) <- gsub("_epithelial", "", names(color_scale))

y <- log2_norm_cpm
purist <- calc_purist(log2_norm_cpm, method="puristn")
x <- bind_cols(s, select(purist, purist_score=score, purist_prob=prob, purist_subtype=subtype))

p <- ggplot(x, aes(x=fct_reorder(patient, purist_prob), y=purist_prob, fill=path)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=path), pch=21, size=2, alpha=0.8) +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black")) +
  scale_fill_manual(values = color_scale) + 
  facet_grid(~ study, scale="free_x") +
  ggtitle("pdac purist score")
f <- file.path(plot_dir, paste0("purist_patient"))
ggsave(paste0(f, ".pdf"), p, width=12, height=6)

p <- ggplot(x, aes(x=fct_reorder(aoi, purist_prob), y=purist_prob, fill=path)) +
  geom_col() +
  scale_fill_manual(values = color_scale) + 
  facet_grid(~ patient, scale="free_x") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=4, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black")) +
  theme(strip.text.x = element_text(angle = 90),
        strip.clip = "off") +
  ggtitle("pdac purist score")
f <- file.path(plot_dir, paste0("purist_aoi"))
ggsave(paste0(f, ".pdf"), p, width=12, height=6)



```


## Plot individual genes

```{r plot individual genes}

y <- log2_norm_cpm
s <- fds$samples
s$patient <- factor(s$patient)
s$patient_slide <- factor(paste0(s$patient, "_", s$slide))
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "adm", "panin", "lgd", "hgd", "inv", "pdac"))
s$path_specimen <- factor(s$path_specimen, levels=c("nl", "ipmn_lgd", "ipmn_inv", "pnet", "pdac"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, 
                         levels=c("nl_nl", "ipmn_lgd_nl", "ipmn_inv_nl", 
                                  "pnet_nl", "pdac_nl", "nl_adm", "pdac_adm", 
                                  "nl_panin", "pdac_panin", 
                                  "ipmn_lgd_lgd", "ipmn_inv_lgd", 
                                  "ipmn_inv_hgd", "ipmn_inv_inv", 
                                  "pdac_pdac"))

width = 8
height = 5
genes <- c("MUC5AC", "S100P", "SYTL2", "LRRTM4", "ST6GALNAC2", "TMEM67",
           "EIF4A3", "CLDN18", "MSMB", 
           "KRT17", "LAMB3", "LAMC2", "PNLIP", "AMY1A", "CLU",
           "TNFAIP2", "MMP7", "MALL", "MUC6", "CEACAM5", "CEACAM1",
           "PGC", "GALNT6", "CELA3A", "BPIFB1", "EPPK1", "PITX1", "LIF", 
           "CTRB1", "NKX6-2", "OLFM4", "CD55", "POSTN", "RBP4", "POU5F1")
#genes <- c("CD72", "DOK5", "ENO1", "POU2F2", "SUPT6H", "DNAJC17")
for (g in genes) {
  if (!(g %in% rownames(y))) { next }
  p <- plot_gene(s, y, g, path_patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
}

#s <- s %>% filter(study == "ipmn")
#y <- y[, s$aoi]
genes <- c("CLU", "TNFAIP2", "MUC3A", "COL1A1", "MALL")
for (g in genes) {
  p <- plot_gene2(s, y, g, path_patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
  p <- plot_gene2(s, y, g, patient, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_by_patient_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
  p <- plot_gene2(s, y, g, patient_slide, path_patient) + 
    scale_fill_manual(values=color_scales$path_patient)
  f <- file.path(plot_dir, paste0("boxplot_gene_by_patient_slide_", g, ".pdf"))
  ggsave(f, p, width=width, height=height)
}


```

# old LE analysis

```{r le analysis old}

padj_cutoff <- 0.01
pathways <- c(
  "Collisson.Exocrine",
  "ICGC.ADEX.Up",
  "Moffitt.F3_Exocrine.top250"
)

x <- filter(gsea_pdacr, analysis=="clusterHRvLR", pathway %in% pathways)
m1 <- gsea_leading_edge_matrix(x, my_padj_cutoff=padj_cutoff, min_gene_sets=2)
colnames(m1) <- paste0("HRvLR_",colnames(m1))

x <- filter(gsea_pdacr, analysis=="clusterHRvNL", pathway %in% pathways)
m2 <- gsea_leading_edge_matrix(x, my_padj_cutoff=padj_cutoff, min_gene_sets=2)
colnames(m2) <- paste0("HRvNL_",colnames(m2))

# combine matrices
# select genes present in both analysis
genes <- intersect(rownames(m1), rownames(m2))
m1 <- m1[genes,]
m2 <- m2[genes,]
m <- cbind(m1, m2)

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 5
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_exocrine.pdf"), width=w, height=h)

#
# basal-like genes
#
padj_cutoff <- 0.05
pathways <- c("Moffitt.F6_BasalLike.top250",
              "Collisson.QM",
              "ICGC.Squamous.Up",
              "Chan_Seng_Yue.BasalA",
              "Chan_Seng_Yue.BasalB")

x <- filter(gsea_pdacr, analysis=="clusterHRvNL", pathway %in% pathways)
m1 <- gsea_leading_edge_matrix(x, my_padj_cutoff=padj_cutoff, min_gene_sets=2)
colnames(m1) <- paste0("HRvNL_",colnames(m1))

x <- filter(gsea_pdacr, analysis=="clusterHRvLR", pathway %in% pathways)
m2 <- gsea_leading_edge_matrix(x, my_padj_cutoff=padj_cutoff, min_gene_sets=2)
colnames(m2) <- paste0("HRvLR_",colnames(m2))

# combine matrices
# select genes present in both analysis
genes <- intersect(rownames(m1), rownames(m2))
m1 <- m1[genes,]
m2 <- m2[genes,]
m <- cbind(m1, m2)

# show intersection with CPTAC
length(intersect(genes, gs_cptac$CPTAC_RNA_ALL_UP))
length(intersect(genes, gs_cptac$CPTAC_PROT_ALL_UP))
length(intersect(genes, gs_hpa_panc$HPA_PANCREATIC_CANCER_UNFAV_PROGNOSIS))
# agreement across studies
table(rowSums(sign(m)))

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 8
w <- 6
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_basal.pdf"), width=w, height=h)


x <- filter(gsea_pdacr, analysis=="clusterHRvNL", pathway %in% pathways)
m1 <- gsea_leading_edge_matrix(x, my_padj_cutoff=padj_cutoff, min_gene_sets=2)
colnames(m1) <- paste0("HRvNL_",colnames(m1))

x <- filter(gsea_pdacr, analysis=="clusterLRvNL", pathway %in% pathways)
m2 <- gsea_leading_edge_matrix(x, my_padj_cutoff=padj_cutoff, min_gene_sets=2)
colnames(m2) <- paste0("LRvNL_",colnames(m2))

# combine matrices
# select genes present in both analysis
genes <- intersect(rownames(m1), rownames(m2))
m1 <- m1[genes,]
m2 <- m2[genes,]
m <- cbind(m1, m2)

```


# extra GSEA plots

```{r gsea plots}

gmt_to_tbl <- function(gmt) {
  gs_to_tbl <- function(gs_name, gs_genes) {
    bind_cols(gs_name = rep(gs_name, length(gs_genes)),
              gene_symbol = gs_genes)
  }
  tbl <- NULL
  for (i in 1:length(gmt)) {
    tbl <- bind_rows(tbl, gs_to_tbl(names(gmt)[i], gmt[[i]]))
  }
  return(tbl)
}

plot_gsea_volcano <- function(x, padj_cutoff = 0.01, max.overlaps = Inf) {
  p <- ggplot(x, aes(x=NES, y=-log10(padj), color=analysis)) + 
    geom_point(data=subset(x, padj > padj_cutoff), size=1, alpha=0.4) + 
    geom_point(data=subset(x, padj <= padj_cutoff), size=2, alpha=0.8) +
    geom_text_repel(data=subset(x, padj <= padj_cutoff), color="black", size=3, aes(label=pathway), max.overlaps=max.overlaps) +
    ylab("-log10(adjusted p-value)") +
    xlab("NES") +
    theme_minimal() +
    theme(legend.position="bottom") + 
    theme(axis.line = element_line(color = "black")) +
    coord_flip() 
  return(p)
}

plot_gsea_barplot <- function(x, padj_cutoff = 0.01) {
  x <- mutate(x, sig = ifelse(padj < padj_cutoff, ifelse(NES < 0, "dn", "up"), "no"),
              neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA))
  p <- ggplot(x, aes(x=reorder(pathway, NES), y=NES, fill=neglog10padj)) +
    geom_col() +
    scale_fill_gradient(low = "blue", high = "cyan", na.value="#aaaaaa") +
    coord_flip() +
    labs(x="Analysis", y="Normalized Enrichment Score") +
    theme_minimal()
  return(p)
}



#
# pdacr plots
#
unique(gsea_pdacr$analysis)
padj_cutoff <- 0.01
plotanalyses <- unique(gsea_pdac$analysis)

donor_analyses <- c("pdacpanin_vs_pdacnl", "pdacpdac_vs_pdacpanin",
                   "pdacpdac_vs_pdacnl", "nlpanin_vs_nlnl",  "nladm_vs_nlnl", 
                   "nlpanin_vs_nladm", "pdacpanin_vs_nlpanin", "pdacnl_vs_nlnl")
ipmn_analyses <- c("invinv_vs_invnl", "invhgd_vs_invnl", "invlgd_vs_invnl", 
                  "invinv_vs_invlgd", "invhgd_vs_invlgd", "invinv_vs_invhgd", 
                  "lgdlgd_vs_lgdnl", "invinv_vs_lgdnl", "invhgd_vs_lgdnl", 
                  "invlgd_vs_lgdnl", "invinv_vs_lgdlgd", "invhgd_vs_lgdlgd", 
                  "invlgd_vs_lgdlgd", "invnl_vs_lgdnl")

x <- filter(gsea_pdac, analysis %in% plotanalyses)
x <- x %>% mutate( 
  sig = ifelse(padj < padj_cutoff, ifelse(NES < 0, "dn", "up"), "no"),
  neglog10padj = -log10(x$padj),
  score = NES * -log10(x$padj),
  fillneglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  fillnes = ifelse(padj < padj_cutoff, NES, NA)
)

x <- x %>% mutate(
  pathway = factor(pathway),
  analysis = fct_reorder(analysis, NES, .fun = function(x) { max(abs(x)) }),
  facet = case_when(analysis %in% ipmn_analyses ~ "ipmn",
                    analysis %in% donor_analyses ~ "donor_panc")
)

p <- ggplot(x, aes(x=analysis, y=NES, fill=fillneglog10padj)) +
  geom_bar(stat="identity", position=position_dodge()) +
  scale_fill_viridis_c() +
  labs(x="Analysis", y="Normalized Enrichment Score",
       title="GSEA PDAC DE genes") + 
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1, size=6),
        strip.clip = "off"
  ) +
  facet_grid(pathway ~ facet, scales = "free", space="free")
p
ggsave(file.path(plot_dir, "gsea_pdac_barplots.pdf"), p, width=8, height=10)



#
# GSEA hallmark bar plots
#
table(gsea_hallmark$analysis)
x <- gsea_hallmark
x$pathway <- gsub("HALLMARK_", "", x$pathway)

padj_cutoff <- 0.01
w <- 6
h <- 6

for (a in unique(x$analysis)) {
  n <- nrow(filter(x, analysis == a, padj < padj_cutoff))
  if (n == 0) { next }
  p <- plot_gsea_barplot(filter(x, analysis == a), padj_cutoff = padj_cutoff) +
    theme(axis.text.y = element_text(size=6))
  f <- file.path(plot_dir, paste0("gsea_hallmark_", a, ".pdf"))
  ggsave(f, plot=p, width=w, height=h)
}


#
# GSEA GOBP
#
table(gsea_gobp$analysis)
x <- gsea_gobp
x$pathway <- gsub("GOBP_", "", x$pathway)

padj_cutoff <- 0.01
ntop <- 25
w <- 10
h <- 6

# my_ggp + theme(axis.text.x = element_text(size = 20))

for (a in unique(x$analysis)) {
  res_up <- filter(x, analysis == a) %>%
    slice_max(NES, n=ntop)
  res_dn <- filter(x, analysis == a) %>%
    slice_min(NES, n=ntop)
  res <- bind_rows(res_up, res_dn) %>% distinct(pathway, .keep_all=TRUE)
  n <- nrow(filter(res, padj < padj_cutoff))
  if (n == 0) { next }
  p <- plot_gsea_barplot(res, padj_cutoff = padj_cutoff) + 
    theme(axis.text.y = element_text(size=6))
  f <- file.path(plot_dir, paste0("gsea_gobp_", a, ".pdf"))
  ggsave(f, plot=p, width=w, height=h)
}


#
# cptac plots comparing analysis
#
unique(gsea_cptac$analysis)
padj_cutoff <- 0.01
plotanalyses <- unique(gsea_cptac$analysis)

donor_analyses <- c("pdacpanin_vs_pdacnl", "pdacpdac_vs_pdacpanin",
                   "pdacpdac_vs_pdacnl", "nlpanin_vs_nlnl",  "nladm_vs_nlnl", 
                   "nlpanin_vs_nladm", "pdacpanin_vs_nlpanin", "pdacnl_vs_nlnl")
ipmn_analyses <- c("invinv_vs_invnl", "invhgd_vs_invnl", "invlgd_vs_invnl", 
                  "invinv_vs_invlgd", "invhgd_vs_invlgd", "invinv_vs_invhgd", 
                  "lgdlgd_vs_lgdnl", "invinv_vs_lgdnl", "invhgd_vs_lgdnl", 
                  "invlgd_vs_lgdnl", "invinv_vs_lgdlgd", "invhgd_vs_lgdlgd", 
                  "invlgd_vs_lgdlgd", "invnl_vs_lgdnl")

x <- filter(gsea_cptac, analysis %in% plotanalyses)
x <- x %>% mutate( 
  sig = ifelse(padj < padj_cutoff, ifelse(NES < 0, "dn", "up"), "no"),
  neglog10padj = -log10(x$padj),
  score = NES * -log10(x$padj),
  fillneglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  fillnes = ifelse(padj < padj_cutoff, NES, NA)
)

x <- x %>% mutate(
  pathway = factor(gsub("CPTAC_RNA_ALL_", "", pathway), levels=c("DN", "UP")), 
  analysis = fct_reorder(analysis, NES, .fun = function(x) { max(abs(x)) }),
  facet = case_when(analysis %in% ipmn_analyses ~ "ipmn",
                    analysis %in% donor_analyses ~ "donor_panc")
)

p <- ggplot(x, aes(x=analysis, y=NES, fill=fillneglog10padj)) +
  geom_bar(stat="identity", position=position_dodge()) +
  scale_fill_viridis_c() +
  labs(x="Analysis", y="Normalized Enrichment Score",
       title="PDAC DE Genes (CPTAC RNA-Seq)") + 
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1, size=6),
        strip.clip = "off"
  ) +
  facet_grid(pathway ~ facet, scales = "free", space="free")
p
ggsave(file.path(plot_dir, "gsea_cptac_barplots.pdf"), p, width=12, height=5)


#
# pdac plots
#
unique(gsea_pdac$analysis)
padj_cutoff <- 0.01
plotanalyses <- unique(gsea_pdac$analysis)

donor_analyses <- c("pdacpanin_vs_pdacnl", "pdacpdac_vs_pdacpanin",
                   "pdacpdac_vs_pdacnl", "nlpanin_vs_nlnl",  "nladm_vs_nlnl", 
                   "nlpanin_vs_nladm", "pdacpanin_vs_nlpanin", "pdacnl_vs_nlnl")
ipmn_analyses <- c("invinv_vs_invnl", "invhgd_vs_invnl", "invlgd_vs_invnl", 
                  "invinv_vs_invlgd", "invhgd_vs_invlgd", "invinv_vs_invhgd", 
                  "lgdlgd_vs_lgdnl", "invinv_vs_lgdnl", "invhgd_vs_lgdnl", 
                  "invlgd_vs_lgdnl", "invinv_vs_lgdlgd", "invhgd_vs_lgdlgd", 
                  "invlgd_vs_lgdlgd", "invnl_vs_lgdnl")

x <- filter(gsea_pdac, analysis %in% plotanalyses)
x <- x %>% mutate( 
  sig = ifelse(padj < padj_cutoff, ifelse(NES < 0, "dn", "up"), "no"),
  neglog10padj = -log10(x$padj),
  score = NES * -log10(x$padj),
  fillneglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  fillnes = ifelse(padj < padj_cutoff, NES, NA)
)

x <- x %>% mutate(
  pathway = factor(pathway),
  analysis = fct_reorder(analysis, NES, .fun = function(x) { max(abs(x)) }),
  facet = case_when(analysis %in% ipmn_analyses ~ "ipmn",
                    analysis %in% donor_analyses ~ "donor_panc")
)

p <- ggplot(x, aes(x=analysis, y=NES, fill=fillneglog10padj)) +
  geom_bar(stat="identity", position=position_dodge()) +
  scale_fill_viridis_c() +
  labs(x="Analysis", y="Normalized Enrichment Score",
       title="GSEA PDAC DE genes") + 
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1, size=6),
        strip.clip = "off"
  ) +
  facet_grid(pathway ~ facet, scales = "free", space="free")
p
ggsave(file.path(plot_dir, "gsea_pdac_barplots.pdf"), p, width=8, height=10)


```


# gsea clusters scatter plots

```{r extra code}



# categorize further
x <- gsea_pdacr %>% mutate(
  study = case_when(startsWith(pathway, "Moffitt") ~ "moffitt",
                    startsWith(pathway, "Chan_Seng_Yue") ~ "chan_seng_yue",
                    startsWith(pathway, "Collisson") ~ "collison",
                    startsWith(pathway, "ICGC") ~ "icgc",
                    TRUE ~ "other")
)


gsea <- x %>%
  filter(analysis %in% contrasts, study != "other")
gsea <- gsea %>% mutate(
  neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  sig = padj < padj_cutoff)

# only keep significant results (at least one analysis)
gsea <- gsea %>% group_by(pathway) %>%
  filter(any(padj < padj_cutoff)) %>%
  ungroup()

# arrange by clusterHR for readability
gsea$pathway <- factor(gsea$pathway, levels=gsea %>% filter(analysis == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway))

p <- ggplot(gsea, aes(x=pathway, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color="black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(axis.text.y = element_text(color="black", size=7)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(study ~ analysis) + 
  labs(x="Pathway", y="Normalized Enrichment Score (NES)")


f <- paste0("gsea_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=5)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=5)



# Hallmark
contrasts <- c("cluster1v2", "cluster1v3", "cluster2v3")
padj_cutoff <- 0.01

gsea <- gsea_hallmark %>%
  filter(analysis %in% contrasts)
gsea <- mutate(gsea, pathway = gsub("HALLMARK_", "", pathway))
gsea <- gsea %>% mutate(
  neglog10padj = ifelse(padj < padj_cutoff, -log10(padj), NA),
  sig = padj < padj_cutoff)

# merge de analysis by gene
gsea <- gsea %>% select(pathway, analysis, padj, neglog10padj, sig, ES, NES) %>%
  pivot_wider(names_from = analysis, values_from = c(padj, neglog10padj, sig, ES, NES))

#
# scatter plot: c1 vs c2 (x) and c2 vs c3 (y)
#
label_data <- gsea %>% 
  filter(!is.na(neglog10padj_cluster1v3) | !is.na(neglog10padj_cluster1v2))

p <- ggplot(gsea, aes(x=NES_cluster1v2, y=NES_cluster1v3, 
                      color=neglog10padj_cluster1v3, 
                      fill=neglog10padj_cluster1v2)) +
  geom_point(data=filter(gsea, sig_cluster1v3 | sig_cluster1v2), pch=21, size=3) + 
  geom_point(data=filter(gsea, !(sig_cluster1v3 | sig_cluster1v2)), pch=21, size=1) +
  geom_text_repel(data=label_data, aes(label=pathway), color="black", size=3, max.overlaps=Inf) +
  scale_fill_viridis_c(limits=c(2,4), oob=scales::oob_squish) +
  scale_color_viridis_c(limits=c(2,4), oob=scales::oob_squish) +
  geom_vline(xintercept=0) +
  geom_hline(yintercept=0) + 
  xlim(-2.5, 2.5) + 
  ylim(-2.5, 2.5)
f <- paste0("gsea_scatter_cluster1_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=10, height=8)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=10, height=8)



#
# scatter plot: c1 vs c2 (x) and c2 vs c3 (y)
#
label_data <- gsea %>% 
  filter(!is.na(neglog10padj_cluster1v2) | !is.na(neglog10padj_cluster2v3))

p <- ggplot(gsea, aes(x=NES_cluster1v2, y=NES_cluster2v3, 
                      color=neglog10padj_cluster2v3, 
                      fill=neglog10padj_cluster1v2)) +
  geom_point(data=filter(gsea, sig_cluster2v3 | sig_cluster1v2), pch=21, size=3, stroke=2) + 
  geom_point(data=filter(gsea, !(sig_cluster2v3 | sig_cluster1v2)), pch=21, size=1, stroke=1) +
  geom_text_repel(data=label_data, aes(label=pathway), color="black", size=3) +
  scale_fill_viridis_c(limits=c(2,4), oob=scales::oob_squish) +
  scale_color_viridis_c(limits=c(2,4), oob=scales::oob_squish) +
  geom_vline(xintercept=0) +
  geom_hline(yintercept=0) + 
  xlim(-2.5, 2.5) + 
  ylim(-2.5, 2.5)
f <- paste0("gsea_scatter_cluster2_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=4)

```


# PDAC Purist score


```{r purist score}

calc_purist <- function(x, method="purist") {
  
  purist_baseline <- -6.815
  purist_tbl <- tribble(
    ~a, ~b, ~w,
    "GPR87", "REG4", 1.994,
    "KRT6A", "ANXA10", 2.031,
    "BCAR3", "GATA6", 1.618,
    "PTGES", "CLDN18", 0.922,
    "ITGA3", "LGALS4", 1.059,
    "C16orf74", "DDC", 0.929,
    "S100A2", "SLC40A1", 2.505,
    "KRT5", "CLRN3", 0.485
  )
  
  puristn_baseline <- -12.414
  puristn_tbl <- tribble(
    ~a, ~b, ~w,
    "GPR87", "REG4", 3.413,
    "KRT6A", "ANXA10", 3.437,
    "KRT17", "LGALS4", 2.078,
    "S100A2", "TFF1", 2.651,
    "C16orf74", "DDC", 0.901,
    "KRT15", "PLA2G10", 2.677,
    "PTGES", "CDH17", 2.911,
    "DCBLD2", "TSPAN8", 1.903
  )

  if (method == "purist") {
    tbl <- purist_tbl
    baseline <- purist_baseline
  } else {
    tbl <- puristn_tbl
    baseline <- puristn_baseline
  }
  
  calc_score <- function(x) {
    sum(tbl$w * as.integer(x[tbl$a] > x[tbl$b])) + baseline
  }

  scores <- apply(x, 2, calc_score)
  probs <- exp(scores) / (1 + exp(scores))
  subtypes <- if_else(scores > 0, "basal-like", "classical")
  as_tibble(bind_cols(sample=colnames(x),
                      score=scores, 
                      prob=probs, 
                      subtype=subtypes))
}


s <- fds$samples
y <- log2_norm_cpm
purist <- calc_purist(y, method="purist")
puristn <- calc_purist(y, method="puristn")

x <- bind_cols(
  s, 
  select(purist, purist_score=score, purist_prob=prob, purist_subtype=subtype),
  select(puristn, puristn_score=score, puristn_prob=prob, puristn_subtype=subtype)
)

p <- ggplot(x, aes(x=fct_reorder(patient, purist_prob), y=purist_prob, fill=path_aoi)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=path_aoi), pch=21, size=2, alpha=0.8) +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black")) +
  scale_fill_manual(values = color_scales$path_aoi)



s <- fds$samples
s$path <- gsub("_epithelial", "", s$path)
s$path <- factor(s$path, levels=c("nl_nl", "pnet_nl", "ipmn_lgd_nl", 
                                  "ipmn_inv_nl", "pdac_nl", 
                                  "nl_adm", "nl_panin",
                                  "ipmn_lgd_lgd", "ipmn_inv_lgd", "pdac_adm",
                                  "pdac_panin", "ipmn_inv_hgd", "ipmn_inv_inv",
                                  "pdac_pdac"),
                 ordered=TRUE)

color_scale <- color_scales$path
names(color_scale) <- gsub("_epithelial", "", names(color_scale))

y <- log2_norm_cpm
purist <- calc_purist(log2_norm_cpm, method="puristn")
x <- bind_cols(s, select(purist, purist_score=score, purist_prob=prob, purist_subtype=subtype))

p <- ggplot(x, aes(x=fct_reorder(patient, purist_prob), y=purist_prob, fill=path)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=path), pch=21, size=2, alpha=0.8) +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black")) +
  scale_fill_manual(values = color_scale) + 
  facet_grid(~ study, scale="free_x") +
  ggtitle("pdac purist score")
f <- file.path(plot_dir, paste0("purist_patient"))
ggsave(paste0(f, ".pdf"), p, width=12, height=6)

p <- ggplot(x, aes(x=fct_reorder(aoi, purist_prob), y=purist_prob, fill=path)) +
  geom_col() +
  scale_fill_manual(values = color_scale) + 
  facet_grid(~ patient, scale="free_x") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=4, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black")) +
  theme(strip.text.x = element_text(angle = 90),
        strip.clip = "off") +
  ggtitle("pdac purist score")
f <- file.path(plot_dir, paste0("purist_aoi"))
ggsave(paste0(f, ".pdf"), p, width=12, height=6)



```


