---
title: "Spatial Transcriptomics of IPMN Reveals Divergent Indolent and Malignant Progenitor Phenotypes"
author: Matthew K. Iyer
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{pja_ipmn_wta}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Setup

## Library imports

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(tidyverse)
library(readxl)
library(writexl)
library(patchwork)
library(ggrepel)
library(ggridges)
library(pROC)
library(Matrix)
library(class)

# bioinformatics packages
library(pheatmap)

# bioconductor packages
library(factoextra)
library(limma)
library(fgsea)
library(GSVA)
library(GeomxTools)
library(slingshot)
library(bluster)

```


## Input files and parameter settings

```{r input files, include=FALSE}

# set to location of manuscript files
working_dir <- "/Users/mkiyer/University of Michigan Dropbox/Matthew Iyer/manuscripts/duke_stgeomx_ipmn_wta/resubmission/analysis"

# input directories
geomx_dir <- file.path(working_dir, "geomx")
data_dir <- file.path(working_dir, "external_data")

# WTA PKC file
wta_pkc_file <- file.path(geomx_dir, "Hs_R_NGS_WTA_v1.0.pkc")

# ipmn dataset and annotations
pja_ipmn_geomx_xlsx <- file.path(geomx_dir, "pja_ipmn_wta_resubmission_2024-08-05.xlsx")
pja_ipmn_geomx_annot_xlsx <- file.path(geomx_dir, "pja_ipmn_wta_annot_resubmission_2024-08-08.xlsx")

# Carbone IPMN WTA datasets
cc_ipmn_geomx_dcc_tar <- file.path(geomx_dir, "GSE229752_RAW.tar")
cc_ipmn_geomx_annot_xlsx <- file.path(geomx_dir, "GSE229752_carbone_ipmn_annot_2024-06-30.xlsx")

# Spatial Organ Atlas dataset
soa_panc_geomx_xlsx <- file.path(geomx_dir, "soa_panc_2022-02-21.xlsx")
soa_panc_geomx_annot_xlsx <- file.path(geomx_dir, "soa_panc_annot_2024-04-24.xlsx")

# GSE226829 (donor pancreata)
umich_gol_pub_dcc_tar <- file.path(geomx_dir, "GSE226829_RAW.tar")
umich_gol_pub_annot_xlsx <- file.path(geomx_dir, "umich_gol_pub_2024-04-24.xlsx")


# external resource files
hpa_path_file <- file.path(data_dir, "hpa_pathology.tsv")
cptac_panc_file <- file.path(data_dir, "cptac_panc_supp_mmc3.xlsx")
pdacr_gene_list_file <- file.path(data_dir, "pdacR_gene_lists.rds")

# qc parameters
geomx_negprobe_lod_quantile = 0.9
geomx_min_counts = 30000
geomx_min_auc = 0.60
geomx_aoi_min_frac_expr = 0.2
geomx_gene_min_frac_expr = 0.15

# normalization
bgcorrect_method = "qq"
bgcorrect_bw_adjust = 2
bgcorrect_bg_quant = 0.5
norm_method = "qn"

# clustering parameters
# hc_func <- "hclust"
# hc_method <- "average"

# de parameters
de_padj_cutoff <- 0.05
de_log2fc_cutoff <- 1

# gsea params
gsea_nperms <- 10000
gsea_padj_cutoff <- 0.01
gsea_log2fc_cutoff <- 1

# external dataset parameters
cptac_log2fc_cutoff <- 1
cptac_padj_cutoff <- 0.05

# highly variable gene params
loess_span <- 0.5

# results
results_dir <- file.path(working_dir, "results")
plot_dir <- file.path(results_dir, "plots")

for (d in c(results_dir, plot_dir)) {
  if (!dir.exists(d)) {
    dir.create(d)
  }
}

# tmp dir for extracting dcc files
dcc_dir <- file.path(working_dir, "dcc")

```

## Plot scales (color, shape)

```{r color scales, include=FALSE}

color_scales = list(
  study = c(duke="#aa0000", soa="#008800", cc="#6600ff", gol="#ffee00"),
  segment = c(panck="#0000ff", cd3="#ff8800", cd45="#ff0000", otherstroma="#880088"),
  traj_clust = c(c1="#ff0000", c2="#ff8800", c3="#63c5da", c4="#710193"),
  traj_clust_name = c(cPDAC="#ff0000", cTR="#ff8800", cLR="#63c5da", cNL="#710193"),
  cluster = c(c1="#fc6a03", c2="#63c5da", c3="#710193"),
  cluster_name = c(cHR="#fc6a03", cLR="#63c5da", cNL="#710193", cNA="#aaaaaa"),
  cluster_nllgd_name = c(cHR="#fc6a03", cLR="#63c5da", cNL="#710193"),
  path_aoi = c(nl="#0044ff", lgd="#00ffff", hgd="#ffff00", inv="#ff0000",
               igd="#00ff00", duct="#0044ff", panin="#88ff88", pdac="#ff0000"),
  path_specimen = c(ipmn_lg="#00ffff", ipmn_pdac="#cc0000",
                    ipmn_ig="#00ff00", ipmn_hg="#ffff00",
                    pdac="#ff0000", donor="#0000ff", pnet="#880088"),
  path = c(ipmn_lg_nl ="#0044ff", 
           ipmn_lg_lgd ="#00ffff", 
           ipmn_pdac_nl ="#ae00ff", 
           ipmn_pdac_lgd ="#42ff00", 
           ipmn_pdac_hgd ="#ffff00", 
           ipmn_pdac_inv ="#ff0000"),
  de_cptac_rna = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"), 
  de_cptac_prot = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"),
  prognostic = c(no="#aaaaaa", fav="#00FFFF", unfav="#FFFF00", na="#FFFFFF"),
  moffitt_basallike = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"),
  moffitt_classical = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF"),
  moffitt_exocrine = c(no="#aaaaaa", dn="#00FFFF", up="#FFFF00", na="#FFFFFF")
)

shape_scales <- list(
  study = c(duke=16, cc=15, soa=17, gol=18),
  path_specimen = c(donor=15, 
                    ipmn_lg=17,
                    ipmn_ig=17,
                    ipmn_hg=16,
                    ipmn_pdac=16,
                    pdac=16,
                    pnet=15),
  segment = c(panck=16, cd3=4, cd45=8, otherstroma=15)
)

```

# Functions

## functions stgeomx package

```{r stgeomx}

stgeomx_init <- function(samples, meta, counts, x) {
  ds <- list(
    samples = samples,
    meta = meta,
    counts = counts,
    x = x
  )
  class(ds) <- "stgeomx"
  return(ds)
}


stgeomx_copy <- function(ds) {
  x <- list(
    samples = ds$samples,
    meta = ds$meta,
    counts = ds$counts,
    x = ds$x
  )
  class(x) <- "stgeomx"
  return(x)
}


stgeomx_prepare_input <- function(samples, counts,
                          slide_key_col,
                          roi_key_col,
                          segment_key_col,
                          aoi_key_col,
                          probe_key_col,
                          gene_key_col,
                          negprobe_regex) {
  # ensure all samples found in counts
  inds <- match(pull(samples, {{ aoi_key_col }}), colnames(counts))
  stopifnot(all(!is.na(inds)))

  # break counts into metadata and count data
  meta <- counts %>%
    select("probe" = all_of({{ probe_key_col }}), 
           "gene" = all_of({{ gene_key_col }})) %>%
    mutate(bg = stringr::str_detect(.data$gene, negprobe_regex), .after="gene")

  # select valid samples
  counts <- select(counts, all_of(inds))
  # ensure samples and count tables are aligned
  stopifnot(all(pull(samples, {{ aoi_key_col }}) == colnames(counts)))

  # create new identifiers for slide, roi, segment, aoi
  slide_keys <- pull(samples, {{ slide_key_col }})
  samples <- samples %>% mutate(
    slide = sprintf("s%02d", match(slide_keys, unique(slide_keys))),
    roi = sprintf("r%s", .data[[roi_key_col]]),
    segment = tolower(gsub(" ", "", .data[[segment_key_col]], fixed = TRUE)),
    aoi = sprintf("%s_%s_%s", .data$slide, .data$roi, .data$segment)
  )
  stopifnot(n_distinct(samples$aoi) == nrow(samples))

  # relabel columns of count matrix with new aoi id
  colnames(counts) <- samples$aoi
  stopifnot(all(samples$aoi == colnames(counts)))

  # create dataset
  ds <- stgeomx_init(samples, meta, counts, counts)
  return(ds)
}


stgeomx_read_xlsx <- function(xlsx_file,
                      sample_sheet = "SegmentProperties",
                      count_sheet = "BioProbeCountMatrix",
                      slide_key_col = "ScanLabel",
                      roi_key_col = "ROILabel",
                      segment_key_col = "SegmentLabel",
                      aoi_key_col = "SegmentDisplayName",
                      probe_key_col = "ProbeDisplayName",
                      gene_key_col = "TargetName",
                      negprobe_regex = "^NegProbe") {
  # read samples (remove duplicates based on aoi key)
  samples <- readxl::read_excel(xlsx_file, sheet=sample_sheet,
                                na = c("", "NaN")) %>%
    distinct(.data[[aoi_key_col]], .keep_all=TRUE)

  # read counts
  counts <- readxl::read_excel(xlsx_file, sheet=count_sheet,
                               .name_repair = "minimal")

  # prepare dataset
  ds <- stgeomx_prepare_input(samples, counts, slide_key_col, roi_key_col,
                      segment_key_col, aoi_key_col, probe_key_col,
                      gene_key_col, negprobe_regex)
  return(ds)
}


geomean <- function(x) {
  return(exp(mean(log(x))))
}


compute_quantiles_kde <- function(x, bg, bw.adjust=2) {
  # setup
  n <- 2^13
  xlog <- log2(x+1)
  xmin <- 0
  xbgmax <- max(xlog[bg])
  # select bandwidth
  bwbg <- stats::bw.nrd0(xlog[bg]) * bw.adjust
  # smooth ecdf functions for neg probes
  dbg <- stats::density(xlog[bg], bw=bwbg, n=n, from=0, to=xbgmax)
  dbgcdf <- cumsum(dbg$y) / sum(dbg$y)
  # find bg quantiles
  q <- stats::approx(dbg$x, dbgcdf, xout=xlog,
                     yleft=0, yright=1, rule=2, ties="ordered")$y
  return(q)
}


calc_bg_quantiles <- function(ds) {
  # calculate probe quantiles relative to negative probes
  bg_quants <- bind_cols(select(ds$meta, "bg"), ds$counts) %>%
    reframe(across(-"bg", ~ compute_quantiles_kde(.x, .data$bg)))
  return(bg_quants)
}


remove_empty_aois <- function(ds, min_count = 1) {
  # initial filter of "empty" samples with essential zero counts
  count_max = apply(ds$counts, 2, max)
  keep <- count_max > min_count
  fds <- stgeomx_init(
    samples = filter(ds$samples, keep),
    meta = ds$meta,
    counts = ds$counts[, keep],
    x = ds$x[, keep]
  )
  return(fds)
}


calc_qc_metrics <- function(samples, meta, counts, bg_lod_quantile) {
  # unpack dataset for readability
  bg <- meta$bg
  x <- counts[!bg, ]
  xbg <- counts[bg, ]

  # add background metrics
  samples <- samples %>% mutate(
    bg_counts = colSums(xbg),
    bg_cpm = 1e6 * colSums(xbg) / colSums(counts),
    bg_geomean = apply(xbg, 2, geomean),
    bg_median = apply(xbg, 2, stats::median),
    bg_q3 = apply(xbg, 2, stats::quantile, 0.75),
    bg_q95 = apply(xbg, 2, stats::quantile, 0.95),
    bg_max = apply(xbg, 2, max),
    bg_lod = apply(xbg, 2, stats::quantile, bg_lod_quantile)
  )
  # add gene metrics
  samples <- samples %>% mutate(
    num_counts = colSums(x),
    complexity = apply(x, 2, function(x) length(unique(x))),
    count_mad = apply(x, 2, stats::mad),
    count_iqr = apply(x, 2, stats::IQR),
    count_geomean = apply(x, 2, geomean),
    count_median = apply(x, 2, stats::median),
    count_q3 = apply(x, 2, stats::quantile, 0.75),
    count_q95 = apply(x, 2, stats::quantile, 0.95),
    count_max = apply(x, 2, max)
  )
  # gene probe vs background probe metrics (signal-to-noise)
  samples <- samples %>% mutate(
    snr_geomean = .data$count_geomean / .data$bg_geomean,
    snr_median = .data$count_median / .data$bg_median,
    snr_q3 = .data$count_q3 / .data$bg_q3,
    snr_q95 = .data$count_q95 / .data$bg_q95
  )
  return(samples)
}


calc_frac_expr <- function(counts, bg, bg_lod) {
  x <- sweep(counts, 2, bg_lod, FUN="-")
  x <- apply(x, 2, pmax, 0)
  x <- x > 0
  gene_frac_expr <- rowSums(x) / ncol(x)
  y <- x[bg == FALSE,]
  sample_frac_expr <- colSums(y) / nrow(y)
  return(list(s=sample_frac_expr, g=gene_frac_expr))
}


calc_snauc <- function(meta, counts) {
  # calculate AUC of gene vs background (negative probe)
  calc_roi_auc <- function(responses, predictors) {
    myroc <- pROC::roc(responses, predictors, levels=c(TRUE, FALSE), direction="<")
    myauc <- as.numeric(myroc$auc)
    return(myauc)
  }
  bg_auc <- bind_cols(select(meta, "bg"), counts) %>%
    summarize(across(-"bg", ~ calc_roi_auc(.data$bg, .x))) %>%
    unlist()
  return(bg_auc)
}


stgeomx_preprocess <- function(ds, bg_lod_quantile=0.9) {
  # first remove empty/control AOIs
  ds <- remove_empty_aois(ds)

  # unpack dataset
  samples <- ds$samples
  meta <- ds$meta
  bg <- meta$bg
  counts <- ds$counts

  # if dataset is 1-based, convert to 0-based for consistency
  if (min(ds$counts) > 0) {
    counts <- counts - min(ds$counts)
  }

  # compute basic qc metrics (returns sample table)
  samples <- calc_qc_metrics(samples, meta, counts, bg_lod_quantile)

  # separate bg and gene counts
  x <- counts[!bg, ]
  xbg <- counts[bg, ]

  # probes expressed over detection limit
  x <- calc_frac_expr(counts, bg, samples$bg_lod)
  samples$frac_expr <- x$s
  meta$frac_expr <- x$g

  # calculate AUC of gene vs background (negative probe)
  samples$bg_auc <- calc_snauc(meta, counts)

  # KS test of gene vs background
  bg_ks_dist <- NULL
  bg_ks_pval <- NULL
  for (i in 1:ncol(counts)) {
    x <- unlist(counts[!bg, i])
    xbg <- unlist(counts[bg, i])
    res <- suppressWarnings(stats::ks.test(x, xbg, alternative="less"))
    bg_ks_dist[i] <- res$statistic
    bg_ks_pval[i] <- res$p.value
  }
  samples <- mutate(samples,
    bg_ks_dist = bg_ks_dist,
    bg_ks_pval = bg_ks_pval
  )

  # filtering defaults
  samples$keep <- TRUE
  meta$keep <- TRUE

  ds <- stgeomx_init(
    samples = samples,
    meta = meta,
    counts = counts,
    x = counts
  )
  return(ds)
}


stgeomx_merge_probes <- function(ds) {
  # convenience function for geomean
  geomean_fast <- function(x) {
    ifelse(length(x) == 1, x, 2^(mean(log2(x+1))))
  }

  # do not merge the negative probes, instead just move the probe name
  # to the gene name
  ybg <- bind_cols(ds$meta, ds$counts) %>%
    filter(.data$bg) %>%
    rowwise() %>%
    summarize(
      gene = .data$probe,
      num_probes = 1,
      probes = .data$probe,
      bg = TRUE,
      frac_expr = .data$frac_expr,
      across(colnames(ds$counts), geomean_fast)
    ) %>% arrange(gene)
  bgmeta <- ybg %>% select(-colnames(ds$counts))
  bgcounts <- ybg %>% select(colnames(ds$counts))

  # merge gene probes
  y <- bind_cols(ds$meta, ds$counts) %>%
    filter(!.data$bg) %>%
    group_by(.data$gene) %>%
    summarize(
      num_probes = n(),
      probes = stringr::str_flatten(.data$probe, collapse=","),
      bg = FALSE,
      frac_expr = mean(.data$frac_expr),
      across(colnames(ds$counts), geomean_fast)
    ) %>% arrange(gene)
  meta <- y %>% select(-colnames(ds$counts))
  counts <- y %>% select(colnames(ds$counts))

  ds <- stgeomx_init(
    samples = ds$samples,
    meta = bind_rows(meta, bgmeta),
    counts = bind_rows(counts, bgcounts),
    x = bind_rows(counts, bgcounts)
  )
  return(ds)
}



stgeomx_set_thresholds <- function(ds,
                           min_counts = 0,
                           min_auc = 0.5,
                           aoi_min_frac_expr = 0.0,
                           gene_min_frac_expr = 0.0) {
  # apply filtering parameters to samples
  ds$samples <- ds$samples %>% mutate(
    keep = ((.data$num_counts > min_counts) &
              (.data$bg_auc > min_auc) &
              (.data$frac_expr > aoi_min_frac_expr))
  )
  # apply filtering parameters to genes
  ds$meta <- ds$meta %>% mutate(
    keep = .data$frac_expr > gene_min_frac_expr,
    expressed = case_when(.data$bg ~ "bg",
                          !(.data$keep) ~ "no",
                          TRUE ~ "yes")
  )
  return(ds)
}


stgeomx_filter_aois <- function(ds) {
  fds <- stgeomx_init(
    samples = filter(ds$samples, .data$keep),
    meta = ds$meta,
    counts = ds$counts[, ds$samples$keep],
    x = ds$x[, ds$samples$keep]
  )
  return(fds)
}

stgeomx_filter_genes <- function(ds) {
  keep_genes <- ds$meta$keep & (!ds$meta$bg)
  fds <- stgeomx_init(
    samples = ds$samples,
    meta = filter(ds$meta, keep_genes),
    counts = ds$counts[keep_genes,],
    x = ds$x[keep_genes,]
  )
  return(fds)
}


stgeomx_apply_filters <- function(ds) {
  keep_genes <- ds$meta$keep & (!ds$meta$bg)
  fds <- stgeomx_init(
    samples = filter(ds$samples, .data$keep),
    meta = filter(ds$meta, keep_genes),
    counts = ds$counts[keep_genes, ds$samples$keep],
    x = ds$x[keep_genes, ds$samples$keep]
  )
  return(fds)
}


stgeomx_plot_aoi_filter <- function(ds,
                            min_counts=0,
                            min_auc=0.5,
                            min_frac_expr=0) {
  # apply cutoffs
  s <- ds$samples %>% mutate(
    keep = ((.data$num_counts > min_counts) &
              (.data$bg_auc > min_auc) &
              (.data$frac_expr > min_frac_expr))
  )

  keep_aois <- table(s$keep)
  keep_pct <- 100 * keep_aois["TRUE"] / nrow(s)
  subtitle <- sprintf("Retained %d/%d (%.2f%%) AOIs", keep_aois["TRUE"], nrow(s), keep_pct)

  p1 <- ggplot(s, aes(x=.data$num_counts+1, y=.data$frac_expr, color=.data$bg_cpm)) +
    geom_point(alpha=0.6) +
    geom_vline(xintercept = min_counts, linetype = "dashed", color = "red") +
    geom_hline(yintercept = min_frac_expr, linetype = "dashed", color = "red") +
    scale_color_viridis_c() +
    scale_x_log10() +
    labs(x="Counts+1", y="Frac detectable genes", color="BG CPM",
         title="AOI QC Metrics",
         subtitle=subtitle) +
    theme_minimal()

  p2 <- ggplot(ds$samples, aes(x=.data$num_counts+1, y=.data$bg_auc, color=.data$bg_cpm)) +
    geom_point(alpha=0.6) +
    geom_vline(xintercept = min_counts, linetype = "dashed", color = "red") +
    geom_hline(yintercept = min_auc, linetype = "dashed", color = "red") +
    scale_color_viridis_c() +
    scale_x_log10() +
    labs(x="Counts+1", y="snAUC", color="BG CPM") +
    theme_minimal()

  p3 <- ggplot(ds$samples, aes(x=.data$bg_auc, y=.data$frac_expr, color=.data$keep)) +
    geom_point(alpha=0.6) +
    geom_vline(xintercept = min_auc, linetype = "dashed", color = "red") +
    geom_hline(yintercept = min_frac_expr, linetype = "dashed", color = "red") +
    scale_color_manual(values = pals::cols25()) +
    scale_x_log10() +
    labs(x="snAUC", y="Frac detectable genes", color="QC Filter") +
    theme_minimal()

  p <- p1 + p2 + p3 + patchwork::plot_layout(guides="collect")
  return(p)
}



stgeomx_plot_gene_filter <- function(ds, min_frac_expr=0) {
  # apply cutoffs
  meta <- ds$meta %>% mutate(
    keep = .data$frac_expr > min_frac_expr,
    expressed = case_when(.data$bg ~ "bg",
                          !.data$keep ~ "no",
                          TRUE ~ "yes")
  )

  # ROC curve analysis for expressed vs bg
  myroc <- pROC::roc(meta$bg, meta$frac_expr, levels=c(TRUE, FALSE), direction="<")
  mystats <- pROC::coords(myroc, x=min_frac_expr, input="threshold", ret="all", transpose=FALSE)

  label_auc <- paste0("AUC = ", round(myroc$auc, 3))
  label_fpr <- paste0("FPR = ", round(mystats$fpr, 3))
  label_stats <- sprintf("Threshold=%.2f Expr=%d Filtered=%d Recall=%.2f",
                         min_frac_expr, mystats$tp, mystats$fn, mystats$recall)

  # density plot showing bg, undetectable (filtered), and detected (expressed)
  p1 <- ggplot(meta, aes(x=.data$frac_expr, y=.data$expressed, fill=.data$expressed)) +
    ggridges::geom_density_ridges(scale=5, alpha=0.5) +
    geom_vline(xintercept = min_frac_expr, linetype="dashed", color="red") +
    scale_fill_manual(values = pals::cols25()) +
    ggridges::theme_ridges() +
    labs(x="Frac detectable above BG", y="Expressed", fill="Expressed",
         title="Gene detection distribution",
         subtitle=label_stats)

  # threshold vs false positive rate
  x <- pROC::coords(myroc, x="all", input="threshold", ret=c("threshold", "fpr", transpose = FALSE))
  p2 <- ggplot(x, aes(x=.data$threshold, y=.data$fpr)) +
    geom_line(linewidth=1, alpha=0.5, color="blue") +
    geom_vline(xintercept = min_frac_expr, linetype="dashed", color="red") +
    geom_hline(yintercept = mystats$fpr, linetype="dashed", color="red") +
    #annotate("text", label=label_fpr, x=min_frac_expr, y=mystats$fpr, hjust=0, vjust=-0.25) +
    scale_x_continuous(breaks = seq(0, 1, by=0.1)) +
    theme_minimal() +
    theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) +
    labs(x="Frac Detectable", y="FPR", title=label_fpr)

  p3 <- pROC::ggroc(myroc) +
    geom_vline(xintercept = mystats$specificity, linetype="dashed", color="red") +
    geom_hline(yintercept = mystats$sensitivity, linetype="dashed", color="red") +
    theme_minimal() +
    labs(title = label_auc)

  p <- p1 / (p2 + p3)
  return(p)
}


stgeomx_plot_bgcorrect <- function(ds, quantiles=c(0.10, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99)) {
  # gene metadata
  bg <- ds$meta$bg
  keep <- ds$meta$keep
  # labels for plot legend
  quant_names <- paste0("q", round(100 * quantiles))
  # sample columns used for plot
  s <- select(ds$samples, "aoi", "slide", "keep")
  # compute background level per aoi
  s$bg_cpm <- 1e6 * colSums(ds$counts[bg,]) / colSums(ds$counts)
  
  #x <- ds$x
  #counts <- ds$counts
  # filter genes
  x <- ds$x[(!bg) & keep,]
  counts <- ds$counts[(!bg) & keep,]

  # gene cpm before bgcorrect
  xgene <- as_tibble(cpm(counts))
  # xgene <- filter(ds$counts, !bg)
  # xgene <- as_tibble(cpm(xgene))
  qgene <- xgene %>%
    reframe(across(everything(), ~ stats::quantile(.x, quantiles))) %>%
    mutate(name="gene", quant=quant_names)

  # cpm after bgcorrect
  y <- as_tibble(cpm(x))
  #y <- as_tibble(cpm(x[!bg, ]))
  qy <- y %>%
    reframe(across(everything(), ~ stats::quantile(.x, quantiles))) %>%
    mutate(name="y", quant=quant_names)

  # combine bg and gene quantiles
  q <- bind_rows(qgene, qy) %>%
    tidyr::pivot_longer(colnames(x), names_to="aoi", values_to="value") %>%
    tidyr::pivot_wider(names_from="name", values_from="value", names_prefix="") %>%
    inner_join(s, by=c("aoi"="aoi"), suffix=c("_gene", "_aoi"))

  p1 <- ggplot(filter(q, .data$keep), aes(x=.data$bg_cpm+1, y=.data$gene+1, color=.data$quant, shape=.data$keep)) +
  #p1 <- ggplot(q, aes(x=.data$bg_cpm+1, y=.data$gene+1, color=.data$quant, shape=.data$keep)) +
    geom_point(alpha=0.6) +
    geom_smooth(data=filter(q, .data$keep), linetype='dashed', method = 'lm',
                formula = y ~ x, se=FALSE, alpha=0.6) +
    scale_color_viridis_d(option="plasma") +
    #scale_x_log10() +
    #scale_y_log10() +
    theme_minimal() + 
    labs(x="Noise (CPM)", y="Raw Counts (CPM)")

  p2 <- ggplot(filter(q, .data$keep), aes(x=.data$bg_cpm+1, y=.data$y+1, color=.data$quant, shape=.data$keep)) +
#  p2 <- ggplot(q, aes(x=.data$bg_cpm+1, y=.data$y+1, color=.data$quant, shape=.data$keep)) +
    geom_point(alpha=0.6) +
    geom_smooth(data=filter(q, .data$keep), linetype='dashed', method = 'lm',
                formula = y ~ x, se=FALSE, alpha=0.6) +
    scale_color_viridis_d(option="plasma") +
    #scale_x_log10() +
    #scale_y_log10() +
    theme_minimal() +
    labs(x="Noise (CPM)", y="Normalized Counts (CPM)")
  return(p1 + p2)
}


stgeomx_plot_expr_dist <- function(ds, bg.lod.quantile=0.9) {
  s <- select(ds$samples, "aoi", "slide", "keep")
  x <- ds$x
  bg <- ds$meta$bg

  # compute background level per aoi
  s$bg_lod <- apply(x[bg,], 2, stats::quantile, bg.lod.quantile)

  # quantiles
  quantiles <- c(0.25, 0.5, 0.75, 0.90, 0.95, 0.99)
  quant_names <- paste0("q", round(100 * quantiles))

  q <- as_tibble(x) %>%
    filter(!bg) %>%
    reframe(across(everything(), ~ stats::quantile(.x, quantiles))) %>%
    mutate(quant=quant_names) %>%
    tidyr::pivot_longer(colnames(x), names_to="aoi", values_to="value") %>%
    tidyr::pivot_wider(names_from="quant", values_from="value", names_prefix="") %>%
    inner_join(s, by=c("aoi"="aoi"), suffix=c("_gene", "_aoi"))

  data1 <- q %>%
    rowwise() %>%
    arrange(.data$keep, .data$q50) %>%
    mutate(x=factor(.data$aoi, .data$aoi)) %>%
    ungroup()

  data2 <- data1 %>%
    select("x", "slide", "keep", "q50", "q90", "q95", "q99") %>%
    tidyr::pivot_longer(cols = c("q50", "q90", "q95", "q99"),
                        names_to="quant", values_to="value")

  p <- ggplot() +
    geom_errorbar(data=data1, aes(x=.data$x, ymin=.data$q25+1, ymax=.data$q75+1, color=.data$keep), alpha=0.75) +
    #geom_segment(data=data1, aes(x=x, xend=x, y=q25, yend=q75), color="grey", alpha=0.75) +
    geom_point(data=data1, aes(x=.data$x, y=.data$bg_lod+1), color="black", shape=4, alpha=0.75) +
    geom_point(data=data2, aes(x=.data$x, y=.data$value+1, fill=.data$quant), shape=21, alpha=0.75) +
    scale_color_manual(values=pals::cols25()) +
    scale_fill_viridis_d() +
    scale_y_log10() +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank()) +
    facet_grid(~ slide, scales="free_x") +
    labs(x = "AOI", y="Gene Expression")

  return(p)
}


bgcorrect_qq <- function(x, bg, bw.adjust=1, bg.quant=0.5) {
  # subtract constant background noise
  bg.loq <- stats::quantile(x[bg], bg.quant)
  x <- pmax(0, x - bg.loq)

  # setup
  n <- 2^11
  xlog <- log2(x+1)

  # select bandwidth
  bwbg <- stats::bw.nrd0(xlog[bg]) * bw.adjust
  bw <- stats::bw.nrd0(xlog[!bg]) * bw.adjust
  bw <- max(bw, bwbg)

  xmin <- 0
  xmax <- max(xlog[!bg])
  xbgmax <- max(xlog[bg])

  # smooth ecdf functions for gene probes and neg probes
  d <- stats::density(xlog[!bg], bw=bw, n=n, from=0, to=xmax)
  dcdf <- cumsum(d$y) / sum(d$y)

  dbg <- stats::density(xlog[bg], bw=bw, n=n, from=0, to=xbgmax)
  dbgcdf <- cumsum(dbg$y) / sum(dbg$y)

  # map gene probes to quantiles
  q <- stats::approx(d$x, dcdf, xout=xlog, ties="ordered")$y
  # map quantiles to background probes
  log2noise <- stats::approx(dbgcdf, dbg$x, yleft=0, yright=xbgmax, rule=2,
                             xout=q, ties="ordered")$y
  # convert from log2 to linear scale and round up to discrete counts
  noise <- ceiling(2^log2noise - 1)
  #noise <- ceiling(2^log2noise - 1)
  # ensure noise is not greater than x
  noise <- pmin(noise, x)
  y <- x - noise

  res <- list(
    x = x,
    bg = bg,
    bw.adjust = bw.adjust,
    bg.quant = bg.quant,
    bg.loq = bg.loq,
    bw = bw,
    y = y,
    noise = noise,
    dx = d$x,
    dy = d$y,
    dcdf = dcdf,
    dbgx = dbg$x,
    dbgy = dbg$y,
    dbgcdf = dbgcdf
  )
}


stgeomx_bgcorrect <- function(ds, method=c("qq", "bgsub", "none"),
                      bw.adjust=1, bg.quant=0.5) {
  # check method
  method <- match.arg(method)
  bg <- ds$meta$bg
  x <- ds$counts

  apply_bgcorrect_bgsub <- function(x) {
    xsub <- ceiling(stats::quantile(x[bg], bg.quant))
    y <- pmax(0, x - xsub)
    return(y)
  }
  apply_bgcorrect_qq <- function(x) {
    res <- bgcorrect_qq(x, bg, bw.adjust=bw.adjust, bg.quant=bg.quant)
    return(res$y)
  }

  f <- switch(method,
              qq = apply_bgcorrect_qq,
              bgsub = apply_bgcorrect_bgsub,
              none = function(x) return(x))
  x <- apply(x, MARGIN=2, FUN=f)
  ds$x <- x
  return(ds)
}


cpm <- function(x) {
  sweep(x, 2, colSums(x) / 1e6, FUN="/")
}

cpm2counts <- function(x, tot_counts) {
  sweep(x, 2, tot_counts / 1e6, FUN="*")
}


qnorm <- function(x) {
  # convert to relative counts (counts-per-million)
  xcpm <- apply(x, 2, function(x) { 1e6 * x / sum(x) })
  # compute quantiles (adding a pseudocount)
  xquant <- apply(xcpm, 2, sort)
  xquant <- apply(xquant, 1, function(x) { 2^(mean(log2(x+1))) })
  xquant <- 1e6 * xquant / sum(xquant)
  # ranks (with ties present)
  xrank <- as.data.frame(apply(x, 2, rank, ties.method="max"))
  # apply quantile normalization procedure using new ranks
  xnorm <- apply(xrank, 2, function(x) { xquant[x] })
  # renormalize (shift and scale) such that columns sum to one million
  renorm <- function(x) {
    x <- x - min(x)
    return(1e6 * x / sum(x))
  }
  xnorm <- apply(xnorm, 2, renorm)
  return(xnorm)
}


#'
#' edgeR package
#' Copied from https://rdrr.io/bioc/edgeR/src/R/calcNormFactors.R
#' Cite Mark Robinson
#' Scale factors as in Anders et al (2010)
#' Mark Robinson
#'
#' @param data matrix of counts
calc_norm_factors_rle <- function(data) {
  gm <- exp(rowMeans(log(data)))
  apply(data, 2, function(u) stats::median((u/gm)[gm > 0]))
}


stgeomx_normalize <- function(ds, method=c("qn", "rle", "cpm", "q3", "none")) {
  # check method
  method <- match.arg(method)
  x <- ds$x

  if (method == "none") {
    # do nothing
    x <- x
  } else if (method == "qn") {
    x <- qnorm(x)
  } else if (method == "cpm") {
    x <- cpm(x)
  } else {
    # methods Q3, RLE, and TMM produce normalization factors
    sf <- colSums(x) / 1e6

    if (method == "rle") {
      nf <- calc_norm_factors_rle(x) / sf
    } else if (method == "q3") {
      nf <- pmax(1, apply(x, 2, stats::quantile, 0.75)) / sf
    }
    # norm factors should multiply to 1
    nf <- nf/exp(mean(log(nf)))
    x <- sweep(x, 2, sf * nf, FUN="/")
  }

  rownames(x) <- ds$meta$gene
  ds$x <- x
  return(ds)
}


```


## functions geomx pkc/dcc files

```{r read dcc and pkc}

stgeomx_read_pkc <- function(pkc_file) {
  pkc <- yaml::read_yaml(pkc_file)
  probes <- tibble(target = pkc$Targets)
  probes <- probes %>% tidyr::unnest_wider(target) %>%
    dplyr::rename(TargetName = DisplayName) %>%
    tidyr::unnest_longer(Probes) %>%
    tidyr::unnest_wider(Probes) %>%
    dplyr::rename(ProbeDisplayName = DisplayName) %>%
    dplyr::mutate(bg = startsWith(CodeClass, "Negative"))
  return(probes)
}

stgeomx_read_dcc <- function(dcc_tar_file, dcc_dir) {
  # cleanup
  if (file.exists(dcc_dir)) {
    unlink(dcc_dir, recursive = TRUE)
  }
  # extract dcc tar file
  untar(dcc_tar_file, exdir=dcc_dir)
  # read dcc files
  dcc_files <- list.files(dcc_dir, full.names=TRUE)
  dcc_data <- sapply(dcc_files, GeomxTools::readDccFile, simplify=FALSE)
  dcc_data <- tibble(dcc = dcc_data) %>%
    tidyr::unnest_wider(dcc, simplify=FALSE)
  
  samples <- dcc_data %>% 
    dplyr::select(Header, Scan_Attributes, NGS_Processing_Attributes) %>%
    tidyr::unnest(cols = c(Header, Scan_Attributes, NGS_Processing_Attributes))
  
  counts <- dcc_data %>%
    dplyr::select(Scan_Attributes, Code_Summary) %>% 
    tidyr::hoist(Scan_Attributes, SampleID="SampleID") %>% 
    dplyr::select(-Scan_Attributes) %>%
    tidyr::unnest(Code_Summary) %>% 
    tidyr::pivot_wider(names_from=SampleID, values_from=Count, values_fill=1)
  
  # TODO: not sure whether to leave values at zero or 1
  # TODO: could add one to all counts
  # TODO: needs to be consistent with the way DSP server processes
  # cleanup
  if (file.exists(dcc_dir)) {
    unlink(dcc_dir, recursive = TRUE)
  }
  return(list(samples=samples,
              counts=counts))
}

stgeomx_join_pkc_dcc <- function(probes, samples, counts) {
  # join probes with counts
  counts <- left_join(probes, counts, by="RTS_ID")
  # ensure all samples found in counts
  inds <- match(pull(samples, SampleID), colnames(counts))
  stopifnot(all(!is.na(inds)))
  # break counts into metadata and count data
  meta <- select(counts, rts=RTS_ID, probe=ProbeDisplayName, gene=TargetName, bg)
  # select valid samples
  counts <- select(counts, all_of(inds))
  # ensure samples and count tables are aligned
  stopifnot(all(pull(samples, SampleID) == colnames(counts)))
  # create new identifiers for slide, roi, segment, aoi
  slide_keys <- pull(samples, SlideName)
  samples$slide <- sprintf("s%02d", match(slide_keys, unique(slide_keys)))
  samples <- samples %>% mutate(
    roi = sprintf("r%s", ROILabel),
    segment = tolower(gsub(" ", "", SegmentLabel, fixed = TRUE)),
    aoi = sprintf("%s_%s_%s", slide, roi, segment)
  )
  stopifnot(n_distinct(samples$aoi) == nrow(samples))
  # relabel columns of count matrix with new aoi id
  colnames(counts) <- samples$aoi
  stopifnot(all(samples$aoi == colnames(counts)))

  ds <- list(samples=samples,
             meta=meta,
             counts=counts)
  return(ds)
}

stgeomx_add_study_field <- function(ds, study_name) {
  s <- ds$samples
  s <- s %>% mutate(study = study_name,
                    aoi = sprintf("%s_%s", study_name, aoi),
                    roi = sprintf("%s_%s_%s", study_name, slide, roi),
                    slide = sprintf("%s_%s", study_name, slide),
                    patient = sprintf("%s_%s", study_name, patient))
  colnames(ds$counts) <- s$aoi
  colnames(ds$x) <- s$aoi
  ds <- stgeomx_init(samples = s,
                     meta = ds$meta, 
                     counts = ds$counts,
                     x = ds$x)
  return(ds)
}


```

## functions merge datasets

```{r functions merge datasets}

merge_ds <- function(a, b) {
  # merge sample tables
  sample_cols <- intersect(colnames(a$samples), colnames(b$samples))
  s <- bind_rows(select(a$samples, all_of(sample_cols)),
                 select(b$samples, all_of(sample_cols)))
  # merge count data
  a_counts <- bind_cols(select(a$meta, probe, gene, bg), a$counts)
  b_counts <- bind_cols(select(b$meta, probe, gene, bg), b$counts)
  counts <- inner_join(a_counts, b_counts, by=c("probe", "gene", "bg"),
                       relationship="one-to-one")
  # break into metadata and count data
  meta <- select(counts, -all_of(s$aoi))
  counts <- select(counts, all_of(s$aoi))
  # fraction expressed 
  ds <- stgeomx_init(samples=s, meta=meta, counts=counts, x=counts)
  return(ds)
}

```

## functions highly variable genes

```{r highly variable gene functions}

get_var_tbl <- function(x, span=0.5) {
  tbl <- bind_cols(
    gene = rownames(x),
    u = rowMeans(x),
    v = apply(x, 1, function(x) var(x))
  )
  lovar <- loess(v ~ u, data=tbl, span=span)
  tbl$lovar <- predict(lovar)
  tbl$vadj <- tbl$v / tbl$lovar
  tbl <- tbl %>% mutate(rank = min_rank(-vadj))
  return(tbl)
}

most_var_genes <- function(x, ntop=500, span=0.50) {
  tbl <- bind_cols(
    u = rowMeans(x),
    v = apply(x, 1, function(x) var(x))
  )
  lovar <- loess(v ~ u, data=tbl, span=span)
  tbl$lovar <- predict(lovar)
  tbl$vadj <- tbl$v / tbl$lovar
  keep <- order(tbl$vadj, decreasing=TRUE)[seq_len(min(ntop, nrow(tbl)))]
  return(x[keep,])
}

maplot <- function(y, ntop=1000, nlabel=100, myspan=0.5) {
  tbl <- get_var_tbl(y, span=myspan)
  label_data <- filter(tbl, rank <= nlabel)
  top_data <- filter(tbl, rank < ntop)
  bot_data <- filter(tbl, rank >= ntop)
  p <- ggplot(tbl) +
    geom_point(data=bot_data, aes(u, v), alpha=0.5, size=1, color="#aaaaaa") +
    geom_point(data=top_data, aes(u, v), alpha=0.8, size=2, color="#ffaaaa") +
    geom_line(aes(u, lovar), color="blue") +
    geom_text_repel(data=label_data, aes(u, v, label=gene), color="black", size=3, max.overlaps=Inf) +
    theme_minimal()
  return(p)
}

heatmap_genes <- function(s, x, filename="heatmap.pdf", width=10, height=10) {
  s <- s %>% mutate(
    noise_quantile = dplyr::ntile(dplyr::desc(bg_auc), 8)
  )
  annot_col <- column_to_rownames(s, "aoi") %>%
    select(study, path_specimen, path_aoi, segment, patient) %>%
    as.data.frame()
  p <- pheatmap(x,
                scale="row",
                show_rownames=TRUE,
                show_colnames=FALSE,
                border_color=NA,
                clustering_distance_rows = "euclidean",
                clustering_distance_cols = "euclidean",
                clustering_method = "ward.D2",
                # clustering_method = "average",
                # clustering_distance_rows = "correlation",
                # clustering_distance_cols = "correlation",
                breaks = seq(-3, 3, 0.05),
                color = colorRampPalette(c("purple3", "black", "yellow2"))(120),
                annotation_col = annot_col,
                annotation_colors = color_scales,
                fontsize=4,
                filename=filename,
                width=width,
                height=height)
  return(p)
}

```


## functions plotting

```{r plotting functions}

pca_plot <- function(x, xvar, yvar, color_by, shape_by) {
  color_by_string <- rlang::as_string(rlang::ensym(color_by))
  shape_by_string <- rlang::as_string(rlang::ensym(shape_by))
  p <- ggplot(x, aes(x={{xvar}}, y={{yvar}}, color={{color_by}}, shape={{shape_by}})) +
    geom_point(alpha = 0.8, size=2) +
    scale_color_manual(values = color_scales[[color_by_string]]) +
    scale_shape_manual(values = shape_scales[[shape_by_string]]) +
    coord_fixed() +
    theme_bw()
  return(p)
}

save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```


# Read GeoMx datasets

## Read WTA probes

```{r wta probes}

# read pkc file
wta_probes <- stgeomx_read_pkc(wta_pkc_file)

```


## Read Carbone IPMN geomx data

```{r read carbone ipmn geomx data}

add_study_field <- function(ds, study_name) {
  s <- ds$samples
  s <- s %>% mutate(study = study_name,
                    aoi = sprintf("%s_%s", study_name, aoi),
                    roi = sprintf("%s_%s_%s", study_name, slide, roi),
                    slide = sprintf("%s_%s", study_name, slide),
                    patient = sprintf("%s_%s", study_name, patient))
  colnames(ds$counts) <- s$aoi
  return(list(samples = s, meta = ds$meta, counts = ds$counts))
}

# extract dcc data from tar file
ds <- stgeomx_read_dcc(cc_ipmn_geomx_dcc_tar, dcc_dir)

# read and join sample annotation
samples <- read_xlsx(cc_ipmn_geomx_annot_xlsx) %>%
  inner_join(ds$samples, by=join_by(SampleID))

ds <- stgeomx_join_pkc_dcc(wta_probes, samples, ds$counts)
ds <- add_study_field(ds, "cc")
ds <- stgeomx_init(samples = ds$samples,
                   meta = ds$meta, 
                   counts = ds$counts[, ds$samples$aoi],
                   x = ds$counts[, ds$samples$aoi])

# name dataset
cc_ipmn_ds <- ds

```

## Read Duke IPMN WTA geomx data

```{r duke ipmn wta geomx data, include=FALSE}

# read data
ds <- stgeomx_read_xlsx(pja_ipmn_geomx_xlsx)
# read and join sample annotation
annot <- read_excel(pja_ipmn_geomx_annot_xlsx)
ds$samples <- ds$samples %>% 
  inner_join(annot, by="SegmentDisplayName")
# for type compatibility during merge
ds$samples$ROILabel <- as.character(ds$samples$ROILabel)

ds <- stgeomx_init(samples = ds$samples,
                    meta = ds$meta, 
                    counts = ds$counts,
                    x = ds$counts)
ds <- stgeomx_add_study_field(ds, "duke")

# augment annotations
ds$samples <- ds$samples %>% mutate(
  path = paste0(path_specimen, "_", path_aoi),
  roi_group = paste0(patient, "_", str_pad(roi_group, width=2, pad="0")),
  roi_region = paste0(patient, "_", roi_region)
)

path_levels <- c("ipmn_lg_nl", "ipmn_lg_lgd", 
                 "ipmn_pdac_nl", "ipmn_pdac_lgd", "ipmn_pdac_hgd", "ipmn_pdac_inv")
# factors
ds$samples$path <- factor(ds$samples$path, levels=path_levels)

# name dataset
duke_ipmn_ds <- ds

```


## Read Spatial Organ Atlas pancreas geomx data

```{r soa panc geomx data, include=FALSE}

# read data
ds <- stgeomx_read_xlsx(soa_panc_geomx_xlsx)
# read and join sample annotation
annot <- read_excel(soa_panc_geomx_annot_xlsx)
ds$samples <- ds$samples %>% 
  inner_join(annot, by="SegmentDisplayName")
ds <- stgeomx_init(samples = ds$samples,
                   meta = ds$meta, 
                   counts = ds$counts,
                   x = ds$counts)
ds <- stgeomx_add_study_field(ds, "soa")

# adjust segment annotation
ds$samples <- ds$samples %>% mutate(
  segment = case_when(segment == "panck+" ~ "panck",
                      TRUE ~ segment)
)

# name dataset
soa_panc_ds <- ds

```


## Read UM GoL published study

```{r read umich gol, include=FALSE}

# extract dcc data from tar file
ds <- stgeomx_read_dcc(umich_gol_pub_dcc_tar, dcc_dir)

# read and join sample annotation
samples <- read_xlsx(umich_gol_pub_annot_xlsx) %>%
  inner_join(ds$samples, by=join_by(SampleID))

ds <- stgeomx_join_pkc_dcc(wta_probes, samples, ds$counts)
ds <- add_study_field(ds, "gol")
# for type compatibility during merge
ds$samples$ROILabel <- as.character(ds$samples$ROILabel)

ds <- stgeomx_init(samples = ds$samples,
                   meta = ds$meta, 
                   counts = ds$counts[, ds$samples$aoi],
                   x = ds$counts[, ds$samples$aoi])

# name dataset
gol_panc_ds <- ds

# extract epithelial duct/pdac aois
x <- ds$samples %>% filter(segment == "panck", path_aoi %in% c("duct", "panin", "pdac"))
gol_panc_panck_ds <- stgeomx_init(samples = x,
                                  meta = ds$meta, 
                                  counts = ds$counts[, x$aoi],
                                  x = ds$counts[, x$aoi])

```


# Read external gene sets

```{r read external datasets}

get_top_gene_sets <- function(tbl, log2fc_cutoff, padj_cutoff) {
  ntopgenes <- c(50, 100, 250, 500)
  gs <- list()
  for (a in unique(tbl$analysis)) {
    up_genes <- tbl %>% filter(
      analysis == a,
      log2fc >= log2fc_cutoff,
      fdr < padj_cutoff
    )
    dn_genes <- tbl %>% filter(
      analysis == a,
      log2fc <= -log2fc_cutoff,
      fdr < padj_cutoff
    )
    gs_name <- paste0(toupper(a), "_ALL_UP")
    gs[[gs_name]] <- pull(up_genes, gene_symbol)
    gs_name <- paste0(toupper(a), "_ALL_DN")
    gs[[gs_name]] <- pull(dn_genes, gene_symbol)

    for (n in ntopgenes) {
      x <- up_genes %>% slice_max(log2fc, n=n) %>% pull(gene_symbol)
      if (length(x) < n) { next }
      gs_name <- paste0(toupper(a), "_TOP", n, "_UP")
      gs[[gs_name]] <- x

      x <- dn_genes %>% slice_min(log2fc, n=n) %>% pull(gene_symbol)
      if (length(x) < n) { next }
      gs_name <- paste0(toupper(a), "_TOP", n, "_DN")
      gs[[gs_name]] <- x
    }
  }
  return(gs)
}

read_cptac_pdac_data <- function(filename, log2fc_cutoff, padj_cutoff) {
  rna <- read_excel(filename, sheet="Diff_exp_RNA")
  rna <- rna %>%
    dplyr::rename(gene_symbol = GeneSymbol,
                  log2fc = MedianLog2FoldChange,
                  pval = `p value`,
                  fdr = FDR) %>%
    dplyr::mutate(analysis = "cptac_rna", 
                  de = case_when(fdr > padj_cutoff ~ "no",
                                 abs(log2fc) <= log2fc_cutoff ~ "no",
                                 log2fc < -log2fc_cutoff ~ "dn",
                                 log2fc > log2fc_cutoff ~ "up",
                                 TRUE ~ "no"))
  
  # GeneSymbol	MedianLog2FoldChange	p value	FDR	
  # MedianLog2FoldChange.duct	p value.duct	FDR.duct	
  # Expression Coefficient	p value of adjusted expression	FDR of adjusted expression
  # add additional data?
  prot <- read_excel(cptac_panc_file, sheet="Diff_exp_prot")
  prot <- prot %>%
    dplyr::rename(
      gene_symbol = GeneSymbol,
      log2fc = MedianLog2FoldChange,
      pval = `p value`,
      fdr = FDR
    ) %>%
    select(gene_symbol, log2fc, pval, fdr) %>%
    mutate(analysis = "cptac_prot",
           de = case_when(fdr > padj_cutoff ~ "no",
                          abs(log2fc) <= log2fc_cutoff ~ "no",
                          log2fc < -log2fc_cutoff ~ "dn",
                          log2fc > log2fc_cutoff ~ "up",
                          TRUE ~ "no"))
  return(bind_rows(rna, prot))
}


read_hpa_data <- function(hpa_path_file) {
  x <- read.table(hpa_path_file, header=TRUE, sep="\t")
  colnames(x) <- c("ens_gene_id", "gene_symbol", "cancer_type",
                   "ihc_high", "ihc_medium", "ihc_low", "ihc_not_detected",
                   "prog_fav", "unprog_fav", "prog_unfav", "unprog_unfav")
  x <- as_tibble(x) %>%
    distinct(gene_symbol, cancer_type, .keep_all=TRUE) %>%
    mutate(cancer_type = str_replace(cancer_type, " ", "_")) %>%
    mutate(prognostic = case_when(!is.na(prog_fav) ~ "fav",
                                  !is.na(prog_unfav) ~ "unfav",
                                  TRUE ~ "no")) %>%
    select(ens_gene_id, gene_symbol, cancer_type, prognostic, prog_fav, prog_unfav)
  return(x)
}


get_hpa_top_gene_sets <- function(tbl, padj_cutoff=1) {
  ntopgenes <- c(50, 100, 250, 500)
  gs <- list()

  for (catype in unique(tbl$cancer_type)) {
    fav_genes <- tbl %>% dplyr::filter(
      cancer_type == catype,
      prognostic == "fav",
      prog_fav <= padj_cutoff
    )
    unfav_genes <- tbl %>% dplyr::filter(
      cancer_type == catype,
      prognostic == "unfav",
      prog_unfav <= padj_cutoff
    )
    gs_name <- paste0("HPA_PROGNOSIS_", toupper(catype), "_UNFAV_ALL")
    gs[[gs_name]] <- pull(unfav_genes, gene_symbol)
    gs_name <- paste0("HPA_PROGNOSIS_", toupper(catype), "_FAV_ALL")
    gs[[gs_name]] <- pull(fav_genes, gene_symbol)
    
    for (ntop in ntopgenes) {
      x <- unfav_genes %>% slice_min(prog_unfav, n=ntop, na_rm=TRUE) %>% pull(gene_symbol)
      if (length(x) < ntop) { next }
      gs_name <- paste0("HPA_PROGNOSIS_", toupper(catype), "_UNFAV_TOP_", ntop)
      gs[[gs_name]] <- x

      x <- fav_genes %>% slice_min(prog_fav, n=ntop, na_rm=TRUE) %>% pull(gene_symbol)
      if (length(x) < ntop) { next }
      gs_name <- paste0("HPA_PROGNOSIS_", toupper(catype), "_FAV_TOP_", ntop)
      gs[[gs_name]] <- x
    }
  }
  return(gs)
}

read_msigdb_json_file <- function(filename) {
  x <- jsonlite::fromJSON(filename)
  y <- tibble(gs_name = names(x), gs = x) %>% 
    unnest_wider(gs) %>%
    unnest_longer(geneSymbols) %>%
    select(gs_cat = collection,
           gs_name = gs_name,
           gene_symbol = geneSymbols)
}

# read cptac data
cptac_de_data <- read_cptac_pdac_data(cptac_panc_file, cptac_log2fc_cutoff, cptac_padj_cutoff)
# pivot to merge protein/rna data
cptac_de_merged <- cptac_de_data %>%
  pivot_wider(names_from = analysis, values_from = c(log2fc, pval, fdr, de))

# read hpa data
hpa <- read_hpa_data(hpa_path_file)
hpa_pdac <- filter(hpa, cancer_type == "pancreatic_cancer")

# read msigdb data
f <- file.path(data_dir, "msigdb.v2023.2.Hs.json")
msigdb_gene_sets <- read_msigdb_json_file(f)

```


# QC Duke IPMN dataset

```{r qc duke dataset}

ds <- duke_ipmn_ds
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)
# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)

```

## QC plots all AOIs

```{r all segments qc, include=FALSE}

nrow(ds$samples)
table(ds$samples$subset)
# nuclei count
summary(ds$samples$AOINucleiCount)
# read counts
summary(ds$samples$num_counts)

# figure: scatter plot of nuclei count vs total counts
x <- cor.test(ds$samples$AOINucleiCount, ds$samples$num_counts)
corlab <- paste0("r = ", round(x$estimate, 2))
p <- ggplot(ds$samples, aes(AOINucleiCount, num_counts, color=path)) + 
  geom_point(aes(shape=segment)) +
  geom_smooth(method = 'lm', 
              formula = y ~ x, 
              se = FALSE, 
              color = "black", 
              linetype = "dashed") +
  scale_x_log10() +
  scale_y_log10() +
  scale_color_manual(values = color_scales$path) +
  scale_shape_manual(values = shape_scales$segment) +
  labs(x="Nuclei Count", y="Total Counts") +
  annotate("text", x=500, y=1e4, label = corlab) +
  theme_minimal() +
  theme(aspect.ratio=1)
p
f <- file.path(plot_dir, "qc_duke_scatter_nuclei_vs_counts.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 3)

# figure: scatter plot of surface area versus total counts
x <- cor.test(ds$samples$AOISurfaceArea, ds$samples$num_counts)
corlab <- paste0("r = ", round(x$estimate, 2))
p <- ggplot(ds$samples, aes(AOISurfaceArea, num_counts, color=path)) + 
  geom_point(aes(shape=segment)) +
  geom_smooth(method = 'lm', 
              formula = y ~ x, 
              se = FALSE, 
              color = "black", 
              linetype = "dashed") +
  scale_x_log10() +
  scale_y_log10() +
  scale_color_manual(values = color_scales$path) +
  scale_shape_manual(values = shape_scales$segment) +
  labs(x="Surface Area", y="Total Counts") +
  annotate("text", x=1e5, y=1e4, label = corlab) +
  theme_minimal() +
  theme(aspect.ratio=1)
p
f <- file.path(plot_dir, "qc_duke_scatter_surfacearea_vs_counts.pdf")
ggsave(f, plot=p, device="pdf", width = 5, height = 3)


# figure: boxplot of segment nuclei count
p1 <- ggplot(ds$samples, aes(segment, AOINucleiCount)) +
  geom_boxplot(outlier.size=0, width=0.5) + 
  geom_point(position=position_jitter(width=0.1)) + 
  scale_y_log10() +
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) + 
  labs(x="Segment", y="Nuclei Count")

# figure: boxplot of segment num counts
p2 <- ggplot(ds$samples, aes(segment, num_counts)) +
  geom_boxplot(outlier.size=0, width=0.5) + 
  geom_point(position=position_jitter(width=0.1)) + 
  geom_hline(yintercept = geomx_min_counts, linetype="dashed", color="red") +
  scale_y_log10() +
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) +
  labs(x="Segment", y="Total Counts")

# figure: boxplot of segment bg_auc
p3 <- ggplot(ds$samples, aes(segment, bg_auc)) +
  geom_boxplot(outlier.size=0, width=0.5) + 
  geom_point(position=position_jitter(width=0.1)) + 
  geom_hline(yintercept = geomx_min_auc, linetype="dashed", color="red") +
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) +
  labs(x="Segment", y="snAUC")
p <- p1 + p2 + p3
f <- file.path(plot_dir, "qc_duke_boxplots_segment.pdf")
ggsave(f, plot=p, device="pdf", width = 6, height = 3)

```

## QC plots epithelial AOIs

```{r qc epithelial AOIs, include=FALSE}

s <- ds$samples %>% filter(segment == "panck")

# figure: boxplot nuclei count
p1 <- ggplot(s, aes(path, AOINucleiCount, fill=path)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(position=position_jitter(width=0.1)) + 
  scale_y_log10() +
  scale_fill_manual(values = color_scales$path) +
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) +
  theme(legend.text = element_text(size=7)) +
  labs(x="Path", y="Nuclei Count")

# figure: boxplot path counts
p2 <- ggplot(s, aes(path, num_counts, fill=path)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(position=position_jitter(width=0.1)) + 
  geom_hline(yintercept = geomx_min_counts, linetype="dashed", color="red") +
  scale_y_log10() +
  scale_fill_manual(values = color_scales$path) +
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) +
  theme(legend.text = element_text(size=7)) +
  labs(x="Path", y="UMI Counts")

# figure: boxplot path vs auc
p3 <- ggplot(s, aes(path, bg_auc, fill=path)) +
  geom_boxplot(outlier.shape = NA, width=0.75) +
  geom_point(position=position_jitter(width=0.1)) + 
  geom_hline(yintercept = geomx_min_auc, linetype="dashed", color="red") +
  scale_fill_manual(values = color_scales$path) +
  theme_minimal() +
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust=0.5, hjust=1, size=6)) +
  theme(legend.text = element_text(size=7)) +
  labs(x="Path", y="snAUC")

p <- p1 + p2 + p3
f <- file.path(plot_dir, "qc_duke_boxplots_epithelial_path.pdf")
ggsave(f, plot=p, device="pdf", width = 9, height = 3)

```

## QC Duke  AOIs

```{r aoi qc}

# epithelial cells
x <- ds$samples %>% filter(segment == "panck")
x <- stgeomx_init(samples = x,
                   meta = ds$meta, 
                   counts = ds$counts[, x$aoi],
                   x = ds$counts[, x$aoi])
table(x$samples$keep)
sum(x$samples$keep) / nrow(x$samples)
p1 <- stgeomx_plot_aoi_filter(x, 
                              min_counts=geomx_min_counts,
                              min_auc=geomx_min_auc,
                              min_frac_expr=geomx_aoi_min_frac_expr)

# stromal cells
x <- ds$samples %>% filter(segment == "otherstroma")
x <- stgeomx_init(samples = x,
                   meta = ds$meta, 
                   counts = ds$counts[, x$aoi],
                   x = ds$counts[, x$aoi])
table(x$samples$keep)
sum(x$samples$keep) / nrow(x$samples)
p2 <- stgeomx_plot_aoi_filter(x, 
                              min_counts=geomx_min_counts,
                              min_auc=geomx_min_auc,
                              min_frac_expr=geomx_aoi_min_frac_expr)

# immune cells
x <- ds$samples %>% filter(segment %in% c("cd3", "cd45"))
x <- stgeomx_init(samples = x,
                   meta = ds$meta, 
                   counts = ds$counts[, x$aoi],
                   x = ds$counts[, x$aoi])
table(x$samples$keep)
sum(x$samples$keep) / nrow(x$samples)
p3 <- stgeomx_plot_aoi_filter(x, 
                              min_counts=geomx_min_counts,
                              min_auc=geomx_min_auc,
                              min_frac_expr=geomx_aoi_min_frac_expr)

p <- p1 / p2 / p3
f <- file.path(plot_dir, "qc_duke_aoi_filtering_by_segment.pdf")
ggsave(f, plot=p, device="pdf", width = 8, height = 8)


```

# Duke stromal (PanCK-) dataset

```{r duke stromal dataset, include=FALSE}

# create duke stromal dataset
ds <- duke_ipmn_ds
s <- ds$samples %>% filter(segment == "otherstroma")
ds <- stgeomx_init(samples=s,
                   meta=ds$meta,
                   counts=ds$counts[, s$aoi],
                   x=ds$counts[, s$aoi])
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)
# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)
duke_ipmn_stroma_meta <- ds$meta

# gene filtering
p <- stgeomx_plot_gene_filter(ds, geomx_gene_min_frac_expr)
p
f <- file.path(plot_dir, "qc_duke_stroma_gene_filtering")
w <- 7
h <- 7
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

# background correction must use original unfiltered matrix
fds <- stgeomx_bgcorrect(ds, method=bgcorrect_method, 
                         bw.adjust=bgcorrect_bw_adjust, 
                         bg.quant=bgcorrect_bg_quant)
# normalize
fds <- stgeomx_normalize(fds, norm_method)

# figure: background correction
p <- stgeomx_plot_bgcorrect(fds)
w <- 8
h <- 5
f <- file.path(plot_dir, "qc_duke_stroma_bgcorrect_norm")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

# background correction must use original unfiltered matrix
fds <- stgeomx_bgcorrect(ds, method=bgcorrect_method,
                         bw.adjust=bgcorrect_bw_adjust,
                         bg.quant=bgcorrect_bg_quant)
# apply filters before normalizing
fds <- stgeomx_apply_filters(fds)
# normalize
fds <- stgeomx_normalize(fds, norm_method)

# filtering stats
nrow(ds$samples)
nrow(fds$samples)
table(ds$meta$keep)

# roi filtering per study
table(ds$samples$study, ds$samples$keep)
table(ds$samples$study, ds$samples$keep) / rowSums(table(ds$samples$study, ds$samples$keep))

# normalized dataset
fds_duke_stroma <- fds
fds_duke_stroma$log2_norm_cpm <- log2(fds_duke_stroma$x + 1)

```

# Duke epithelial (PanCK+) dataset normalization

```{r duke epithelial, include=FALSE}

# create duke epithelial dataset
ds <- duke_ipmn_ds
s <- ds$samples %>% filter(segment == "panck")
ds <- stgeomx_init(samples=s,
                   meta=ds$meta,
                   counts=ds$counts[, s$aoi],
                   x=ds$counts[, s$aoi])
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)
# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)
duke_ipmn_panck_meta <- ds$meta

# gene filtering
p <- stgeomx_plot_gene_filter(ds, geomx_gene_min_frac_expr)
p
f <- file.path(plot_dir, "qc_duke_panck_gene_filtering")
w <- 7
h <- 7
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

# background correction must use original unfiltered matrix
fds <- stgeomx_bgcorrect(ds, method=bgcorrect_method, 
                         bw.adjust=bgcorrect_bw_adjust, 
                         bg.quant=bgcorrect_bg_quant)
# normalize
fds <- stgeomx_normalize(fds, norm_method)

# figure: background correction
p <- stgeomx_plot_bgcorrect(fds)
w <- 8
h <- 5
f <- file.path(plot_dir, "qc_duke_panck_bgcorrect_norm")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

```


# QC Carbone dataset

```{r cc dataset}

ds <- cc_ipmn_ds
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)

# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)
ds$samples <- ds$samples %>% mutate(
  path = paste0(path_specimen, "_", path_aoi)
)

# factors
path_levels <- c("ipmn_lgd_lgd", "ipmn_igd_lgd", "ipmn_igd_igd", "ipmn_hgd_hgd")
ds$samples$path <- factor(ds$samples$path, levels=path_levels)


p <- stgeomx_plot_aoi_filter(ds, 
                             min_counts=geomx_min_counts,
                             min_auc=geomx_min_auc,
                             min_frac_expr=geomx_aoi_min_frac_expr)
f <- file.path(plot_dir, "qc_cc_aoi_filtering")
ggsave(paste0(f, ".pdf"), plot=p, width = 8, height = 3)

p <- stgeomx_plot_gene_filter(ds, geomx_gene_min_frac_expr)
w <- 7
h <- 7
f <- file.path(plot_dir, "qc_cc_gene_filtering")
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

```

# QC Spatial Organ Atlas Dataset

```{r qc soa dataset, include=FALSE}

ds <- soa_panc_ds
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)

# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)

# epithelial
x <- ds$samples %>% filter(segment == "panck")
x <- stgeomx_init(samples = x,
                   meta = ds$meta, 
                   counts = ds$counts[, x$aoi],
                   x = ds$counts[, x$aoi])
table(x$samples$keep)
sum(x$samples$keep) / nrow(x$samples)
p <- stgeomx_plot_aoi_filter(x, 
                             min_counts=geomx_min_counts,
                             min_auc=geomx_min_auc,
                             min_frac_expr=geomx_aoi_min_frac_expr)
f <- file.path(plot_dir, "qc_soa_panck_aoi_filtering")
ggsave(paste0(f, ".pdf"), plot=p, width = 8, height = 3)


```

# QC UM GoL study

```{r qc umich gift of life dataset, include=FALSE}

ds <- gol_panc_panck_ds
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)

# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)

table(ds$samples$keep)
sum(ds$samples$keep) / nrow(ds$samples)
p <- stgeomx_plot_aoi_filter(ds,
                             min_counts=geomx_min_counts,
                             min_auc=geomx_min_auc,
                             min_frac_expr=geomx_aoi_min_frac_expr)
f <- file.path(plot_dir, "qc_gol_panck_aoi_filtering")
ggsave(paste0(f, ".pdf"), plot=p, width = 8, height = 3)

```


# Epithelial (PanCK+) study datasets

## Merged dataset

```{r merged geomx dataset, include=FALSE}

# merge datasets
ds <- duke_ipmn_ds %>%
  merge_ds(cc_ipmn_ds) %>%
  merge_ds(soa_panc_ds) %>% 
  merge_ds(gol_panc_panck_ds)

# ensure no missing probes
stopifnot(!apply(ds$counts, 1, function(x) any(is.na(x))))

# filter epithelial aois from merged full dataset
s <- ds$samples %>% filter(segment == "panck")
ds <- stgeomx_init(samples=s,
                   meta=ds$meta,
                   counts=ds$counts[, s$aoi],
                   x=ds$counts[, s$aoi])

# augment sample annotation
ds$samples <- ds$samples %>% mutate(
  path = paste0(path_specimen, "_", path_aoi),
)

path_levels <- c("donor_duct", "donor_panin", "pnet_duct", 
                 "ipmn_lg_nl", "ipmn_lg_lgd", 
                 "ipmn_ig_lgd", "ipmn_ig_igd", "ipmn_hg_hgd",
                 "ipmn_pdac_nl", "ipmn_pdac_lgd", "ipmn_pdac_hgd", "ipmn_pdac_inv",
                 "pdac_duct", "pdac_panin", "pdac_pdac")
# factors
ds$samples$path <- factor(ds$samples$path, levels=path_levels)

# process dataset
ds <- stgeomx_preprocess(ds, geomx_negprobe_lod_quantile)
ds <- stgeomx_merge_probes(ds)

# set filtering thresholds
ds <- stgeomx_set_thresholds(ds,
                             min_counts = geomx_min_counts,
                             min_auc = geomx_min_auc,
                             aoi_min_frac_expr = geomx_aoi_min_frac_expr,
                             gene_min_frac_expr = geomx_gene_min_frac_expr)

# override gene metadata with duke metadata so that external datasets are not
# a factor in determining which genes are expressed/filtered
stopifnot(all(ds$meta$gene == duke_ipmn_panck_meta$gene))
ds$meta <- duke_ipmn_panck_meta


# background correction must use original unfiltered matrix
fds <- stgeomx_bgcorrect(ds, method=bgcorrect_method, 
                         bw.adjust=bgcorrect_bw_adjust, 
                         bg.quant=bgcorrect_bg_quant)
# normalize
fds <- stgeomx_normalize(fds, norm_method)

#
# background correction/normalization
#
p <- stgeomx_plot_bgcorrect(fds)
w <- 8
h <- 5
f <- file.path(plot_dir, "qc_panck_bgcorrect_norm")
ggsave(paste0(f, ".png"), p, width=w, height=h)
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

# background correction must use original unfiltered matrix
fds <- stgeomx_bgcorrect(ds, method=bgcorrect_method,
                         bw.adjust=bgcorrect_bw_adjust,
                         bg.quant=bgcorrect_bg_quant)
# apply filters before normalizing
fds <- stgeomx_apply_filters(fds)
# normalize
fds <- stgeomx_normalize(fds, norm_method)

# filtering stats
nrow(ds$samples)
nrow(fds$samples)
table(ds$meta$keep)

# roi filtering per study
table(ds$samples$study, ds$samples$keep)
table(ds$samples$study, ds$samples$keep) / rowSums(table(ds$samples$study, ds$samples$keep))

# normalized dataset
fds_panck <- fds
fds_panck$log2_norm_cpm <- log2(fds$x + 1)

```



## Study specific normalized datasets

```{r study specific normalized datasets}

# duke
s <- fds_panck$samples %>% filter(study == "duke")
fds_duke_panck <- stgeomx_init(samples = s,
                               meta = fds_panck$meta,
                               counts = fds_panck$counts[, s$aoi],
                               x = fds_panck$x[, s$aoi])
# final normalized gene expression matrix
fds_duke_panck$log2_norm_cpm <- log2(fds_duke_panck$x + 1)

# cc
s <- fds_panck$samples %>% filter(study == "cc")
fds_cc_panck <- stgeomx_init(samples = s,
                             meta = fds_panck$meta,
                             counts = fds_panck$counts[, s$aoi],
                             x = fds_panck$x[, s$aoi])
fds_cc_panck$log2_norm_cpm <- log2(fds_cc_panck$x + 1)

# spatial organ atlas
s <- fds_panck$samples %>% filter(study == "soa")
fds_soa_panck <- stgeomx_init(samples = s,
                              meta = fds_panck$meta,
                              counts = fds_panck$counts[, s$aoi],
                              x = fds_panck$x[, s$aoi])
fds_soa_panck$log2_norm_cpm <- log2(fds_soa_panck$x + 1)

# umich gol
s <- fds_panck$samples %>% filter(study == "gol")
fds_gol_panck <- stgeomx_init(samples = s,
                              meta = fds_panck$meta,
                              counts = fds_panck$counts[, s$aoi],
                              x = fds_panck$x[, s$aoi])
fds_gol_panck$log2_norm_cpm <- log2(fds_gol_panck$x + 1)

```



# Duke IPMN epithelial analysis

## Color scales for patients

```{r duke color scales}

# set patient color scales
x <- pals::cols25(n_distinct(fds_duke_panck$samples$patient))
color_scales$patient = setNames(x, sort(unique(fds_duke_panck$samples$patient)))

```

## Filter genes in external gene sets

```{r filter genes in external datasets}

# define gene universe
gene_universe <- fds_duke_panck$meta$gene

# cptac: get gene sets after filtering gene universe
cptac_de_data <- filter(cptac_de_data, gene_symbol %in% gene_universe)
gs_cptac <- get_top_gene_sets(cptac_de_data, cptac_log2fc_cutoff, cptac_padj_cutoff)

# hpa: get gene sets after filtering gene universe
hpa <- filter(hpa, gene_symbol %in% gene_universe)
gs_hpa <- get_hpa_top_gene_sets(hpa)
gs_hpa_panc <- get_hpa_top_gene_sets(hpa_pdac)

# msigdb: get gene sets after filtering gene universe
msigdb_gene_sets <- msigdb_gene_sets %>%
  dplyr::filter(gene_symbol %in% gene_universe) %>%
  distinct(gs_cat, gs_name, gene_symbol, .keep_all = TRUE)

# gene set categories
gs_hallmark <- dplyr::filter(msigdb_gene_sets, gs_cat == "H")
gs_hallmark <- split(x = gs_hallmark$gene_symbol, f = gs_hallmark$gs_name)

# pancreatic cancer subtypes gene lists
x <- readRDS(pdacr_gene_list_file)
gs_pdacr <- list()
for (gs_name in names(x)) {
  genes <- x[[gs_name]]
  if (typeof(genes) == "character") {
    genes <- intersect(x[[gs_name]], gene_universe)
    gs_pdacr[[gs_name]] <- genes
  } else {
    print(paste0("error reading gene set ", gs_name))
  }
}

# define gene sets used in analysis
moffitt_pathways <- c("Moffitt.F3_Exocrine.top250", 
                      "Moffitt.F8_Classical.top250",
                      "Moffitt.F6_BasalLike.top250")
collison_pathways <- c("Collisson.Exocrine",
                       "Collisson.Classical", 
                       "Collisson.QM")
icgc_pathways <- c("ICGC.ADEX.Down", 
                   "ICGC.ADEX.Up",
                   "ICGC.Immunogenic.Down",
                   "ICGC.Immunogenic.Up",
                   "ICGC.Progenitor.Down",
                   "ICGC.Progenitor.Up",
                   "ICGC.Squamous.Down",
                   "ICGC.Squamous.Up")
chan_seng_yue_pathways <- c("Chan_Seng_Yue.ClassicalA",
                            "Chan_Seng_Yue.ClassicalB", 
                            "Chan_Seng_Yue.BasalA", 
                            "Chan_Seng_Yue.BasalB")
pdacr_pathways <- c(moffitt_pathways, collison_pathways, icgc_pathways, chan_seng_yue_pathways)

```


## Highly variable genes

```{r highly variable genes, include=FALSE}

ntop <- 500
nlabel <- 100
s <- fds_duke_panck$samples
y <- fds_duke_panck$log2_norm_cpm

# maplot
p <- maplot(y, ntop=ntop, nlabel=nlabel, myspan=loess_span)
f <- file.path(plot_dir, paste0("maplot"))
ggsave(paste0(f, ".pdf"), plot=p, width=10, height=10)
ggsave(paste0(f, ".png"), plot=p, width=10, height=10)

# select most variable genes
y <- most_var_genes(y, ntop=ntop, span=loess_span)

# heatmap
f <- file.path(plot_dir, "heatmap_highvar_genes.pdf")
p <- heatmap_genes(s, y, f)

```

## PCA

```{r pca}

# subset most variable genes
s <- fds_duke_panck$samples
y <- most_var_genes(fds_duke_panck$log2_norm_cpm, ntop=ntop, span=loess_span)
hvgenes <- rownames(y)

# plot dimensions
width <- 4
height <- 4

# PCA
pca_res <- prcomp(t(y), center=TRUE, scale.=FALSE)
pca_res.var <- round(100 * pca_res$sdev^2 / sum( pca_res$sdev^2 ), 2)
pca_res.var_txt <- paste(colnames(pca_res$x), " (", paste(as.character(pca_res.var), "%", ")", sep=""), sep="")

# fviz plots
fviz_ntop <- 50

# percent variance of each PC
p <- fviz_eig(pca_res, col.var="blue", addlabels=TRUE)
f <- file.path(plot_dir, paste0("pca_duke_fviz_eig.pdf"))
ggsave(f, p, width=4, height=4)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  pca_res,
  axes = c(1,2),
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("pca_duke_fviz_pca_var_pc12.pdf"))
ggsave(f, p, width=7, height=7)


#
# PCA plots
#
custom_theme_pca <- function() {
  theme_bw() +
  theme(aspect.ratio=1) +
    theme(legend.title = element_text(size=8), 
          legend.text = element_text(size=7))
}

x <- bind_cols(s, as_tibble(pca_res$x))

# bg_ks_dist
p <- ggplot(x, aes(x=PC1, y=PC2, color=bg_auc)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_viridis_c() +
  custom_theme_pca() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_duke_bg_auc.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# patient
p <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values = color_scales$patient) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_duke_patient.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path specimen
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values=color_scales$path_specimen) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_duke_path_specimen.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path aoi
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_duke_path_aoi.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path
p <- ggplot(x, aes(x=PC1, y=PC2, color=path, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values=color_scales$path) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])
f <- file.path(plot_dir, paste0("pca_duke_path.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

```


### PCA clustering

```{r pca clustering}
# the following guides were used in this section:
# guide: https://rpubs.com/KarolinaSzczesna/862710
# guide: https://rpubs.com/Bury/ClusteringOnPcaResults

# npcs
npcs <- 2
nclusters <- 3

# assess optimal number of clusters
pca_tbl <- pca_res$x[, 1:npcs]
p <- fviz_nbclust(pca_tbl, FUNcluster=cluster::pam, method="silhouette")
f <- file.path(plot_dir, paste0("pca_duke_fviz_nbclust_silhouette.pdf"))
ggsave(f, plot=p, device="pdf", width=4, height=4)

# pam
pamx <- cluster::pam(dist(pca_tbl), k=nclusters)
memb <- paste0("c", pamx$clustering)

# add cluster membership to sample properties
fds_duke_panck$samples$cluster <- memb

# plot clustering
x <- bind_cols(fds_duke_panck$samples, as_tibble(pca_tbl))

# pca plot clustering
p <- ggplot(x, aes(x=PC1, y=PC2, color=cluster, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values=color_scales$cluster) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()
f <- file.path(plot_dir, paste0("pca_duke_cluster_unidentified.pdf"))
ggsave(f, plot=p, device="pdf", width=6, height=6)

#
# identify clusters based on pca relationships
#
label_clusters <- function(s, memb) {
  s <- s %>% mutate(
    cluster_name = case_match(memb, "c1" ~ "cLR", "c2" ~ "cNL", "c3" ~ "cHR")
  )
  return(s)
}

fds_duke_panck$samples <- label_clusters(fds_duke_panck$samples, memb)
x <- bind_cols(fds_duke_panck$samples, as_tibble(pca_tbl))

#
# single multipanel plot for figure
# 
# path specimen
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

# path aoi
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

# patient
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = pals::cols25()) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

# cluster
p4 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster_name) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=pca_res.var_txt[1], y=pca_res.var_txt[2])

p <- p1 + p2 + p3 + p4 + plot_layout(nrow=1, guides="collect")
f <- file.path(plot_dir, paste0("pca_duke_panel_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=16, height=12)

# examine clustering properties
table(fds_duke_panck$samples$path_specimen, fds_duke_panck$samples$cluster_name)
table(fds_duke_panck$samples$path_aoi, fds_duke_panck$samples$cluster_name)

# test for cluster enrichment
x <- table(fds_duke_panck$samples$path_aoi, fds_duke_panck$samples$cluster_name)
chisq <- chisq.test(x)

```


### PCA validation

```{r pca validation}

predict_pam <- function(pam_res, traindata, newdata, metric="euclidean") {
  nclus <- length(pam_res$id.med)
  d = cluster::daisy(rbind(traindata[pam_res$id.med,], newdata), metric=metric)
  max.col(-as.matrix(d)[-c(1:nclus),1:nclus])
}

# merged epithelial (panck+) data
s <- fds_panck$samples
# get matrix of highly variable genes
y <- fds_panck$log2_norm_cpm[hvgenes,]

# predict external datasets
pca_tbl.validation <- predict(pca_res, t(y))[,1:npcs]

# assign clusters
memb <- predict_pam(pamx, pca_tbl, pca_tbl.validation)
memb <- paste0("c", memb)
# identify/label clusters
fds_panck$samples$cluster <- memb
fds_panck$samples <- label_clusters(fds_panck$samples, memb)


# 
# plots for CC IPMN and SOA datasets
#
tbl <- bind_cols(fds_panck$samples, as_tibble(pca_tbl.validation))

x <- tbl %>% filter(study %in% c("duke", "cc", "soa"))
p1 <- ggplot(x, aes(x=PC1, y=PC2, shape=study)) +
  geom_point(data=filter(x, study == "duke"), size=2, alpha=0.3) +
  geom_point(data=filter(x, study != "duke"), aes(color=study), size=3, alpha=0.8) +
  scale_color_manual(values = color_scales$study) +
  scale_shape_manual(values = shape_scales$study) +
  custom_theme_pca()

# bg_auc
p2 <- ggplot(x, aes(x=PC1, y=PC2, shape=study)) +
  geom_point(data=filter(x, study == "duke"), color="black", size=2, alpha=0.3) +
  geom_point(data=filter(x, study != "duke"), aes(color=bg_auc), size=3, alpha=0.8) +
  scale_color_viridis_c() +
  scale_shape_manual(values = shape_scales$study) +
  custom_theme_pca()

# patient
p3 <- ggplot(x, aes(x=PC1, y=PC2, shape=study)) +
  geom_point(data=filter(x, study == "duke"), color="black", size=2, alpha=0.3) +
  geom_point(data=filter(x, study != "duke"), aes(color=patient), size=3, alpha=0.8) +
  scale_color_manual(values = as.vector(pals::alphabet2())) +
  scale_shape_manual(values = shape_scales$study) +
  custom_theme_pca()

# path aoi
p4 <- ggplot(x, aes(x=PC1, y=PC2, shape=study)) +
  geom_point(data=filter(x, study == "duke"), color="black", size=2, alpha=0.3) +
  geom_point(data=filter(x, study != "duke"), aes(color=path_aoi), size=3, alpha=0.8) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values = shape_scales$study) +
  custom_theme_pca()

# cluster
p5 <- ggplot(x, aes(x=PC1, y=PC2, shape=study)) +
  geom_point(data=filter(x, study == "duke"), color="black", size=2, alpha=0.3) +
  geom_point(data=filter(x, study != "duke"), aes(color=cluster_name), size=3, alpha=0.8) +
  scale_color_manual(values = color_scales$cluster_name) +
  scale_shape_manual(values = shape_scales$study) +
  custom_theme_pca()

p <- p1 + p2 + p3 + p4 + p5 + plot_layout(nrow=1, guides="collect")
f <- file.path(plot_dir, paste0("pca_cc_soa_validation_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=20, height=12)



# 
# plots for gol study
#
x <- tbl %>% filter(study %in% c("duke", "gol"))

# GoL path aoi
p1 <- ggplot(x, aes(x=PC1, y=PC2, shape=path_specimen)) +
  geom_point(data=filter(x, study == "duke"), color="black", size=2, alpha=0.3) +
  geom_point(data=filter(x, study == "gol"), aes(color=path_aoi), size=3, alpha=0.8) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

# GoL cluster
p2 <- ggplot(x, aes(x=PC1, y=PC2, shape=path_specimen)) +
  geom_point(data=filter(x, study == "duke"), color="black", size=2, alpha=0.3) +
  geom_point(data=filter(x, study == "gol"), aes(color=cluster_name), size=3, alpha=0.8) +
  scale_color_manual(values = color_scales$cluster_name) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  custom_theme_pca()

p <- p1 + p2 + plot_layout(nrow=1, guides="collect")
f <- file.path(plot_dir, paste0("pca_gol_validation_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=10, height=12)


# cc clustering stats
x <- filter(fds_panck$samples, study == "cc")
table(x$cluster_name, x$path_aoi, x$path_specimen)
x <- table(x$path_aoi, x$cluster_name)
chisq <- chisq.test(x)
# just igd and hgd
x <- fds_panck$samples %>% filter(study == "cc", path_aoi %in% c("igd", "hgd"))
x <- table(x$path_aoi, x$cluster_name)
chisq <- chisq.test(x)


# gol clustering stats
x <- filter(fds_panck$samples, study == "gol")
table(x$cluster_name, x$path_aoi, x$path_specimen)

# gol donor pancreas specimens
x <- filter(fds_panck$samples, study == "gol")
chisq <- chisq.test(table(x$path_aoi, x$cluster_name))

x <- filter(fds_panck$samples, study == "gol", path_specimen == "donor")
chisq <- chisq.test(table(x$path_aoi, x$cluster_name))

table(filter(fds_panck$samples, study == "gol") %>% pull(path_aoi),
      filter(fds_panck$samples, study == "gol") %>% pull(cluster_name))

# test for cluster enrichment
x <- table(fds_duke_panck$samples$path_aoi, fds_duke_panck$samples$cluster_name)
chisq <- chisq.test(x)


table(pull(fds_panck$samples, study), pull(fds_panck$samples, cluster_name))


```

### Trajectory analysis

```{r slingshot trajectory inference}

# number of principal components to use for clustering
npcs.traj <- 2
# set number of clusters
nclusters.traj <- 4

# pam
pca_tbl.traj <- pca_res$x[, 1:npcs.traj]
pamx.traj <- cluster::pam(dist(pca_tbl.traj), k=nclusters.traj)
memb <- paste0("c", pamx.traj$clustering)

# add cluster membership to sample properties
s <- fds_duke_panck$samples
s$traj_clust <- memb

# plot clustering
x <- bind_cols(s, as_tibble(pca_tbl.traj))

# pca plot clustering
p <- ggplot(x, aes(x=PC1, y=PC2, color=traj_clust, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values=color_scales$traj_clust) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()
f <- file.path(plot_dir, paste0("pca_duke_traj_cluster_unidentified.pdf"))
ggsave(f, plot=p, device="pdf", width=6, height=6)

#
# identify clusters based on pca relationships
#
s <- s %>% mutate(
  traj_clust_name = case_match(memb, 
                               "c1" ~ "cTR", 
                               "c2" ~ "cLR", 
                               "c3" ~ "cNL",
                               "c4" ~ "cPDAC")
)
# store updated sample metadata
fds_duke_panck$samples <- s

# pca plot clustering
x <- bind_cols(s, as_tibble(pca_tbl.traj))
p <- ggplot(x, aes(x=PC1, y=PC2, color=traj_clust_name, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values=color_scales$traj_clust_name) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()
f <- file.path(plot_dir, paste0("pca_duke_traj_clust.pdf"))
ggsave(f, plot=p, device="pdf", width=4, height=3)

# examine clustering properties
table(s$path_specimen, s$traj_clust_name)
table(s$path_aoi, s$traj_clust_name)

# test for cluster enrichment
x <- table(s$path_aoi, s$traj_clust)
chisq <- chisq.test(x)
chisq

# run trajectory analysis
res <- slingshot(pca_tbl.traj, s$traj_clust_name, 
                 start.clus="cNL", 
                 end.clus="cPDAC")

x <- slingAvgPseudotime(res)
stopifnot(all(s$aoi == rownames(x)))
x <- bind_cols(s, pca_tbl.traj)
x <- x %>% add_column(pseudotime = slingAvgPseudotime(res))
curves <- slingCurves(res, as.df=TRUE)
mst <- slingMST(res, as.df = TRUE)

#
# plot trajectory analysis
#
p <- ggplot(x, aes(x=PC1, y=PC2, color=pseudotime)) + 
  geom_point(alpha=0.8, size=3) + 
  scale_color_viridis_c(option="turbo")

p <- p + geom_path(data = curves %>% arrange(Order),
                   aes(group=Lineage), 
                   color="black",
                   linewidth=1,
                   arrow=grid::arrow(length=grid::unit(0.1, "inches"), type="closed"))
p <- p + geom_point(data = mst, size = 3, color="black") + 
  geom_path(data = mst %>% arrange(Order),
            aes(group=Lineage), linewidth=1, 
            color="black") + 
  custom_theme_pca()

f <- file.path(plot_dir, "pca_duke_traj")
ggsave(paste0(f, ".pdf"), p)
ggsave(paste0(f, ".png"), p)


# path aoi
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

# cluster
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=traj_clust_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$traj_clust_name) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

p <- p2 + p3 + p + plot_layout(guides="collect")
f <- file.path(plot_dir, paste0("pca_duke_traj_panel_plots"))
ggsave(paste0(f, ".pdf"), plot=p, width=12, height=12)
ggsave(paste0(f, ".png"), plot=p, width=12, height=12)


```


## Differential expression

### DE analysis

```{r}

run_limma_trend <- function(y, design, contrasts, padj_cutoff, log2fc_cutoff) {
  fit <- lmFit(y, design)
  fit <- contrasts.fit(fit, contrasts)
  fit <- eBayes(fit, trend=TRUE)

  x <- NULL
  for (coef in colnames(contrasts)) {
    res <- topTable(fit, coef=coef, number=Inf, sort.by="none")
    res <- as_tibble(res, rownames="gene")
    res <- res %>%
      select(gene,
             log2fc=logFC,
             avgexpr=AveExpr,
             pval=P.Value,
             padj=adj.P.Val) %>%
      mutate(de = case_when(padj > padj_cutoff ~ "no",
                            log2fc < -log2fc_cutoff ~ "dn",
                            log2fc > log2fc_cutoff ~ "up",
                            TRUE ~ "no"),
             contrast = coef)
    x <- bind_rows(x, res)
  }
  return(x)
}

# init de results
de <- NULL

#
# DE analysis of clusters
# 
s <- fds_duke_panck$samples
y <- fds_duke_panck$log2_norm_cpm

s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
s$path_specimen <- factor(s$path_specimen, levels=c("ipmn_lg", "ipmn_pdac"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient)
s$cluster <- factor(s$cluster_name, levels=c("cNL", "cLR", "cHR"))

design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(clusterHRvLR = clustercHR - clustercLR,
                           clusterHRvNL = clustercHR - clustercNL,
                           clusterLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de <- bind_rows(de, res)
# write cluster results
write_xlsx(res, file.path(results_dir, "de_clusters.xlsx"))


```

### DE Volcano plots

```{r de volcano plots}

plot_de_volcano <- function(res, nlabel=50) {
  # label the top n up/down genes
  label_data <- bind_rows(filter(res, de == "up") %>% slice_max(log2fc, n=nlabel),
                          filter(res, de == "dn") %>% slice_min(log2fc, n=nlabel))
  label_data <- distinct(label_data)

  p <- ggplot(res, aes(x=log2fc, y=-log10(padj), color=de, size=de)) +
    geom_point(alpha=0.7) +
    geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +
    scale_color_manual(values=c("no"="grey", "dn"="blue", "up"="red")) +
    scale_size_manual(values=c("no"=1, "dn"=2, "up"=2)) +
    theme_minimal() +
    theme(legend.position="bottom") +
    theme(axis.line = element_line(color = "black")) +
    labs(x="log2fc", y="-log10(padj)")
  return(p)
}

# volcano plots
for (mycontrast in unique(de$contrast)) {
  print(mycontrast)
  x <- filter(de, contrast == mycontrast)
  p <- plot_de_volcano(x) + 
    labs(title="Volcano Plot", subtitle=mycontrast)
  f <- paste0("volcano_de_", mycontrast)
  ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=8)
  ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=8)
}

```

### DE Heatmaps

```{r de heatmaps, include=FALSE}

# heatmap functions
plot_pheatmap <- function(m, 
                          annot_col, 
                          annot_row, 
                          annot_colors,
                          filename, width, height,
                          xmin=-6, xmax=6) {
  pheatmap(m,
           scale = "none",
           color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
           breaks = seq(xmin, xmax, length.out=51),
           annotation_col = annot_col,
           annotation_row = annot_row,
           annotation_colors = annot_colors,
           border_color = NA,
           cluster_rows = TRUE,
           cluster_cols = TRUE,
           clustering_distance_rows = "euclidean",
           clustering_distance_cols = "euclidean",
           clustering_method = "ward.D2",
#           cellwidth = 6,
#           cellheight = 6,
           fontsize = 6,
           fontsize_row = 6,
           fontsize_col = 6,
           filename = filename, 
           width = width,
           height = height)
}

# setup heatmap annotations
gene_annot <- fds_duke_panck$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot) %>%
  column_to_rownames("gene") %>%
  as.data.frame()
annot_col <- column_to_rownames(fds_duke_panck$samples, "aoi") %>% 
  select(patient, path_aoi, path_specimen, cluster_name) %>%
  as.data.frame()

#
# IPMN clusters
#
# de gene params
padj_cutoff <- 0.01
log2fc_cutoff <- 3
ntop <- 10
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

g <- filter(de, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)
cluster_marker_genes <- g

# centered gene expression (subtract median)
s <- fds_duke_panck$samples
x <- fds_duke_panck$log2_norm_cpm[g, s$aoi]
u <- apply(x, 1, mean)
x <- sweep(x, 1, u, FUN="-")
xmin <- -6
xmax <- 6

# heatmap
f <- file.path(plot_dir,  paste0("heatmap_de_cluster_markers_", ngenes, "_genes"))
h <- 6
w <- 6
p <- plot_pheatmap(x, 
                   annot_col[s$aoi,], 
                   annot_row[g,], 
                   annot_colors=color_scales,
                   filename=paste0(f, ".png"), 
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


```


### PCA feature plots for marker genes

```{r pca feature plot for each marker gene, include=FALSE}

s <- fds_duke_panck$samples
y <- fds_duke_panck$log2_norm_cpm
feature_plot_genes <- c("CTRB1", "CRISP3", "TRIM29", "PGC")

for (gene in feature_plot_genes) {
  print(gene)
  x <- bind_cols(s, as_tibble(pca_res$x), gene=unlist(y[gene,]))
  p <- ggplot(x, aes(x=PC1, y=PC2, color=gene, shape=path_specimen)) +
    geom_point(alpha = 0.8, size=3) +
    scale_color_viridis_c() +
    scale_shape_manual(values = shape_scales$path_specimen) +
    theme_minimal() +
    theme(aspect.ratio=1) +
    labs(x="PC1", y="PC2", title=gene)
  f <- file.path(plot_dir, paste0("pca_featureplot_", gene))
  ggsave(paste0(f, ".pdf"), plot=p, width=4, height=3)
  ggsave(paste0(f, ".png"), plot=p, width=4, height=3)
}

```


### DE 2D scatter plots

```{r ipmn de 2d scatter plots}

plot_de_scatter2d <- function(tbl, xcontrast, ycontrast, nlabel=10, 
                              highlight_genes=NULL) {
  if (is.null(highlight_genes)) {
    highlight_genes <- c()
  }
  xde <- sym(paste0("de_", xcontrast))
  yde <- sym(paste0("de_", ycontrast))
  xlog2fc <- sym(paste0("log2fc_", xcontrast))
  ylog2fc <- sym(paste0("log2fc_", ycontrast))

  x <- tbl %>% 
    mutate(subtype = paste0("x", {{xde}}, "_", "y", {{yde}}),
           log2fc_abs_x = case_when({{xde}} == "no" ~ 0,
                                    TRUE ~ abs({{xlog2fc}})),
           log2fc_abs_y = case_when({{yde}} == "no" ~ 0,
                                    TRUE ~ abs({{ylog2fc}})),
           log2fc_abs_max = pmax(log2fc_abs_x, log2fc_abs_y))
  
  color_scale = c(xno_yno = "#888888",
                  xup_yup = "#FF0000",
                  xdn_ydn = "#0000aa",
                  xup_ydn = "#eeee00",
                  xdn_yup = "#ff00ff",
                  xup_yno = "#32cd32",
                  xdn_yno = "#670092",
                  xno_yup = "#ff6700",
                  xno_ydn = "#16ccc4")
  size_scale = c(xno_yno = 1, 
                 xup_yup = 2,
                 xdn_ydn = 2,
                 xup_ydn = 2,
                 xdn_yup = 2,
                 xup_yno = 2,
                 xdn_yno = 2,
                 xno_yup = 2,
                 xno_ydn = 2)
  
  annot_data <- filter(x, subtype != "xno_yno")
  none_data <- filter(x, subtype == "xno_yno")
  label_data <- annot_data %>% 
    group_by(subtype) %>%
    slice_max(log2fc_abs_max, n=nlabel) %>%
    ungroup()
  highlight_data <- filter(x, gene %in% highlight_genes)

  p <- ggplot(x, aes(x={{xlog2fc}}, 
                     y={{ylog2fc}},
                     color=subtype, 
                     size=subtype)) +
    geom_point(data=none_data, alpha=0.4) + 
    geom_point(data=annot_data, alpha=0.8) +
    geom_text_repel(data=label_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) + 
    geom_point(data=highlight_data, pch=21, fill=NA, size=5, colour="#ffff00", stroke=1) + 
    geom_point(data=highlight_data, pch=21, fill=NA, size=4, colour="#222222", stroke=1, alpha=0.5) +
    geom_text_repel(data=highlight_data, color="black", size=3, aes(label=gene), max.overlaps=Inf) +     
    scale_color_manual(values = color_scale) +
    scale_size_manual(values = size_scale, guide = "none") + 
    theme_minimal() +
    xlab(rlang::as_string(xlog2fc)) +
    ylab(rlang::as_string(ylog2fc)) +
    labs(color = "DE Genes")
  return(p)
}


# merge de analysis by gene
de_merged <- de %>% select(gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))
nlabel <- 10
width <- 8
height <- 8

#
# IPMN Clusters
#
p <- plot_de_scatter2d(de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterHRvLR", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_clusterHRvNL_y_clusterHRvLR")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterLRvNL", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterLRvNL (y-axis)")
f <- file.path(plot_dir, "scatter2d_de_x_clusterHRvNL_y_clusterLRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

```


### DE GoL Validation

```{r de gol validation}

#
# DE analysis of clusters
# 
s <- fds_panck$samples %>% filter(study == "gol")
y <- fds_panck$log2_norm_cpm[, s$aoi]

s$cluster <- factor(s$cluster_name, levels=c("cNL", "cLR", "cHR"))

design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(clusterHRvLR = clustercHR - clustercLR,
                           clusterHRvNL = clustercHR - clustercNL,
                           clusterLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de_gol <- res

# write cluster results
write_xlsx(res, file.path(results_dir, "de_clusters_gol.xlsx"))




# volcano plots
for (mycontrast in unique(de_gol$contrast)) {
  print(mycontrast)
  x <- filter(de_gol, contrast == mycontrast)
  p <- plot_de_volcano(x) + 
    labs(title="Volcano Plot", subtitle=mycontrast)
  f <- paste0("gol_volcano_de_", mycontrast)
  ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=8)
  ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=8)
}




# setup heatmap annotations
gene_annot <- fds_panck$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot) %>%
  column_to_rownames("gene") %>%
  as.data.frame()
annot_col <- column_to_rownames(fds_panck$samples, "aoi") %>% 
  select(patient, path_aoi, path_specimen, cluster_name) %>%
  as.data.frame()

# de gene params
padj_cutoff <- 0.01
log2fc_cutoff <- 3
ntop <- 10
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

g <- filter(de_gol, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)
gol_cluster_marker_genes <- g

# centered gene expression (subtract median)
s <- fds_panck$samples %>% filter(study == "gol")
x <- fds_panck$log2_norm_cpm[g, s$aoi]
u <- apply(x, 1, mean)
x <- sweep(x, 1, u, FUN="-")
xmin <- -6
xmax <- 6

# set patient color scales
gol_color_scales <- color_scales
gol_color_scales$patient = setNames(pals::cols25(n_distinct(s$patient)),
                                    sort(unique(s$patient)))


# heatmap
f <- file.path(plot_dir,  paste0("gol_heatmap_de_cluster_markers_", ngenes, "_genes"))
h <- 6
w <- 6
p <- plot_pheatmap(x, 
                   annot_col[s$aoi,], 
                   annot_row[g,], 
                   annot_colors=gol_color_scales,
                   filename=paste0(f, ".png"), 
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


# merge de analysis by gene
de_gol_merged <- de_gol %>% select(gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))
nlabel <- 10
width <- 8
height <- 8

#
# GoL scatter plots
#
p <- plot_de_scatter2d(de_gol_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterHRvLR", 
                       nlabel=nlabel) +
  labs(title="GoL Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "gol_scatter2d_de_x_clusterHRvNL_y_clusterHRvLR")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_gol_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterLRvNL", 
                       nlabel=nlabel) +
  labs(title="GoL Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterLRvNL (y-axis)")
f <- file.path(plot_dir, "gol_scatter2d_de_x_clusterHRvNL_y_clusterLRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


```



### DE plot validation genes

```{r de validation genes}

plot_gene_boxplot <- function(s, y, g) {
  box_width <- 0.7
  x <- s %>% add_column(gene = y[g,])
  p <- ggplot(x, aes(x=cluster_name, y=gene)) +
    geom_point(aes(color=path_aoi, shape=path_specimen), 
               position=position_jitter(width=0.1)) +
    geom_boxplot(outlier.size=0, width=box_width, alpha=0.3) + 
    scale_color_manual(values = color_scales$path_aoi) +
    #scale_fill_manual(values = color_scales$cluster_name) +
    scale_shape_manual(values = shape_scales$path_specimen) +
    theme_minimal() +
    theme(legend.title=element_text(size=8), 
          legend.text = element_text(size=7)) +
    facet_grid(~ study) + 
    labs(x="", y="log2 normalized expression", title=g)  
}

s <- fds_panck$samples
y <- fds_panck$log2_norm_cpm

# order studies
s <- s %>% mutate(
  study = factor(study, levels=c("duke", "soa", "cc", "gol"))
)


w <- 5
h <- 4
genes <- c("PGC", "CTRB1", "MUC6", "NKX6-2", "TRIM29", "KRT17", "CRISP3", "ANPEP",
           "S100A2", "AREG", "AHNAK2", "CLDN18", "ANXA10", "TFF1", "CLU")
for (g in genes) {
  print(g)
  p <- plot_gene_boxplot(s, y, g)
  f <- file.path(plot_dir, paste0("gene_boxplot_", g))
  ggsave(paste0(f, ".png"), p, width=w, height=h)
  ggsave(paste0(f, ".pdf"), p, width=w, height=h)
}

```


## GSEA


### GSEA write gene sets

```{r gsea write results}

# manuscript gene sets
gs <- c(gs_cptac, 
        gs_hpa_panc, 
        gs_pdacr[pdacr_pathways], 
        gs_hallmark)

# write gene sets to file
tbl <- NULL
for (gs_name in names(gs)) {
  x <- as_tibble_col(gs[[gs_name]], column_name="gene")
  x$gene_set <- gs_name
  tbl <- bind_rows(tbl, x)
}
write_xlsx(tbl, path=file.path(results_dir, "gene_sets.xlsx"))

```


### GSEA analysis

```{r gsea, include=FALSE}

get_de_ranks <- function(xde, mycontrast) {
  ranks <- xde %>% 
    dplyr::filter(contrast == mycontrast) %>%
    select(gene, log2fc, padj) %>%
    mutate(rank = log2fc)
  ranks = sort(setNames(ranks$rank, ranks$gene), decreasing = TRUE)
  return(ranks)
}

plot_gsea_enrichment <- function(x, ranks, a, gs) {
  txt_stat <- paste0("ES=", round(x$ES,2), " NES=", round(x$NES, 2), " padj=", format(x$padj, scientific=TRUE, digits=3))
  txt_title <- paste0("ranks: ", a, " gs: ", x$pathway)
  p <- plotEnrichment(gs[[x$pathway]], ranks) +
    annotate(geom="text", x=150, y=0.1, label=txt_stat, hjust=0) +
    labs(title = txt_title, 
         xlab = "Rank",
         ylab = "Enrichment Score") +
    theme(plot.title = element_text(size=8))
  return(p)
}

fgsea_run_batch <- function(xde, xgs, 
                            fgsea_eps=0,
                            fgsea_nperms=10000, 
                            fgsea_nproc=0) {
  fgsea_scoretypes <- c("pos", "neg")
  gsea <- NULL
  for (xcontrast in unique(xde$contrast)) {
    ranks <- get_de_ranks(xde, xcontrast)
    for (fgsea_scoretype in fgsea_scoretypes) {
      res <- fgsea(pathways=xgs, 
                   stats=ranks, 
                   eps=fgsea_eps, 
                   scoreType = fgsea_scoretype,
                   nPermSimple=fgsea_nperms, 
                   nproc=fgsea_nproc)
      res <- mutate(res, 
                    contrast = xcontrast, 
                    dir = fgsea_scoretype)
      gsea <- bind_rows(gsea, res)      
    }
  }
  return(gsea)
}


#
# Run full analysis
#
gsea <- NULL

# CPTAC
res <- fgsea_run_batch(de, gs_cptac, fgsea_eps=0, fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# HPA
res <- fgsea_run_batch(de, gs_hpa_panc, fgsea_eps=0, fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# pdacr
res <- fgsea_run_batch(de, gs_pdacr[pdacr_pathways], 
                       fgsea_eps=0, fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# hallmark
res <- fgsea_run_batch(de, gs_hallmark, 
                       fgsea_eps=0, fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
gsea_all <- gsea

```

### GSEA write gene sets

```{r}

# write results for clusters
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")
manuscript_pathways <- c(names(gs_cptac), names(gs_hpa_panc), 
                         pdacr_pathways, names(gs_hallmark))
res <- gsea_all %>% filter(contrast %in% contrasts, 
                           pathway %in% manuscript_pathways)

prefix <- "gsea_all"
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(res, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(res, f)

```

### GSEA plots

```{r gsea plots}

# contrasts
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")

# prepare gsea results
gsea <- gsea_all %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff) %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()

         
#
# CPTAC and HPA
#
cptac_pathways <- c("CPTAC_RNA_ALL_UP", "CPTAC_RNA_ALL_DN",
                    "CPTAC_PROT_ALL_UP", "CPTAC_PROT_ALL_DN")
hpa_pathways <- c("HPA_PROGNOSIS_PANCREATIC_CANCER_UNFAV_ALL",  
                  "HPA_PROGNOSIS_PANCREATIC_CANCER_FAV_ALL")
res <- gsea %>% filter(pathway %in% c(cptac_pathways, hpa_pathways))
p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_grid(~ pathway) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_barplot_clusters_cptac_hpa")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=5, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=5, height=4)



# 
# PDAC subtypes
#

#
# Moffitt (main figure)
#
res <- gsea %>% filter(pathway %in% moffitt_pathways)

# arrange for readability
pathway_levels <- res %>% filter(contrast == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=4)

#
# pdacr pathways (supplemental figure)
#
res <- gsea %>% filter(pathway %in% pdacr_pathways)
res$pathway <- factor(res$pathway, levels=pdacr_pathways)
# add author
res <- res %>% mutate(
  author = case_when(pathway %in% moffitt_pathways ~ "Moffitt",
                     pathway %in% collison_pathways ~ "Collison",
                     pathway %in% icgc_pathways ~ "ICGC",
                     pathway %in% chan_seng_yue_pathways ~ "Chan Seng Yue")
)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)") +
  facet_wrap(author ~ pathway, nrow=1)
f <- paste0("gsea_barplot_clusters_pdac_subtypes")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=16, height=5)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=16, height=5)



# 
# Hallmark
#
res <- gsea %>% filter(pathway %in% names(gs_hallmark)) %>% 
  mutate(pathway = gsub("HALLMARK_", "", pathway))
# only keep significant results (at least one contrast)
res <- res %>% group_by(pathway) %>%
  filter(any(padj < gsea_padj_cutoff)) %>%
  ungroup()

# arrange by clusterHRvLR for readability
pathway_levels <- filter(res, contrast == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=7)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ contrast, nrow=1) + 
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")
f <- paste0("gsea_barplot_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=3)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=3)


```


### GSEA LE pdac subtypes

```{r gsea leading edge}

gsea_leading_edge_matrix <- function(gsea_tbl,
                                     de_tbl,
                                     gsea_padj_cutoff=0.05,
                                     de_padj_cutoff=0.01,
                                     min_gene_sets=2,
                                     min_contrasts=2) {
  
  # get leading edge from significant gsea results
  x <- gsea_tbl %>% 
    dplyr::filter(padj < gsea_padj_cutoff) %>%
    unnest_longer(leadingEdge, values_to="gene")
  # join de table
  x <- left_join(x, de_tbl, by=join_by(gene==gene, contrast==contrast), suffix=c("",".de"))
  sig_de_genes <- x %>% 
    dplyr::filter(padj.de < de_padj_cutoff) %>%
    dplyr::distinct(gene) %>%
    pull()
  # keep genes that are significantly de
  x <- x %>% filter(gene %in% sig_de_genes)
  # get leading edge genes from significant gsea results
  x$nes_sign <- factor(sign(x$NES))

  # genes must be in leading edge of at least 'min_contrasts' (same direction)
  x <- x %>% 
    dplyr::group_by(gene, pathway, nes_sign) %>% 
    dplyr::filter(n() >= min_contrasts) %>%
    dplyr::ungroup() %>%
    mutate(contrast_pathway = paste0(contrast, ".", pathway))  

  # gsea result must be significant in at least "min_gene_sets" pathways
  leading_edge_genes <- x %>% 
    dplyr::group_by(contrast, nes_sign) %>%
    dplyr::count(gene) %>% 
    dplyr::filter(n >= min_gene_sets) %>% 
    dplyr::arrange(desc(n)) %>% 
    distinct(gene) %>% 
    pull()
  leading_edge_genes <- factor(leading_edge_genes)
  # filter results for leading edge genes
  x <- x %>% filter(gene %in% leading_edge_genes)
  
  # table with genes (rows) vs contrast-pathways (columns)
  x <- x %>% select(contrast_pathway, gene, log2fc) %>% 
    pivot_wider(id_cols=gene, names_from=contrast_pathway, values_from=log2fc)
  m <- x %>% column_to_rownames(var="gene")
  m[is.na(m)] <- 0
  return(m)
}

# setup heatmap annotations
gene_annot <- fds_duke_panck$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot) %>%
  column_to_rownames("gene") %>%
  as.data.frame()

# 
# Prepare gsea results
#
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")

gsea <- gsea_all %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff) %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()

#
# exocrine genes
#
pathways <- c(
  "Collisson.Exocrine",
  "ICGC.ADEX.Up",
  "Moffitt.F3_Exocrine.top250"
)

x <- filter(gsea, contrast %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix(x, de, gsea_padj_cutoff=gsea_padj_cutoff, 
                              min_gene_sets=2, min_contrasts=1)

# show intersection with CPTAC/prognosis
genes <- rownames(m)
length(genes)
table(annot_row[genes,]$de_cptac_rna)
table(annot_row[genes,]$de_cptac_prot)
table(annot_row[genes,]$de_cptac_rna, annot_row[genes,]$de_cptac_prot)
table(annot_row[genes,]$prognostic)
# agreement across studies
table(rowSums(sign(m) > 0))

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 7
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_exocrine.pdf"), width=w, height=h)

#
# classical genes
#
pathways <- c(
  "Chan_Seng_Yue.ClassicalA",
  "Chan_Seng_Yue.ClassicalB",
  "Collisson.Classical",
  "ICGC.Immunogenic.Up",
  "Moffitt.F8_Classical.top250"
)

x <- filter(gsea, contrast %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix(x, de, gsea_padj_cutoff=gsea_padj_cutoff, 
                              min_gene_sets=2, min_contrasts=1)

# show intersection with CPTAC
genes <- rownames(m)
length(genes)
table(annot_row[genes,]$de_cptac_rna)
table(annot_row[genes,]$de_cptac_prot)
table(annot_row[genes,]$de_cptac_rna, annot_row[genes,]$de_cptac_prot)
table(annot_row[genes,]$prognostic)
# agreement across studies
table(rowSums(sign(m) > 0))
# negatively enriched
apply(sign(m), 1, function(x) { any(x < 0) })

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 10
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_classical.pdf"), width=w, height=h)

#
# basallike genes
#
pathways <- c("Moffitt.F6_BasalLike.top250",
              "Collisson.QM",
              "ICGC.Squamous.Up",
              "Chan_Seng_Yue.BasalA",
              "Chan_Seng_Yue.BasalB")

x <- filter(gsea, contrast %in% contrasts, pathway %in% pathways)
m <- gsea_leading_edge_matrix(x, de, gsea_padj_cutoff=gsea_padj_cutoff, 
                              min_gene_sets=2, min_contrasts=1)

# show intersection with CPTAC
genes <- rownames(m)
length(genes)
table(annot_row[genes,]$de_cptac_rna)
table(annot_row[genes,]$de_cptac_prot)
table(annot_row[genes,]$de_cptac_rna, annot_row[genes,]$de_cptac_prot)
table(annot_row[genes,]$prognostic)
# agreement across studies
table(rowSums(sign(m) > 0))
# negatively enriched
apply(sign(m), 1, function(x) { any(x < 0) })
table(apply(sign(m), 1, function(x) { any(x < 0) }))
table(apply(sign(m), 1, function(x) { all(x <= 0) }))

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 10
w <- 6
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_pdac_basal.pdf"), width=w, height=h)

```

### GSEA LE hallmark

```{r gsea leading edge hallmark}

# setup heatmap annotations
gene_annot <- fds_duke_panck$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na")) %>% 
  mutate(moffitt_basallike = case_when(gene %in% gs_pdacr$Moffitt.F6_BasalLike.top250 ~ "up",
                                       TRUE ~ "no"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot, moffitt_basallike) %>%
  column_to_rownames("gene") %>%
  as.data.frame()


# 
# Prepare gsea results
#
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")
gsea <- gsea_all %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff) %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()


pathways <- c(
  "HALLMARK_INTERFERON_ALPHA_RESPONSE",
  "HALLMARK_INTERFERON_GAMMA_RESPONSE",
  "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION"
)

x <- filter(gsea, pathway %in% pathways)
m <- gsea_leading_edge_matrix(x, de, gsea_padj_cutoff=gsea_padj_cutoff, 
                              de_padj_cutoff=0.01, 
                              min_gene_sets=1, min_contrasts=2)

p <- pheatmap(m, 
              scale = "none",
              color = colorRampPalette(c("blue", "white", "red"))(50),
              breaks = seq(-1, 1, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              border_color = "black",
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "ward.D2",
              annotation_row = annot_row,
              annotation_colors = color_scales,
              fontsize_col = 4,
              fontsize_row = 6)
h <- 10
w <- 5
save_pheatmap_pdf(p, filename=file.path(plot_dir, "gsea_leading_edge_hallmark.pdf"), width=w, height=h)

```

### GSEA GoL Validation

```{r gsea gol validation}

gsea <- NULL

# CPTAC
res <- fgsea_run_batch(de_gol, gs_cptac, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# HPA
res <- fgsea_run_batch(de_gol, gs_hpa_panc, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# pdacr
res <- fgsea_run_batch(de_gol, gs_pdacr[pdacr_pathways], 
                       fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# hallmark
res <- fgsea_run_batch(de_gol, gs_hallmark, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
gsea_gol <- gsea

```

### GSEA GoL Plots

```{r gsea gol plots}

# contrasts
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")

# prepare gsea results
gsea <- gsea_gol %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff) %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()


#
# CPTAC and HPA
#
cptac_pathways <- c("CPTAC_RNA_ALL_UP", "CPTAC_RNA_ALL_DN",
                    "CPTAC_PROT_ALL_UP", "CPTAC_PROT_ALL_DN")
hpa_pathways <- c("HPA_PROGNOSIS_PANCREATIC_CANCER_UNFAV_ALL",  
                  "HPA_PROGNOSIS_PANCREATIC_CANCER_FAV_ALL")
res <- gsea %>% filter(pathway %in% c(cptac_pathways, hpa_pathways))
p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_grid(~ pathway) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gol_gsea_barplot_clusters_cptac_hpa")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=5, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=5, height=4)



# 
# PDAC subtypes
#

#
# Moffitt (main figure)
#
res <- gsea %>% filter(pathway %in% moffitt_pathways)

# arrange for readability
pathway_levels <- res %>% filter(contrast == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gol_gsea_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=4)

#
# pdacr pathways (supplemental figure)
#
res <- gsea %>% filter(pathway %in% pdacr_pathways)
res$pathway <- factor(res$pathway, levels=pdacr_pathways)
# add author
res <- res %>% mutate(
  author = case_when(pathway %in% moffitt_pathways ~ "Moffitt",
                     pathway %in% collison_pathways ~ "Collison",
                     pathway %in% icgc_pathways ~ "ICGC",
                     pathway %in% chan_seng_yue_pathways ~ "Chan Seng Yue")
)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)") +
  facet_wrap(author ~ pathway, nrow=1)
f <- paste0("gol_gsea_barplot_clusters_pdac_subtypes")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=16, height=5)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=16, height=5)



# 
# Hallmark
#
res <- gsea %>% filter(pathway %in% names(gs_hallmark)) %>% 
  mutate(pathway = gsub("HALLMARK_", "", pathway))
# only keep significant results (at least one contrast)
res <- res %>% group_by(pathway) %>%
  filter(any(padj < gsea_padj_cutoff)) %>%
  ungroup()

# arrange by clusterHRvLR for readability
pathway_levels <- filter(res, contrast == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=7)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ contrast, nrow=1) + 
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")
f <- paste0("gol_gsea_barplot_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=3)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=3)

```


## GSVA

```{r duke ipmn gsva analysis}

y <- fds_duke_panck$log2_norm_cpm
gs <- c(gs_cptac, gs_hpa_panc, gs_pdacr, gs_hallmark) 

# run gsva
x <- gsva(GSVA::gsvaParam(y, gs))
rownames(x) <- gsub("HALLMARK_", "H_", rownames(x))
rownames(x) <- gsub("HPA_PROGNOSIS_", "HPA_", rownames(x))
gsva_res <- x

```

### GSVA hallmark pathways

```{r gsva hallmark}

# hallmark pathways
gene_sets <- c("H_TNFA_SIGNALING_VIA_NFKB", 
               "H_INTERFERON_GAMMA_RESPONSE",
               "H_EPITHELIAL_MESENCHYMAL_TRANSITION")

path_aoi_shapes = list(
  "nl"=21,
  "lgd"=25,
  "hgd"=24,
  "inv"=23
)

hallmark_color_scales = list(
  "H_TNFA_SIGNALING_VIA_NFKB" = "#ff0000",
  "H_INTERFERON_GAMMA_RESPONSE" = "#ffff00",
  "H_EPITHELIAL_MESENCHYMAL_TRANSITION" = "#0022ff"
)


s <- fds_duke_panck$samples
s <- bind_cols(s, pca_tbl[,1:npcs])
y <- fds_duke_panck$log2_norm_cpm
x <- as_tibble(gsva_res[gene_sets,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, s, by=join_by(aoi == aoi))
x$path_aoi <- factor(x$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))

x <- x %>% 
  group_by(patient, pathway, path_aoi) %>%
  mutate(medianvalue = median(value)) %>%
  ungroup()

# boxplot individual patients
p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0.05, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "boxplot_gsva_hallmark_inflammation")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# boxplots
#
p1 <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.05, dodge.width=0.75),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = hallmark_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.title = element_text(size=8),
        legend.text = element_text(size=7))
f <- file.path(plot_dir, "boxplot_gsva_hallmark_path")
ggsave(paste0(f, ".pdf"), p1, width=7, height=5)

p2 <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.05, dodge.width=0.75),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = hallmark_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.title = element_text(size=8),
        legend.text = element_text(size=7))
f <- file.path(plot_dir, "boxplot_gsva_hallmark_cluster")
ggsave(paste0(f, ".pdf"), p2, width=6, height=4)

p <- p2 + p1 + patchwork::plot_layout(guides="collect", widths=c(1,2))
f <- file.path(plot_dir, "boxplot_gsva_hallmark_path_cluster")
ggsave(paste0(f, ".pdf"), p, width=8, height=4)

```


### GSVA pdac subtypes

```{r gsva pdac subtypes}

gene_sets <- c("Moffitt.F3_Exocrine.top250",
               "Moffitt.F8_Classical.top250",
               "Moffitt.F6_BasalLike.top250")

moffitt_color_scales = list(
  "Moffitt.F3_Exocrine.top250" = "#4400bb",
  "Moffitt.F8_Classical.top250" = "#ffff00",
  "Moffitt.F6_BasalLike.top250" = "#ff00ff"
)

path_aoi_shapes = list(
  "nl"=21,
  "lgd"=25,
  "hgd"=24,
  "inv"=23
)

s <- bind_cols(fds_duke_panck$samples, pca_tbl[,1:npcs])
y <- fds_duke_panck$log2_norm_cpm
x <- as_tibble(gsva_res[gene_sets,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, s, by=join_by(aoi == aoi))
# x$path <- factor(x$path, levels=c("ipmn_lg_nl_epithelial", "ipmn_lg_lgd_epithelial",
#                                   "ipmn_pdac_nl_epithelial", "ipmn_pdac_lgd_epithelial",
#                                   "ipmn_pdac_hgd_epithelial", "ipmn_pdac_inv_epithelial"))
x$path_aoi <- factor(x$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))
x$pathway <- factor(x$pathway, levels=gene_sets)

p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "boxplot_patient_gsva_pdacr_subtypes")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# PCA plot
#
p <- ggplot(x, aes(x=PC1, y=PC2, color=value, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_minimal() +
  theme(legend.text = element_text(size=7)) +
  labs(x="PC1", y="PC2", title="Moffitt Subtypes") + 
  facet_wrap(~ pathway, nrow=1)
f <- file.path(plot_dir, "pca_featureplot_moffitt")
ggsave(paste0(f, ".pdf"), plot=p, width=9, height=3.5)
ggsave(paste0(f, ".png"), plot=p, width=9, height=3.5)


#
# lineplot
#
x <- x %>% 
  group_by(patient, pathway, path_aoi) %>%
  mutate(medianvalue = median(value)) %>%
  ungroup()

p <- ggplot(x, aes(color=patient, shape=path_specimen, group=patient)) + 
  geom_point(aes(x=path_aoi, y=value)) +
  geom_line(aes(x=path_aoi, y=medianvalue, group=patient)) +
  geom_hline(yintercept=0) +
  scale_color_manual(values = color_scales$patient) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  facet_grid(pathway ~ path_specimen, scales="free_x") + 
  theme_minimal()
f <- file.path(plot_dir, "lineplot_gsva_moffitt_path")
ggsave(paste0(f, ".pdf"), p, width=w, height=h)

#
# boxplots
#
p1 <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75, linewidth=0.5) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_path")
ggsave(paste0(f, ".pdf"), p1, width=5, height=4)

p2 <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_cluster")
ggsave(paste0(f, ".pdf"), p2, width=4, height=4)

p <- p2 + p1 + patchwork::plot_layout(guides="collect", widths=c(1,2))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_path_cluster")
ggsave(paste0(f, ".pdf"), p, width=7, height=4)


```


### GSVA PCA correlations

```{r gsva pca correlation}

gene_sets <- c("Moffitt.F3_Exocrine.top250",
               "Moffitt.F8_Classical.top250",
               "Moffitt.F6_BasalLike.top250")
pc_names <- c("PC1", "PC2", "PC3")

x <- bind_cols(fds_duke_panck$samples, as_tibble(pca_res$x), t(gsva_res))
x <- x %>% select(aoi, path_specimen, path_aoi, cluster_name, 
                  all_of(pc_names), 
                  all_of(gene_sets))

coefs <- matrix(nrow=length(pc_names), ncol=length(gene_sets), dimnames=list(pc_names, gene_sets))
pvals <- matrix(nrow=length(pc_names), ncol=length(gene_sets), dimnames=list(pc_names, gene_sets))
for (i in pc_names) {
  for (j in gene_sets) {
    res <- cor.test(pull(x, i), pull(x, j), method="pearson", exact=TRUE)
    coefs[i, j] <- res$estimate
    pvals[i, j] <- res$p.value
  }
}
coefs <- as_tibble(coefs, rownames="pc_name") %>% 
  pivot_longer(cols=gene_sets, names_to="gs_name", values_to="coef")
pvals <- as_tibble(pvals, rownames="pc_name") %>% 
  pivot_longer(cols=gene_sets, names_to="gs_name", values_to="pvals")


x <- x %>% pivot_longer(cols=all_of(gene_sets), names_to="gs_name", values_to="gsva")
x <- x %>% pivot_longer(cols=all_of(pc_names), names_to="pc_name", values_to="pc_value")
x$gs_name <- factor(x$gs_name, levels=gene_sets)
x$pc_name <- factor(x$pc_name, levels=pc_names)
x <- left_join(x, coefs, by=join_by(pc_name, gs_name))
x <- left_join(x, pvals, by=join_by(pc_name, gs_name))
x <- x %>% mutate(label=paste0("r=", round(coef, 3), " p=", signif(pvals, 3)))

# for formatting
x$gs_name <- factor(x$gs_name, levels=gene_sets)

p <- ggplot(x, aes(x=pc_value, y=gsva)) + 
  geom_point(aes(color=path_aoi, shape=path_specimen)) + 
  geom_smooth(color="black", linetype="dashed", method='lm', formula=y~x) +
  scale_color_manual(values = color_scales$path_aoi) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  geom_text(data=distinct(x, gs_name, pc_name, .keep_all=TRUE), 
            aes(label=label), x=Inf, y=Inf, size=3, vjust="inward", hjust="inward") + 
  theme_minimal() + 
  theme(aspect.ratio=1) + 
  facet_grid(pc_name ~ gs_name)
f <- file.path(plot_dir, "scatter_cor_pca_moffitt")
ggsave(paste0(f, ".png"), p, width=7, height=5)
ggsave(paste0(f, ".pdf"), p, width=7, height=5)


```


### BG AUC PCA correlations

```{r bg auc pca correlation}

pc_names <- c("PC1", "PC2", "PC3")

x <- bind_cols(fds_duke_panck$samples, as_tibble(pca_res$x))
x <- x %>% select(aoi, path_specimen, path_aoi, cluster_name, bg_auc, all_of(pc_names))

coefs <- NULL
pvals <- NULL
for (i in pc_names) {
  res <- cor.test(pull(x, i), x$bg_auc, method="pearson", exact=TRUE)
  coefs[i] <- res$estimate
  pvals[i] <- res$p.value
}

coefs <- as_tibble(coefs, rownames="pc_name") %>% dplyr::rename(coef=value)
pvals <- as_tibble(pvals, rownames="pc_name") %>% dplyr::rename(pval=value)

x <- x %>% pivot_longer(cols=all_of(pc_names), names_to="pc_name", values_to="pc_value")
x$pc_name <- factor(x$pc_name, levels=pc_names)
x <- left_join(x, coefs, by=join_by(pc_name==pc_name))
x <- left_join(x, pvals, by=join_by(pc_name==pc_name))
x <- x %>% mutate(label=paste0("r=", round(coef, 3), " p=", signif(pval, 3)))


p <- ggplot(x, aes(x=pc_value, y=bg_auc)) + 
  geom_point(aes(color=bg_auc, shape=path_specimen)) + 
  geom_smooth(color="black", linetype="dashed", method='lm', formula=y~x) +
  scale_color_viridis_c() +
  scale_shape_manual(values = shape_scales$path_specimen) +
  geom_text(data=distinct(x, pc_name, .keep_all=TRUE), 
            aes(label=label), x=Inf, y=Inf, size=3, vjust="inward", hjust="inward") + 
  theme_minimal() + 
  theme(aspect.ratio=1) + 
  facet_grid(~ pc_name)
f <- file.path(plot_dir, "scatter_cor_pca_bg_auc")
ggsave(paste0(f, ".png"), p, width=7, height=5)
ggsave(paste0(f, ".pdf"), p, width=7, height=5)


```


### GSVA GoL Validation

```{r gsva gol validation}

s <- bind_cols(fds_panck$samples, pca_tbl.validation[,1:npcs])
s <- s %>% filter(study == "gol")
y <- fds_panck$log2_norm_cpm[, s$aoi]
gs <- c(gs_cptac, gs_hpa_panc, gs_pdacr, gs_hallmark) 

# run gsva
x <- gsva(GSVA::gsvaParam(y, gs))
rownames(x) <- gsub("HALLMARK_", "H_", rownames(x))
rownames(x) <- gsub("HPA_PROGNOSIS_", "HPA_", rownames(x))
res <- x


gene_sets <- c("Moffitt.F3_Exocrine.top250",
               "Moffitt.F8_Classical.top250",
               "Moffitt.F6_BasalLike.top250")

moffitt_color_scales = list(
  "Moffitt.F3_Exocrine.top250" = "#4400bb",
  "Moffitt.F8_Classical.top250" = "#ffff00",
  "Moffitt.F6_BasalLike.top250" = "#ff00ff"
)

path_aoi_shapes = list(
  "duct"=21,
  "panin"=25,
  "pdac"=23
)


x <- as_tibble(res[gene_sets,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, s, by=join_by(aoi == aoi))
x$path_aoi <- factor(x$path_aoi, levels=c("duct", "panin", "pdac"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))
x$pathway <- factor(x$pathway, levels=gene_sets)

p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "boxplot_gol_patient_gsva_pdacr_subtypes")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# PCA plot
#
p <- ggplot(x, aes(x=PC1, y=PC2, color=value, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_minimal() +
  theme(legend.text = element_text(size=7)) +
  labs(x="PC1", y="PC2", title="Moffitt Subtypes") + 
  facet_wrap(~ pathway, nrow=1)
f <- file.path(plot_dir, "pca_gol_featureplot_moffitt")
ggsave(paste0(f, ".pdf"), plot=p, width=9, height=3.5)
ggsave(paste0(f, ".png"), plot=p, width=9, height=3.5)


#
# boxplots
#
p1 <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75, linewidth=0.5) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_path")
ggsave(paste0(f, ".pdf"), p1, width=5, height=4)

p2 <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "boxplot_gsva_moffitt_cluster")
ggsave(paste0(f, ".pdf"), p2, width=4, height=4)

p <- p2 + p1 + patchwork::plot_layout(guides="collect", widths=c(1,2))
f <- file.path(plot_dir, "boxplot_gol_gsva_moffitt_path_cluster")
ggsave(paste0(f, ".pdf"), p, width=7, height=4)



#
# hallmark pathways
#
gene_sets <- c("H_TNFA_SIGNALING_VIA_NFKB", 
               "H_INTERFERON_GAMMA_RESPONSE",
               "H_EPITHELIAL_MESENCHYMAL_TRANSITION")
hallmark_color_scales = list(
  "H_TNFA_SIGNALING_VIA_NFKB" = "#ff0000",
  "H_INTERFERON_GAMMA_RESPONSE" = "#ffff00",
  "H_EPITHELIAL_MESENCHYMAL_TRANSITION" = "#0022ff"
)


x <- as_tibble(res[gene_sets,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, s, by=join_by(aoi == aoi))
x$path_aoi <- factor(x$path_aoi, levels=c("duct", "panin", "pdac"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))
x$pathway <- factor(x$pathway, levels=gene_sets)

x <- x %>% 
  group_by(patient, pathway, path_aoi) %>%
  mutate(medianvalue = median(value)) %>%
  ungroup()

# boxplot individual patients
p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0.05, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "gol_boxplot_gsva_hallmark_inflammation")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# boxplots
#
p <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.05, dodge.width=0.75),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = hallmark_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"))
f <- file.path(plot_dir, "gol_boxplot_gsva_hallmark_path")
ggsave(paste0(f, ".pdf"), p, width=7, height=5)

p <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.05, dodge.width=0.75),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = hallmark_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"))
f <- file.path(plot_dir, "gol_boxplot_gsva_hallmark_cluster")
ggsave(paste0(f, ".pdf"), p, width=6, height=4)


```



## Duke epithelial NL-LGD analysis

### NL-LGD PCA

```{r nl_lgd cohort}

# make nl_lgd cohort
s <- fds_duke_panck$samples %>% filter(path_aoi %in% c("nl","lgd"))
y <- fds_duke_panck$log2_norm_cpm[, s$aoi]

# subset most variable genes
ntop <- 500
y <- most_var_genes(y, ntop=ntop, span=loess_span)

# plot dimensions
width <- 4
height <- 3

# PCA
nllgd.pca <- prcomp(t(y), center=TRUE, scale.=FALSE)
nllgd.pca_var <- round(100 * nllgd.pca$sdev^2 / sum( nllgd.pca$sdev^2 ), 2)
nllgd.pca_var_txt <- paste(colnames(nllgd.pca$x), " (", paste(as.character(nllgd.pca_var), "%", ")", sep=""), sep="")

# percent variance of each PC
p <- fviz_eig(nllgd.pca, col.var="blue", addlabels=TRUE)
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_eig.pdf"))
ggsave(f, p)

# contributions of variables to PC1 and PC2
p <- fviz_contrib(nllgd.pca, choice="var", axes = 1:2, top=fviz_ntop) +
  theme(axis.text.x = element_text(angle=90))
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_contrib.pdf"))
ggsave(f, p)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  nllgd.pca,
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("pca_nllgd_fviz_pca_var.pdf"))
ggsave(f, p, width=7, height=7)

#
# PCA plots
#
x <- bind_cols(s, as_tibble(nllgd.pca$x))

# bg_ks_dist
p <- ggplot(x, aes(x=PC1, y=PC2, color=bg_auc)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_bg_auc.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# patient
p <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$patient) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_patient.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path specimen
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_path_specimen.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path aoi
p <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_path_aoi.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)

# path
p <- ggplot(x, aes(x=PC1, y=PC2, color=path, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
f <- file.path(plot_dir, paste0("pca_nllgd_path.pdf"))
ggsave(f, plot=p, device="pdf", width=width, height=height)


```


### NL-LGD PCA Clustering

```{r nllgd cluster}

# make nl_lgd cohort
s <- fds_duke_panck$samples %>% filter(path_aoi %in% c("nl","lgd"))
y <- fds_duke_panck$log2_norm_cpm[, s$aoi]

# set number of pcs
npcs.nllgd <- 2
# set number of clusters
nclusters.nllgd <- 3
pca_tbl.nllgd <- nllgd.pca$x[,1:npcs.nllgd]

# pam
pamx <- cluster::pam(stats::dist(pca_tbl.nllgd), k=nclusters.nllgd)
memb <- paste0("c", pamx$clustering)

# add cluster membership to sample properties
s$cluster_nllgd <- memb
# save clustering
nllgd_samples <- s

```

### NL-LGD PCA cluster plots and naming

```{r nllgd pca cluster plots}

# plot clustering
x <- bind_cols(nllgd_samples, as_tibble(pca_tbl.nllgd))

# cluster
p <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_nllgd, shape=path_specimen)) +
  geom_point(alpha = 0.7, size=2) +
  scale_color_manual(values = color_scales$cluster) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])
p

# identify clusters by plot
nllgd_samples <- nllgd_samples %>% mutate(
  cluster_nllgd_name = case_when(cluster_nllgd == "c1" ~ "cHR",
                                 cluster_nllgd == "c2" ~ "cLR",
                                 cluster_nllgd == "c3" ~ "cNL")
)
x <- bind_cols(nllgd_samples, as_tibble(nllgd.pca$x))

# path specimen
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# path aoi
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# patient
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = pals::cols25()) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# cluster (original)
p4 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster_name) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

# cluster (nl-lgd)
p5 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_nllgd_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$cluster_name) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_bw() +
  theme(aspect.ratio=1) +
  labs(x=nllgd.pca_var_txt[1], y=nllgd.pca_var_txt[2])

p <- p1 + p2 + p3 + p4 + theme(legend.position="none") + p5 + plot_layout(nrow=1)
f <- file.path(plot_dir, paste0("pca_nllgd_pca_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=22, height=6)

# compare clustering
table(nllgd_samples$cluster_nllgd_name, nllgd_samples$cluster_name)


```

### NL-LGD DE

```{r nl_lgd de}

s <- nllgd_samples
y <- fds_duke_panck$log2_norm_cpm[, s$aoi]

s$patient <- factor(s$patient)
s$path_aoi <- factor(s$path_aoi, levels=c("nl", "lgd"))
s$path_specimen <- factor(s$path_specimen, levels=c("ipmn_lg", "ipmn_pdac"))
s$path_patient <- paste0(s$path_specimen, "_", s$path_aoi)
s$path_patient <- factor(s$path_patient, 
                         levels=c("ipmn_lg_nl", "ipmn_lg_lgd", 
                                  "ipmn_pdac_nl", "ipmn_pdac_lgd"))
s$cluster <- factor(s$cluster_nllgd_name, levels=c("cNL", "cLR", "cHR"))

nllgd_de <- NULL

#
# Clusters
#
design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(clusterHRvLR = clustercHR - clustercLR,
                           clusterHRvNL = clustercHR - clustercNL,
                           clusterLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
nllgd_de <- res

# write cluster results
write_xlsx(res, file.path(results_dir, "de_nllgd_clusters.xlsx"))

```

### NL-LGD DE correlations

```{r nllgd de correlations}

# merge de analysis by gene
nllgd_de_merged <- nllgd_de %>% select(gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))

cor.test(nllgd_de_merged$log2fc_clusterHRvNL, nllgd_de_merged$log2fc_clusterLRvNL)
cor.test(nllgd_de_merged$log2fc_clusterHRvNL, nllgd_de_merged$log2fc_clusterHRvLR)
cor.test(nllgd_de_merged$log2fc_clusterLRvNL, nllgd_de_merged$log2fc_clusterHRvLR)


```

### NL-LGD DE plots

```{r nllgd de plots}

# volcano plots
for (mycontrast in unique(nllgd_de$contrast)) {
  print(mycontrast)
  x <- filter(nllgd_de, contrast == mycontrast)
  p <- plot_de_volcano(x) + 
    labs(title="Volcano Plot", subtitle=mycontrast)
  f <- paste0("volcano_nllgd_de_", mycontrast)
  ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=8)
  ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=8)
}


#
# scatter plots
#
nlabel <- 10
width <- 8
height <- 8


# clusters
p <- plot_de_scatter2d(nllgd_de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterHRvLR", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "scatter2d_nllgd_de_x_clusterHRvNL_y_clusterHRvLR")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(nllgd_de_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterLRvNL", 
                       nlabel=nlabel) +
  labs(title="IPMN Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterLRvNL (y-axis)")
f <- file.path(plot_dir, "scatter2d_nllgd_de_x_clusterHRvNL_y_clusterLRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


#
# Heatmap
#
# setup heatmap annotations
gene_annot <- fds_duke_panck$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na")) %>% 
  mutate(moffitt_basallike = case_when(gene %in% gs_pdacr$Moffitt.F6_BasalLike.top250 ~ "up",
                                       TRUE ~ "no"),
         moffitt_classical = case_when(gene %in% gs_pdacr$Moffitt.F8_Classical.top250 ~ "up",
                                       TRUE ~ "no"),
         moffitt_exocrine = case_when(gene %in% gs_pdacr$Moffitt.F3_Exocrine.top250 ~ "up",
                                       TRUE ~ "no"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot, 
         moffitt_basallike, moffitt_classical, moffitt_exocrine) %>%
  column_to_rownames("gene") %>%
  as.data.frame()

annot_col <- column_to_rownames(nllgd_samples, "aoi") %>% 
  select(patient, path_aoi, path_specimen, cluster_name, cluster_nllgd_name) %>%
  as.data.frame()


# de gene params
padj_cutoff <- 0.01
log2fc_cutoff <- 3
ntop <- 5
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

g <- filter(nllgd_de, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)
cluster_marker_genes <- g

# centered gene expression (subtract mean)
s <- nllgd_samples
x <- fds_duke_panck$log2_norm_cpm[g, s$aoi]
u <- apply(x, 1, mean)
x <- sweep(x, 1, u, FUN="-")
xmin <- -6
xmax <- 6

# heatmap
f <- file.path(plot_dir,  paste0("heatmap_nllgd_de_cluster_markers_", ngenes, "_genes"))
h <- 10
w <- 10

p <- pheatmap(x,
              scale = "none",
              color = colorRampPalette(c("navy", "white", "firebrick3"))(50),
              breaks = seq(xmin, xmax, length.out=51),
              cellwidth = 6,
              cellheight = 6,
              annotation_col = annot_col[s$aoi,],
              annotation_row = annot_row[g,],
              annotation_colors = color_scales,
              border_color = NA,
              cluster_rows = TRUE,
              cluster_cols = TRUE,
              clustering_distance_rows = "euclidean",
              clustering_distance_cols = "euclidean",
              clustering_method = "complete",
              fontsize = 6,
              fontsize_row = 6,
              fontsize_col = 6,
              filename = paste0(f, ".png"), 
              width = w,
              height = h)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)



```

### NL-LGD GSEA

```{r nllgd gsea}

gsea <- NULL

# CPTAC
res <- fgsea_run_batch(nllgd_de, gs_cptac, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# HPA
res <- fgsea_run_batch(nllgd_de, gs_hpa_panc, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# pdacr
res <- fgsea_run_batch(nllgd_de, gs_pdacr[pdacr_pathways], 
                       fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
# hallmark
res <- fgsea_run_batch(nllgd_de, gs_hallmark, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
gsea_nllgd <- gsea

```

### NL-LGD GSEA write gene sets

```{r}

# write results for clusters
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")
manuscript_pathways <- c(names(gs_cptac), names(gs_hpa_panc), 
                         pdacr_pathways, names(gs_hallmark))
res <- gsea_nllgd %>% filter(contrast %in% contrasts, pathway %in% manuscript_pathways)

prefix <- "gsea_nllgd"
f <- file.path(results_dir, paste0(prefix, "_results.tsv"))
data.table::fwrite(res, file=f, sep="\t", sep2=c("", " ", ""))
f <- file.path(results_dir, paste0(prefix, "_results.xlsx"))
write_xlsx(res, f)

```


### NL-LGD GSEA plots

```{r nllgd gsea plots}

# contrasts
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")

# prepare gsea results
x <- gsea_all %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff,
         group="all") %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()

y <- gsea_nllgd %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff,
         group="nllgd") %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()
gsea <- bind_rows(x, y)


#
# CPTAC and HPA
#
pathways <- c(cptac_pathways, hpa_pathways)
res <- gsea %>% filter(pathway %in% pathways)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway + group, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_nllgd_barplot_clusters_pdac")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=9, height=3)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=9, height=3)


#
# Hallmark
#
res <- gsea %>% filter(pathway %in% names(gs_hallmark))
res <- res %>% mutate(
  contrast = sub("cluster", "", contrast), 
  pathway = gsub("HALLMARK_", "", pathway)
)

# only keep significant results (at least one analysis)
res <- res %>% group_by(pathway) %>%
  filter(any(padj < gsea_padj_cutoff)) %>%
  ungroup()
# arrange by cluster 1v2 for readability
pathway_levels <- res %>% 
  filter(group == "all", contrast == "HRvLR") %>%
  arrange(NES) %>%
  distinct(pathway) %>%
  pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  geom_vline(xintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=6)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ contrast + group, nrow=1) +
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")
f <- paste0("gsea_nllgd_barplot_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=3)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=3)


#
# Moffitt
# 
res <- gsea %>% filter(pathway %in% moffitt_pathways)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway + group, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("gsea_nllgd_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=5, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=5, height=4)


```


# Duke stromal analysis

## Setup

```{r duke stromal setup}

# get roi grouping from original dataset
# join with epithelial clustering
x <- duke_ipmn_ds$samples %>% select(aoi, roi_group)
x <- left_join(fds_duke_panck$samples, x, by=join_by(aoi))
x <- x %>% select(roi_group, cluster_name)

# join epithelial clustering with stromal samples
s <- left_join(fds_duke_stroma$samples, x, by=join_by(roi_group))
s <- s %>% mutate(
  cluster_name = if_else(is.na(cluster_name), "cNA", cluster_name)
)

# set patient color scales
stroma_color_scales <- color_scales
stroma_color_scales$patient = setNames(pals::cols25(n_distinct(s$patient)),
                                    sort(unique(s$patient)))

fds_duke_stroma$samples <- s

table(fds_duke_stroma$samples$path_aoi)
table(fds_duke_stroma$samples$cluster_name)

```


## HVG

```{r duke stromal hvg}

ntop <- 500
nlabel <- 100

s <- fds_duke_stroma$samples
y <- fds_duke_stroma$log2_norm_cpm

# maplot
p <- maplot(y, ntop=ntop, nlabel=nlabel, myspan=loess_span)
f <- file.path(plot_dir, paste0("stroma_maplot"))
ggsave(paste0(f, ".pdf"), plot=p, width=6, height=6)
ggsave(paste0(f, ".png"), plot=p, width=6, height=6)

# select most variable genes
y <- most_var_genes(y, ntop=ntop, span=loess_span)

# heatmap
f <- file.path(plot_dir, "stroma_heatmap_highvar_genes.pdf")
p <- heatmap_genes(s, y, f)

```

## PCA hvgs

```{r duke stromal pca}

s <- fds_duke_stroma$samples

# subset most variable genes
y <- most_var_genes(fds_duke_stroma$log2_norm_cpm, ntop=ntop, span=loess_span)
hvgenes.stroma <- rownames(y)

# plot dimensions
width <- 4
height <- 4

# PCA
pca_res.stroma <- prcomp(t(y), center=TRUE, scale.=FALSE)
pca_res.var.stroma <- round(100 * pca_res.stroma$sdev^2 / sum( pca_res.stroma$sdev^2 ), 2)
pca_res.var_txt.stroma <- paste(colnames(pca_res.stroma$x), " (", paste(as.character(pca_res.var.stroma), "%", ")", sep=""), sep="")

# fviz plots
fviz_ntop <- 50

# percent variance of each PC
p <- fviz_eig(pca_res.stroma, col.var="blue", addlabels=TRUE)
f <- file.path(plot_dir, paste0("duke_stroma_pca_fviz_eig.pdf"))
ggsave(f, p, width=4, height=4)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  pca_res.stroma,
  axes = c(1,2),
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("duke_stroma_pca_fviz_var_pc12.pdf"))
ggsave(f, p, width=7, height=7)


#
# PCA plots
#
custom_theme_pca <- function() {
  theme_bw() +
  theme(aspect.ratio=1) +
    theme(legend.title = element_text(size=8), 
          legend.text = element_text(size=7))
}

x <- bind_cols(s, as_tibble(pca_res.stroma$x))

# bg_ks_dist
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=bg_auc)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma[1], y=pca_res.var_txt.stroma[2])

# patient
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$patient) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma[1], y=pca_res.var_txt.stroma[2])

# path specimen
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma[1], y=pca_res.var_txt.stroma[2])

# path aoi
p4 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma[1], y=pca_res.var_txt.stroma[2])

# path
p5 <- ggplot(x, aes(x=PC1, y=PC2, color=path, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma[1], y=pca_res.var_txt.stroma[2])

# clustering
p6 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$cluster_name) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma[1], y=pca_res.var_txt.stroma[2])

p <- p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(nrow=1, guides="collect")
f <- file.path(plot_dir, paste0("duke_stroma_pca_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=20, height=12)


# stromal pca plots for important genes
feature_plot_genes <- c("CTRB1", "INS", "S100A9", "SFRP2", "COL10A1")
s <- bind_cols(s, as_tibble(pca_res.stroma$x))
          
for (gene in feature_plot_genes) {
  print(gene)
  x <- bind_cols(s, gene=unlist(y[gene,]))
  p <- ggplot(x, aes(x=PC1, y=PC2, color=gene, shape=path_specimen)) +
    geom_point(alpha = 0.8, size=3) +
    scale_color_viridis_c() +
    scale_shape_manual(values = shape_scales$path_specimen) +
    theme_minimal() +
    theme(aspect.ratio=1) +
    labs(x="PC1", y="PC2", title=gene)
  f <- file.path(plot_dir, paste0("duke_stroma_pca_featureplot_", gene))
  ggsave(paste0(f, ".pdf"), plot=p, width=4, height=3)
  ggsave(paste0(f, ".png"), plot=p, width=4, height=3)
}

```

## PCA hvg subset by moffitt stromal signatures

```{r duke stromal pca moffit}

s <- fds_duke_stroma$samples

# stromal gene signature
tbl <- get_var_tbl(fds_duke_stroma$log2_norm_cpm, span = loess_span)
tbl <- tbl %>% mutate(
  moffitt_normal_stroma = gene %in% gs_pdacr$Moffitt.F13_NormalStroma.top250,
  moffitt_activated_stroma = gene %in% gs_pdacr$Moffitt.F5_ActivatedStroma.top250,
)
tbl <- tbl %>% filter(rank <= 500) %>% arrange(rank)
f <- file.path(results_dir, "stromal_hvg_moffitt.xlsx")
write_xlsx(tbl, f)

# instead of all variable genes, use moffitt stromal genes
g1 <- intersect(hvgenes.stroma, gs_pdacr$Moffitt.F5_ActivatedStroma.top250)
g2 <- intersect(hvgenes.stroma, gs_pdacr$Moffitt.F13_NormalStroma.top250)
hvgenes.stroma.moffitt <- union(g1, g2)
y <- fds_duke_stroma$log2_norm_cpm[hvgenes.stroma.moffitt,]

# plot dimensions
width <- 4
height <- 4

# PCA
pca_res.stroma.moffitt <- prcomp(t(y), center=TRUE, scale.=FALSE)
pca_res.var.stroma.moffitt <- round(100 * pca_res.stroma.moffitt$sdev^2 / sum( pca_res.stroma.moffitt$sdev^2 ), 2)
pca_res.var_txt.stroma.moffitt<- paste(colnames(pca_res.stroma.moffitt$x), " (", paste(as.character(pca_res.var.stroma.moffitt), "%", ")", sep=""), sep="")

# fviz plots
fviz_ntop <- 50

# percent variance of each PC
p <- fviz_eig(pca_res.stroma.moffitt, col.var="blue", addlabels=TRUE)
f <- file.path(plot_dir, paste0("duke_stroma_moffitt_pca_fviz_eig.pdf"))
ggsave(f, p, width=4, height=4)

# graph of variables contributing to PCA
p <- fviz_pca_var(
  pca_res.stroma.moffitt,
  axes = c(1,2),
  col.var = "contrib", # Color by the quality of representation
  gradient.cols = c("darkorchid4", "gold", "darkorange"),
  repel=TRUE,
  select.var = list(contrib=fviz_ntop)
)
f <- file.path(plot_dir, paste0("duke_stroma_moffitt_fviz_pca_var_pc12.pdf"))
ggsave(f, p, width=7, height=7)


#
# PCA plots
#
custom_theme_pca <- function() {
  theme_bw() +
  theme(aspect.ratio=1) +
    theme(legend.title = element_text(size=8), 
          legend.text = element_text(size=7))
}

x <- bind_cols(s, as_tibble(pca_res.stroma.moffitt$x))

# bg_ks_dist
p1 <- ggplot(x, aes(x=PC1, y=PC2, color=bg_auc)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  custom_theme_pca() +
  labs(x=pca_res.var_txt.stroma.moffitt[1], y=pca_res.var_txt.stroma.moffitt[2])

# patient
p2 <- ggplot(x, aes(x=PC1, y=PC2, color=patient, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values = color_scales$patient) +
  scale_shape_manual(values = shape_scales$path_specimen) +
  custom_theme_pca()

# path specimen
p3 <- ggplot(x, aes(x=PC1, y=PC2, color=path_specimen, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_specimen) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

# path aoi
p4 <- ggplot(x, aes(x=PC1, y=PC2, color=path_aoi, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path_aoi) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

# path
p5 <- ggplot(x, aes(x=PC1, y=PC2, color=path, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$path) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

# clustering
p6 <- ggplot(x, aes(x=PC1, y=PC2, color=cluster_name, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_manual(values=color_scales$cluster_name) +
  scale_shape_manual(values=shape_scales$path_specimen) +
  custom_theme_pca()

p <- p1 + p2 + p3 + p4 + p5 + p6 + plot_layout(nrow=1, guides="collect")
f <- file.path(plot_dir, paste0("duke_stroma_moffitt_pca_plots.pdf"))
ggsave(f, plot=p, device="pdf", width=20, height=12)



# PCA statistics
tbl <- bind_cols(s, as_tibble(pca_res.stroma.moffitt$x)) %>% mutate(
  path_aoi_group = case_match(path_aoi, c("hgd", "inv") ~ "hgdinv",
                              .default = path_aoi)
)
tbl <- table(tbl$path_aoi_group, tbl$PC1 > 0)
chisq <- chisq.test(tbl)
chisq

tbl <- bind_cols(s, as_tibble(pca_res.stroma.moffitt$x))%>% 
  filter(cluster_name != "cNA")
tbl <- table(tbl$cluster_name, tbl$PC1 > 0)
chisq <- chisq.test(tbl)
chisq


```

## DE

```{r duke stromal de}

s <- fds_duke_stroma$samples
y <- fds_duke_stroma$log2_norm_cpm[, s$aoi]

s$cluster <- factor(s$cluster_name, levels=c("cNA", "cNL", "cLR", "cHR"))
design <- model.matrix(~0 + cluster, s)
contrasts <- makeContrasts(clusterHRvLR = clustercHR - clustercLR,
                           clusterHRvNL = clustercHR - clustercNL,
                           clusterLRvNL = clustercLR - clustercNL,
                           levels=design)
res <- run_limma_trend(y, design, contrasts, 
                       padj_cutoff=de_padj_cutoff, 
                       log2fc_cutoff=de_log2fc_cutoff)
table(res$contrast, res$de)
de_stroma <- res

# write cluster results
write_xlsx(res, file.path(results_dir, "de_clusters_stroma.xlsx"))


# volcano plots
for (mycontrast in unique(de_stroma$contrast)) {
  print(mycontrast)
  x <- filter(de_stroma, contrast == mycontrast)
  p <- plot_de_volcano(x) + 
    labs(title="Volcano Plot", subtitle=mycontrast)
  f <- paste0("duke_stroma_volcano_de_", mycontrast)
  ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=8, height=8)
  ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=8, height=8)
}




# setup heatmap annotations
gene_annot <- fds_duke_stroma$meta %>%
  left_join(hpa_pdac, by=c("gene"="gene_symbol")) %>%
  left_join(cptac_de_merged, by=c("gene"="gene_symbol")) %>%
  mutate(de_cptac_rna = replace_na(de_cptac_rna, "na"),
         de_cptac_prot = replace_na(de_cptac_prot, "na"))
annot_row <- gene_annot %>%
  select(gene, prognostic, de_cptac_rna, de_cptac_prot) %>%
  column_to_rownames("gene") %>%
  as.data.frame()
annot_col <- column_to_rownames(fds_duke_stroma$samples, "aoi") %>% 
  select(patient, path_aoi, path_specimen, cluster_name) %>%
  as.data.frame()

# de gene params
padj_cutoff <- 0.05
log2fc_cutoff <- 1
ntop <- 10
contrasts <- c("clusterHRvLR", "clusterHRvNL", "clusterLRvNL")

g <- filter(de_stroma, contrast %in% contrasts,
            abs(log2fc) > log2fc_cutoff, 
            padj < padj_cutoff) %>%
  group_by(contrast, de) %>%
  slice_max(abs(log2fc), n=ntop) %>%
  ungroup() %>%
  distinct(gene) %>% 
  pull(gene)
ngenes <- length(g)
duke_stroma_cluster_marker_genes <- g

# centered gene expression (subtract median)
x <- fds_duke_stroma$log2_norm_cpm[g, s$aoi]
u <- apply(x, 1, mean)
x <- sweep(x, 1, u, FUN="-")
xmin <- -6
xmax <- 6

# heatmap
f <- file.path(plot_dir,  paste0("duke_stroma_heatmap_de_cluster_markers_", ngenes, "_genes"))
h <- 6
w <- 6
p <- plot_pheatmap(x, 
                   annot_col[s$aoi,], 
                   annot_row[g,], 
                   annot_colors=stroma_color_scales,
                   filename=paste0(f, ".png"), 
                   width=w, height=h,
                   xmin=xmin, xmax=xmax)
save_pheatmap_pdf(p, filename=paste0(f, ".pdf"), width=w, height=h)


# merge de analysis by gene
de_stroma_merged <- de_stroma %>% select(gene, contrast, log2fc, padj, de) %>%
  pivot_wider(names_from = contrast, values_from = c(log2fc, padj, de))
nlabel <- 10
width <- 8
height <- 8

#
# GoL scatter plots
#
p <- plot_de_scatter2d(de_stroma_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterHRvLR", 
                       nlabel=nlabel) +
  labs(title="GoL Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterHRvLR (y-axis)")
f <- file.path(plot_dir, "duke_stroma_scatter2d_de_x_clusterHRvNL_y_clusterHRvLR")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)

p <- plot_de_scatter2d(de_stroma_merged, 
                       xcontrast="clusterHRvNL", 
                       ycontrast="clusterLRvNL", 
                       nlabel=nlabel) +
  labs(title="GoL Specimens", 
       subtitle="clusterHRvNL (x-axis) clusterLRvNL (y-axis)")
f <- file.path(plot_dir, "duke_stroma_scatter2d_de_x_clusterHRvNL_y_clusterLRvNL")
ggsave(paste0(f, ".pdf"), plot=p, width=8, height=8)
ggsave(paste0(f, ".png"), plot=p, width=8, height=8)


```

## GSEA

```{r duke stromal gsea, include=FALSE}

gsea <- NULL

moffitt_stromal_pathways <- c("Moffitt.F13_NormalStroma.top250",
                              "Moffitt.F5_ActivatedStroma.top250")
# pdacr
res <- fgsea_run_batch(de_stroma, gs_pdacr[moffitt_stromal_pathways], 
                       fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)

# hallmark
res <- fgsea_run_batch(de_stroma, gs_hallmark, fgsea_eps=0, 
                       fgsea_nperms=gsea_nperms, fgsea_nproc=10)
gsea <- bind_rows(gsea, res)
gsea_stroma <- gsea

```


### GSEA plots

```{r stromal gsea plots}

# stats to report
gsea_stroma %>% arrange(padj) %>% select(!c(leadingEdge, dir)) %>% filter(pathway == "Moffitt.F5_ActivatedStroma.top250")
gsea_stroma %>% arrange(padj) %>% select(!c(leadingEdge, dir)) %>% filter(pathway == "Moffitt.F13_NormalStroma.top250")

# contrasts
contrasts <- c("clusterLRvNL", "clusterHRvNL", "clusterHRvLR")

# prepare gsea results
gsea <- gsea_stroma %>%
  filter(contrast %in% contrasts) %>%
  mutate(padj = 2 * padj) %>%
  mutate(contrast = factor(contrast, levels=contrasts),
         neglog10padj = ifelse(padj < gsea_padj_cutoff, -log10(padj), NA),
         sig = padj < gsea_padj_cutoff) %>%
  group_by(contrast, pathway) %>%
  slice_min(padj, n=1, with_ties=FALSE) %>%
  ungroup()

#
# Moffitt (main figure)
#
res <- gsea %>% filter(pathway %in% moffitt_stromal_pathways)

# arrange for readability
pathway_levels <- res %>% filter(contrast == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=contrast, y=NES, fill=neglog10padj)) +
  geom_col() + 
  geom_hline(yintercept=0) +
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.x = element_text(color = "black", angle = 90, vjust = 0.5, hjust=1)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ pathway, nrow=1) + 
  labs(x="GSEA Analysis", y="Normalized Enrichment Score (NES)")
f <- paste0("duke_stroma_gsea_barplot_clusters_moffitt")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=3, height=4)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=3, height=4)

# 
# Hallmark
#
res <- gsea %>% filter(pathway %in% names(gs_hallmark)) %>% 
  mutate(pathway = gsub("HALLMARK_", "", pathway))
# only keep significant results (at least one contrast)
res <- res %>% group_by(pathway) %>%
  filter(any(padj < gsea_padj_cutoff)) %>%
  ungroup()

# arrange by clusterHRvLR for readability
pathway_levels <- filter(res, contrast == "clusterHRvLR") %>% arrange(NES) %>% pull(pathway)
res$pathway <- factor(res$pathway, levels=pathway_levels)

p <- ggplot(res, aes(x=NES, y=pathway, fill=neglog10padj)) +
  geom_col() + 
  scale_fill_viridis_c() +
  theme_bw() + 
  theme(axis.text.y = element_text(color="black", size=7)) +
  theme(strip.text = element_text(size=8)) +
  facet_wrap(~ contrast, nrow=1) + 
  labs(x="Normalized Enrichment Score (NES)", y="Pathway")
f <- paste0("duke_stroma_gsea_barplot_clusters_hallmark")
ggsave(file.path(plot_dir, paste0(f, ".pdf")), p, width=6, height=2)
ggsave(file.path(plot_dir, paste0(f, ".png")), p, width=6, height=2)


```



## GSVA

```{r duke stromal gsva analysis}

y <- fds_duke_stroma$log2_norm_cpm
s <- bind_cols(fds_duke_stroma$samples, pca_res.stroma$x)
gs <- c(gs_cptac, gs_hpa_panc, gs_pdacr, gs_hallmark) 

# run gsva
x <- gsva(GSVA::gsvaParam(y, gs))
rownames(x) <- gsub("HALLMARK_", "H_", rownames(x))
rownames(x) <- gsub("HPA_PROGNOSIS_", "HPA_", rownames(x))
res <- x

moffitt_color_scales = list(
  "Moffitt.F13_NormalStroma.top250" = "#006400",
  "Moffitt.F5_ActivatedStroma.top250" = "#b5651d"
)

path_aoi_shapes = list(
  "nl"=21,
  "lgd"=25,
  "hgd"=24,
  "inv"=23
)

x <- as_tibble(res[moffitt_stromal_pathways,], rownames="pathway") %>%
  pivot_longer(cols = !pathway, names_to="aoi")
x <- left_join(x, s, by=join_by(aoi == aoi))

# discard "NA" regions
x <- x %>% filter(cluster_name != "cNA")

x$path_aoi <- factor(x$path_aoi, levels=c("nl", "lgd", "hgd", "inv"))
x$cluster_name <- factor(x$cluster_name, levels=c("cNL", "cLR", "cHR"))
#x$cluster_name <- factor(x$cluster_name, levels=c("cNA", "cNL", "cLR", "cHR"))
x$pathway <- factor(x$pathway, levels=moffitt_stromal_pathways)

p <- ggplot(x, aes(x=cluster_name, y=value)) + 
  geom_boxplot(outlier.shape = NA, width=0.75, alpha=0.5) +
  geom_point(aes(group=path_aoi, fill=path_aoi, shape=path_aoi), size=2, color="black",
             position = position_jitterdodge(jitter.width=0, dodge.width=0.5)) +
  geom_hline(yintercept=0) +
  scale_shape_manual(values = path_aoi_shapes) +
  scale_fill_manual(values = color_scales$path_aoi) +
  facet_grid(pathway ~ path_specimen + patient, scales="free", space="free") + 
  theme_bw()
f <- file.path(plot_dir, "duke_stroma_boxplot_patient_gsva_pdacr_subtypes")
ggsave(paste0(f, ".pdf"), p, width=8, height=7)
ggsave(paste0(f, ".png"), p, width=8, height=7)


#
# PCA plot
#
p <- ggplot(x, aes(x=PC1, y=PC2, color=value, shape=path_specimen)) +
  geom_point(alpha = 0.8, size=3) +
  scale_color_viridis_c() +
  scale_shape_manual(values = shape_scales$path_specimen) +
  theme_minimal() +
  theme(legend.text = element_text(size=7)) +
  labs(x="PC1", y="PC2", title="Moffitt Subtypes") + 
  facet_wrap(~ pathway, nrow=1)
f <- file.path(plot_dir, "duke_stroma_pca_featureplot_moffitt")
ggsave(paste0(f, ".pdf"), plot=p, width=9, height=3.5)
ggsave(paste0(f, ".png"), plot=p, width=9, height=3.5)

#
# boxplots
#
p1 <- ggplot(x, aes(x=path, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75, linewidth=0.5) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Pathology", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "duke_stroma_boxplot_gsva_moffitt_path")
ggsave(paste0(f, ".pdf"), p1, width=5, height=4)

p2 <- ggplot(x, aes(x=cluster_name, y=value, fill=pathway)) +
  geom_boxplot(outlier.shape=NA, width=0.75) +
  geom_point(position=position_jitterdodge(jitter.width=0.1),
             aes(group=pathway), pch=21, size=1, alpha=0.8) +
  scale_fill_manual(values = moffitt_color_scales) +
  labs(x="Cluster", y="Gene Set Score (GSVA)") +
  theme_bw() + 
  theme(axis.text.x = element_text(size=6, color="black", angle=90, vjust=0.5, hjust=1),
        axis.line = element_line(color = "black"),
        legend.text = element_text(size=6))
f <- file.path(plot_dir, "duke_stroma_boxplot_gsva_moffitt_cluster")
ggsave(paste0(f, ".pdf"), p2, width=4, height=4)

p <- p2 + p1 + patchwork::plot_layout(guides="collect", widths=c(1,2))
f <- file.path(plot_dir, "duke_stroma_boxplot_gsva_moffitt_path_cluster")
ggsave(paste0(f, ".pdf"), p, width=6, height=4)




```


